{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "35024d94-1087-4640-af06-d89f349f5601",
   "metadata": {},
   "outputs": [],
   "source": [
    "backup_dir = '/root/datos/maestria/netopaas/luca_explore/surgeries'\n",
    "name = 'Deng'\n",
    "cell_key = 'cell_type_adjusted'\n",
    "integ_features = 3000"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a14891aa-a7c0-4bc9-b05d-8be55a5ad016",
   "metadata": {},
   "source": [
    "Based on SCTransform [v2 vignette](https://satijalab.org/seurat/archive/v4.3/sctransform_v2_vignette)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "719a61ad-eb17-4580-a9c1-945cece78a9b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_23522/3724497324.py:4: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  anndata2ri.activate()\n"
     ]
    }
   ],
   "source": [
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import anndata2ri\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6c29c50a-91ff-4180-aa01-dbb36386b2ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata  = ad.read_h5ad(f'{backup_dir}/{name}/query.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "208e9d07-da7e-4efb-b40e-a95060035b34",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>percent.mt</th>\n",
       "      <th>assigned_cell_type</th>\n",
       "      <th>stage</th>\n",
       "      <th>Ecotype</th>\n",
       "      <th>ecoRecovery</th>\n",
       "      <th>Patient Number</th>\n",
       "      <th>Group</th>\n",
       "      <th>...</th>\n",
       "      <th>T stage</th>\n",
       "      <th>N stage</th>\n",
       "      <th>M stage</th>\n",
       "      <th>Pathological stage</th>\n",
       "      <th>cell_type_tumor</th>\n",
       "      <th>dataset</th>\n",
       "      <th>sample</th>\n",
       "      <th>_scvi_batch</th>\n",
       "      <th>_scvi_labels</th>\n",
       "      <th>cell_type_adjusted</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P001</td>\n",
       "      <td>4348.0</td>\n",
       "      <td>2099</td>\n",
       "      <td>12.531686</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>P001</td>\n",
       "      <td>pGGO</td>\n",
       "      <td>...</td>\n",
       "      <td>T1a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA1</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>P001</td>\n",
       "      <td>P001</td>\n",
       "      <td>12</td>\n",
       "      <td>51</td>\n",
       "      <td>Mast cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P001</td>\n",
       "      <td>5701.0</td>\n",
       "      <td>2614</td>\n",
       "      <td>4.821479</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>CD8.T.cells_S04</td>\n",
       "      <td>P001</td>\n",
       "      <td>pGGO</td>\n",
       "      <td>...</td>\n",
       "      <td>T1a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA1</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>P001</td>\n",
       "      <td>P001</td>\n",
       "      <td>12</td>\n",
       "      <td>51</td>\n",
       "      <td>T cell CD8 effector memory</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P001</td>\n",
       "      <td>5636.0</td>\n",
       "      <td>2489</td>\n",
       "      <td>2.784740</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>CD4.T.cells_S04</td>\n",
       "      <td>P001</td>\n",
       "      <td>pGGO</td>\n",
       "      <td>...</td>\n",
       "      <td>T1a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA1</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>P001</td>\n",
       "      <td>P001</td>\n",
       "      <td>12</td>\n",
       "      <td>51</td>\n",
       "      <td>T cell CD4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P001</td>\n",
       "      <td>5991.0</td>\n",
       "      <td>2591</td>\n",
       "      <td>6.061365</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>CD4.T.cells_S05</td>\n",
       "      <td>P001</td>\n",
       "      <td>pGGO</td>\n",
       "      <td>...</td>\n",
       "      <td>T1a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA1</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>P001</td>\n",
       "      <td>P001</td>\n",
       "      <td>12</td>\n",
       "      <td>51</td>\n",
       "      <td>T cell CD8 activated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P001</td>\n",
       "      <td>3464.0</td>\n",
       "      <td>1625</td>\n",
       "      <td>9.162717</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Mast.cells_S01</td>\n",
       "      <td>P001</td>\n",
       "      <td>pGGO</td>\n",
       "      <td>...</td>\n",
       "      <td>T1a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA1</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>P001</td>\n",
       "      <td>P001</td>\n",
       "      <td>12</td>\n",
       "      <td>51</td>\n",
       "      <td>Mast cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421706</th>\n",
       "      <td>P049</td>\n",
       "      <td>1497.0</td>\n",
       "      <td>723</td>\n",
       "      <td>0.407106</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>P049</td>\n",
       "      <td>SolidN</td>\n",
       "      <td>...</td>\n",
       "      <td>T1b</td>\n",
       "      <td>N2</td>\n",
       "      <td>M0</td>\n",
       "      <td>IIIA</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>P049</td>\n",
       "      <td>P049</td>\n",
       "      <td>60</td>\n",
       "      <td>51</td>\n",
       "      <td>Macrophage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421707</th>\n",
       "      <td>P049</td>\n",
       "      <td>7093.0</td>\n",
       "      <td>541</td>\n",
       "      <td>0.027082</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>P049</td>\n",
       "      <td>SolidN</td>\n",
       "      <td>...</td>\n",
       "      <td>T1b</td>\n",
       "      <td>N2</td>\n",
       "      <td>M0</td>\n",
       "      <td>IIIA</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>P049</td>\n",
       "      <td>P049</td>\n",
       "      <td>60</td>\n",
       "      <td>51</td>\n",
       "      <td>Plasma cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421708</th>\n",
       "      <td>P049</td>\n",
       "      <td>5584.0</td>\n",
       "      <td>2466</td>\n",
       "      <td>5.115492</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>P049</td>\n",
       "      <td>SolidN</td>\n",
       "      <td>...</td>\n",
       "      <td>T1b</td>\n",
       "      <td>N2</td>\n",
       "      <td>M0</td>\n",
       "      <td>IIIA</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>P049</td>\n",
       "      <td>P049</td>\n",
       "      <td>60</td>\n",
       "      <td>51</td>\n",
       "      <td>B cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421709</th>\n",
       "      <td>P049</td>\n",
       "      <td>5451.0</td>\n",
       "      <td>2348</td>\n",
       "      <td>9.361350</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>CD8.T.cells_S04</td>\n",
       "      <td>CD8.T.cells_S04</td>\n",
       "      <td>P049</td>\n",
       "      <td>SolidN</td>\n",
       "      <td>...</td>\n",
       "      <td>T1b</td>\n",
       "      <td>N2</td>\n",
       "      <td>M0</td>\n",
       "      <td>IIIA</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>P049</td>\n",
       "      <td>P049</td>\n",
       "      <td>60</td>\n",
       "      <td>51</td>\n",
       "      <td>T cell CD8 effector memory</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421710</th>\n",
       "      <td>P049</td>\n",
       "      <td>4361.0</td>\n",
       "      <td>2007</td>\n",
       "      <td>6.144832</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>P049</td>\n",
       "      <td>SolidN</td>\n",
       "      <td>...</td>\n",
       "      <td>T1b</td>\n",
       "      <td>N2</td>\n",
       "      <td>M0</td>\n",
       "      <td>IIIA</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>P049</td>\n",
       "      <td>P049</td>\n",
       "      <td>60</td>\n",
       "      <td>51</td>\n",
       "      <td>B cell</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>421711 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       orig.ident  nCount_RNA  nFeature_RNA  percent.mt  assigned_cell_type  \\\n",
       "index                                                                         \n",
       "0            P001      4348.0          2099   12.531686                   5   \n",
       "1            P001      5701.0          2614    4.821479                   0   \n",
       "2            P001      5636.0          2489    2.784740                   0   \n",
       "3            P001      5991.0          2591    6.061365                   0   \n",
       "4            P001      3464.0          1625    9.162717                   5   \n",
       "...           ...         ...           ...         ...                 ...   \n",
       "421706       P049      1497.0           723    0.407106                   1   \n",
       "421707       P049      7093.0           541    0.027082                   4   \n",
       "421708       P049      5584.0          2466    5.115492                   4   \n",
       "421709       P049      5451.0          2348    9.361350                   0   \n",
       "421710       P049      4361.0          2007    6.144832                   4   \n",
       "\n",
       "        stage          Ecotype      ecoRecovery Patient Number   Group  ...  \\\n",
       "index                                                                   ...   \n",
       "0           1       unassigned       unassigned           P001    pGGO  ...   \n",
       "1           1       unassigned  CD8.T.cells_S04           P001    pGGO  ...   \n",
       "2           1       unassigned  CD4.T.cells_S04           P001    pGGO  ...   \n",
       "3           1       unassigned  CD4.T.cells_S05           P001    pGGO  ...   \n",
       "4           1       unassigned   Mast.cells_S01           P001    pGGO  ...   \n",
       "...       ...              ...              ...            ...     ...  ...   \n",
       "421706      9       unassigned       unassigned           P049  SolidN  ...   \n",
       "421707      9       unassigned       unassigned           P049  SolidN  ...   \n",
       "421708      9       unassigned       unassigned           P049  SolidN  ...   \n",
       "421709      9  CD8.T.cells_S04  CD8.T.cells_S04           P049  SolidN  ...   \n",
       "421710      9       unassigned       unassigned           P049  SolidN  ...   \n",
       "\n",
       "       T stage  N stage M stage Pathological stage cell_type_tumor  dataset  \\\n",
       "index                                                                         \n",
       "0          T1a       N0      M0                IA1         Unknown     P001   \n",
       "1          T1a       N0      M0                IA1         Unknown     P001   \n",
       "2          T1a       N0      M0                IA1         Unknown     P001   \n",
       "3          T1a       N0      M0                IA1         Unknown     P001   \n",
       "4          T1a       N0      M0                IA1         Unknown     P001   \n",
       "...        ...      ...     ...                ...             ...      ...   \n",
       "421706     T1b       N2      M0               IIIA         Unknown     P049   \n",
       "421707     T1b       N2      M0               IIIA         Unknown     P049   \n",
       "421708     T1b       N2      M0               IIIA         Unknown     P049   \n",
       "421709     T1b       N2      M0               IIIA         Unknown     P049   \n",
       "421710     T1b       N2      M0               IIIA         Unknown     P049   \n",
       "\n",
       "       sample _scvi_batch _scvi_labels          cell_type_adjusted  \n",
       "index                                                               \n",
       "0        P001          12           51                   Mast cell  \n",
       "1        P001          12           51  T cell CD8 effector memory  \n",
       "2        P001          12           51                  T cell CD4  \n",
       "3        P001          12           51        T cell CD8 activated  \n",
       "4        P001          12           51                   Mast cell  \n",
       "...       ...         ...          ...                         ...  \n",
       "421706   P049          60           51                  Macrophage  \n",
       "421707   P049          60           51                 Plasma cell  \n",
       "421708   P049          60           51                      B cell  \n",
       "421709   P049          60           51  T cell CD8 effector memory  \n",
       "421710   P049          60           51                      B cell  \n",
       "\n",
       "[421711 rows x 28 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "preds = pd.read_csv(f'{backup_dir}/{name}_predicted_leiden.csv')\n",
    "preds.index = adata.obs.index\n",
    "adata.obs[cell_key] = preds[cell_key]\n",
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b70018a0-98da-40b3-a3e5-1a6f3338fb17",
   "metadata": {},
   "outputs": [],
   "source": [
    "%Rdevice svg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5e86d8e5-d45d-4b7e-9ec6-8dadd3590de3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    },
    {
     "data": {
      "text/plain": [
       "Loading required package: SeuratObject\n",
       "Loading required package: sp\n",
       "\n",
       "Attaching package: ‘sp’\n",
       "\n",
       "The following object is masked from ‘package:IRanges’:\n",
       "\n",
       "    %over%\n",
       "\n",
       "\n",
       "Attaching package: ‘SeuratObject’\n",
       "\n",
       "The following object is masked from ‘package:SummarizedExperiment’:\n",
       "\n",
       "    Assays\n",
       "\n",
       "The following object is masked from ‘package:GenomicRanges’:\n",
       "\n",
       "    intersect\n",
       "\n",
       "The following object is masked from ‘package:GenomeInfoDb’:\n",
       "\n",
       "    intersect\n",
       "\n",
       "The following object is masked from ‘package:IRanges’:\n",
       "\n",
       "    intersect\n",
       "\n",
       "The following object is masked from ‘package:S4Vectors’:\n",
       "\n",
       "    intersect\n",
       "\n",
       "The following object is masked from ‘package:BiocGenerics’:\n",
       "\n",
       "    intersect\n",
       "\n",
       "The following objects are masked from ‘package:base’:\n",
       "\n",
       "    intersect, t\n",
       "\n",
       "\n",
       "Attaching package: ‘Seurat’\n",
       "\n",
       "The following object is masked from ‘package:SummarizedExperiment’:\n",
       "\n",
       "    Assays\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i adata -i backup_dir -i name\n",
    "library(Seurat)\n",
    "adata.seurat <- as.Seurat(adata, data = \"X\", counts='X')\n",
    "SaveSeuratRds(adata.seurat, file.path(backup_dir,name,\"query.rds\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8f1abdb-3282-48d0-b4e6-46ca4bbaa23d",
   "metadata": {},
   "source": [
    "SCTransform processing: 2008.358 sec elapsed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbec834a-302b-4ca4-b394-187733415d6e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  |                                                  | 0 % ~calculating   |++                                                | 2 % ~19s           |+++                                               | 4 % ~26s           |++++                                              | 6 % ~24s           |+++++                                             | 8 % ~23s           |++++++                                            | 10% ~20s           |+++++++                                           | 12% ~18s           |++++++++                                          | 14% ~18s           |+++++++++                                         | 16% ~18s           |++++++++++                                        | 18% ~18s           |+++++++++++                                       | 20% ~18s           |++++++++++++                                      | 22% ~17s           |+++++++++++++                                     | 24% ~17s           |++++++++++++++                                    | 27% ~17s           |+++++++++++++++                                   | 29% ~16s           |++++++++++++++++                                  | 31% ~15s           |+++++++++++++++++                                 | 33% ~15s           |++++++++++++++++++                                | 35% ~14s           |+++++++++++++++++++                               | 37% ~14s           |++++++++++++++++++++                              | 39% ~13s           |+++++++++++++++++++++                             | 41% ~12s           |++++++++++++++++++++++                            | 43% ~12s           |+++++++++++++++++++++++                           | 45% ~12s           |++++++++++++++++++++++++                          | 47% ~12s           |+++++++++++++++++++++++++                         | 49% ~12s           |++++++++++++++++++++++++++                        | 51% ~12s           |+++++++++++++++++++++++++++                       | 53% ~11s           |++++++++++++++++++++++++++++                      | 55% ~11s           |+++++++++++++++++++++++++++++                     | 57% ~10s           |++++++++++++++++++++++++++++++                    | 59% ~10s           |+++++++++++++++++++++++++++++++                   | 61% ~09s           |++++++++++++++++++++++++++++++++                  | 63% ~09s           |+++++++++++++++++++++++++++++++++                 | 65% ~09s           |++++++++++++++++++++++++++++++++++                | 67% ~08s           |+++++++++++++++++++++++++++++++++++               | 69% ~08s           |++++++++++++++++++++++++++++++++++++              | 71% ~07s           |+++++++++++++++++++++++++++++++++++++             | 73% ~07s           |++++++++++++++++++++++++++++++++++++++            | 76% ~06s           |+++++++++++++++++++++++++++++++++++++++           | 78% ~06s           |++++++++++++++++++++++++++++++++++++++++          | 80% ~05s           |+++++++++++++++++++++++++++++++++++++++++         | 82% ~05s           |++++++++++++++++++++++++++++++++++++++++++        | 84% ~04s           |+++++++++++++++++++++++++++++++++++++++++++       | 86% ~04s           |++++++++++++++++++++++++++++++++++++++++++++      | 88% ~03s           |+++++++++++++++++++++++++++++++++++++++++++++     | 90% ~03s           |++++++++++++++++++++++++++++++++++++++++++++++    | 92% ~02s           |+++++++++++++++++++++++++++++++++++++++++++++++   | 94% ~02s           |++++++++++++++++++++++++++++++++++++++++++++++++  | 96% ~01s           |+++++++++++++++++++++++++++++++++++++++++++++++++ | 98% ~01s           |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=28s  \n",
      "  |                                                  | 0 % ~calculating   |+                                                 | 1 % ~06d 07h 14m 50s |++                                                | 2 % ~05d 16h 08m 07s |++                                                | 3 % ~05d 08h 59m 37s |+++                                               | 4 % ~05d 12h 14m 01s |+++                                               | 5 % ~05d 06h 22m 07s |++++                                              | 6 % ~05d 04h 01m 52s |++++                                              | 7 % ~05d 00h 12m 01s |+++++                                             | 8 % ~04d 20h 29m 02s |+++++                                             | 9 % ~04d 20h 12m 01s |++++++                                            | 10% ~04d 19h 04m 31s |++++++                                            | 11% ~04d 15h 47m 08s |+++++++                                           | 12% ~04d 13h 59m 48s |+++++++                                           | 13% ~04d 13h 19m 41s |++++++++                                          | 14% ~04d 11h 27m 04s |++++++++                                          | 15% ~04d 08h 08m 44s |+++++++++                                         | 16% ~04d 05h 36m 57s |+++++++++                                         | 17% ~04d 05h 10m 38s |++++++++++                                        | 18% ~04d 05h 43m 17s |++++++++++                                        | 19% ~04d 07h 11m 10s |+++++++++++                                       | 20% ~04d 06h 51m 51s |+++++++++++                                       | 21% ~04d 06h 09m 38s |++++++++++++                                      | 22% ~04d 07h 15m 56s |++++++++++++                                      | 23% ~04d 08h 19m 29s |+++++++++++++                                     | 24% ~04d 05h 33m 08s |+++++++++++++                                     | 26% ~04d 03h 13m 03s |++++++++++++++                                    | 27% ~04d 01h 43m 27s |++++++++++++++                                    | 28% ~04d 00h 39m 40s |+++++++++++++++                                   | 29% ~03d 22h 39m 08s |+++++++++++++++                                   | 30% ~03d 20h 53m 57s |++++++++++++++++                                  | 31% ~03d 19h 10m 03s |++++++++++++++++                                  | 32% ~03d 17h 27m 25s |+++++++++++++++++                                 | 33% ~03d 16h 03m 06s |+++++++++++++++++                                 | 34% ~03d 14h 27m 29s |++++++++++++++++++                                | 35% ~03d 13h 39m 27s |++++++++++++++++++                                | 36% ~03d 12h 17m 23s"
     ]
    }
   ],
   "source": [
    "%%R -i integ_features\n",
    "library(Seurat)\n",
    "library(dplyr)\n",
    "library(purrr)\n",
    "library(tictoc)\n",
    "\n",
    "# Start the timer\n",
    "tic(\"Total execution time\")\n",
    "\n",
    "# Split and transform the Seurat objects\n",
    "# tic(\"SCTransform processing\")\n",
    "# adata.seurat.list <- adata.seurat %>%\n",
    "#   SplitObject(split.by = \"sample\") %>%\n",
    "#   map(~ SCTransform(.x, vst.flavor = \"v2\", verbose = FALSE, assay='originalexp', return.only.var.genes = FALSE)) # Important to return all genes or else integration does not work\n",
    "# toc()  # Ends SCTransform timing\n",
    "\n",
    "# Proceed with further analysis (e.g., integration)\n",
    "tic(\"Integration processing\")\n",
    "features <- SelectIntegrationFeatures(object.list = adata.seurat.list, nfeatures = integ_features)\n",
    "\n",
    "adata.seurat.list <- PrepSCTIntegration(object.list = adata.seurat.list, anchor.features = features)\n",
    "anchors <- FindIntegrationAnchors(\n",
    "  object.list = adata.seurat.list,\n",
    "  normalization.method = \"SCT\",\n",
    "  anchor.features = features\n",
    ")\n",
    "integrated_data <- IntegrateData(\n",
    "  anchorset = anchors,\n",
    "  normalization.method = \"SCT\"\n",
    ")\n",
    "toc()  # Ends Integration timing\n",
    "\n",
    "toc()  # Ends total execution timing\n",
    "integrated_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "612c0662-fadd-4ad9-bd31-25b467888446",
   "metadata": {},
   "source": [
    "### Import new annots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a169452-6605-4578-9b53-55929e19046f",
   "metadata": {},
   "outputs": [],
   "source": [
    "2+2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "cd7ba570-b21a-46ec-9d4b-6609bb277e5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "preds = pd.read_csv(f'{backup_dir}/{name}_predicted_leiden.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "580a8ffc-0399-4cb3-a84b-5f242268d35a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%R -i preds\n",
    "integrated_data@meta.data$cell_type_adjusted <- preds$cell_type_adjusted"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17893966-98b9-4afb-8e8a-0f4d6908fd63",
   "metadata": {},
   "source": [
    "## Markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf57707d-44c1-4bb9-b9c9-4ff18a0e2a5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "tic()\n",
    "integrated_data <- PrepSCTFindMarkers(integrated_data)\n",
    "integrated_data@meta.data$cell_type_adjusted <- as.factor(integrated_data@meta.data$cell_type_adjusted)\n",
    "Idents(object=integrated_data) <- 'cell_type_adjusted'\n",
    "markers <- FindMarkers(object = integrated_data, ident.1 = \"Tumor LUAD_mixed\", assay='SCT', recorrect_umi=F) # We dont recollect UMI because we use UMI counts the SCRAssay method\n",
    "print(head(x = markers, n=15))\n",
    "toc()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "604b17db-3cc0-43e6-b8bd-db9c4a0cda03",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
