{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a80560b2-00f4-4d4f-9d13-2b5c48e057d0",
   "metadata": {},
   "source": [
    "# Download the scRNA data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "32151548-2c85-4ced-887e-2523ab68a156",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = '/root/datos/maestria/netopaas/'\n",
    "backup_dir = '/root/datos/maestria/netopaas/luca_explore/surgeries'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8a4a47b6-e24d-4170-9d4a-e3566be33eae",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_331915/2172773568.py:25: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  anndata2ri.activate()\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import utils.functions as ut\n",
    "import urllib\n",
    "\n",
    "import scanpy as sc\n",
    "\n",
    "import rpy2.rinterface_lib.callbacks\n",
    "import logging\n",
    "from rpy2.robjects import pandas2ri\n",
    "import rpy2.robjects as robjects\n",
    "\n",
    "# Set the graphics device\n",
    "robjects.r('options(bitmapType=\"cairo\")')\n",
    "import anndata2ri\n",
    "\n",
    "# Ignore R warning messages\n",
    "#Note: this can be commented out to get more verbose R output\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)\n",
    "\n",
    "# Automatically convert rpy2 outputs to pandas dataframes\n",
    "pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68ea9660-e216-414c-aca4-11143ca39a33",
   "metadata": {},
   "source": [
    "## Zuani 2024"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a0630a4-3ef6-40a0-aa9d-f6f4abd79cbf",
   "metadata": {},
   "source": [
    "### Download"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "75568c3d-ab43-450d-a95d-035081923f24",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/root/datos/maestria/netopaas/'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4f59dd44-4668-45e2-a0c9-3bae3f4e5a76",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "E-MTAB-13526.idf.txt\tP17_T2-barcodes.tsv.gz\tP22_T1-matrix.mtx.gz\n",
      "E-MTAB-13526.sdrf.txt\tP17_T2-features.tsv.gz\tP23_T1-barcodes.tsv.gz\n",
      "EmptyDrop_P15_T2-.sh\tP17_T2-matrix.mtx.gz\tP23_T1-features.tsv.gz\n",
      "EmptyDrop_P16_T2-.sh\tP17_T3-barcodes.tsv.gz\tP23_T1-matrix.mtx.gz\n",
      "EmptyDrop_P17_T2-.sh\tP17_T3-features.tsv.gz\tP24_T1-barcodes.tsv.gz\n",
      "EmptyDrop_P17_T3-.sh\tP17_T3-matrix.mtx.gz\tP24_T1-features.tsv.gz\n",
      "EmptyDrop_P18_T2-.sh\tP18_T2-barcodes.tsv.gz\tP24_T1-matrix.mtx.gz\n",
      "EmptyDrop_P19_T2-.sh\tP18_T2-features.tsv.gz\tP4_T2-barcodes.tsv.gz\n",
      "EmptyDrop_P20_T2-.sh\tP18_T2-matrix.mtx.gz\tP4_T2-features.tsv.gz\n",
      "EmptyDrop_P21_T1-.sh\tP19_T2-barcodes.tsv.gz\tP4_T2-matrix.mtx.gz\n",
      "EmptyDrop_P21_T2-.sh\tP19_T2-features.tsv.gz\tP4_T3-barcodes.tsv.gz\n",
      "EmptyDrop_P22_T1-.sh\tP19_T2-matrix.mtx.gz\tP4_T3-features.tsv.gz\n",
      "EmptyDrop_P23_T1-.sh\tP20_T2-barcodes.tsv.gz\tP4_T3-matrix.mtx.gz\n",
      "EmptyDrop_P24_T1-.sh\tP20_T2-features.tsv.gz\tP8_T2-barcodes.tsv.gz\n",
      "EmptyDrop_P4_T2-.sh\tP20_T2-matrix.mtx.gz\tP8_T2-features.tsv.gz\n",
      "EmptyDrop_P4_T3-.sh\tP21_T1-barcodes.tsv.gz\tP8_T2-matrix.mtx.gz\n",
      "EmptyDrop_P8_T2-.sh\tP21_T1-features.tsv.gz\tZuani2024.h5ad\n",
      "P15_T2-barcodes.tsv.gz\tP21_T1-matrix.mtx.gz\tdrop_plots\n",
      "P15_T2-features.tsv.gz\tP21_T2-barcodes.tsv.gz\tftp_magetables.sh\n",
      "P15_T2-matrix.mtx.gz\tP21_T2-features.tsv.gz\tmetadata.xlsx\n",
      "P16_T2-barcodes.tsv.gz\tP21_T2-matrix.mtx.gz\toutputEmptyDrops\n",
      "P16_T2-features.tsv.gz\tP22_T1-barcodes.tsv.gz\n",
      "P16_T2-matrix.mtx.gz\tP22_T1-features.tsv.gz\n"
     ]
    }
   ],
   "source": [
    "! ls {data_dir}/Zuani2024"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "77a1cbc9-e3cb-4cdf-9e76-28d3802cc92f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Source Name</th>\n",
       "      <th>Comment[ENA_SAMPLE]</th>\n",
       "      <th>Comment[BioSD_SAMPLE]</th>\n",
       "      <th>Characteristics[organism]</th>\n",
       "      <th>Characteristics[individual]</th>\n",
       "      <th>Characteristics[original source name]</th>\n",
       "      <th>Characteristics[age]</th>\n",
       "      <th>Unit[time unit]</th>\n",
       "      <th>Term Source REF</th>\n",
       "      <th>Term Accession Number</th>\n",
       "      <th>...</th>\n",
       "      <th>Derived Array Data File</th>\n",
       "      <th>Protocol REF.6</th>\n",
       "      <th>Protocol REF.7</th>\n",
       "      <th>Derived Array Data File.1</th>\n",
       "      <th>Protocol REF.8</th>\n",
       "      <th>Protocol REF.9</th>\n",
       "      <th>Derived Array Data File.2</th>\n",
       "      <th>Factor Value[disease]</th>\n",
       "      <th>Factor Value[FACS]</th>\n",
       "      <th>Factor Value[sampling site]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P1_B1</td>\n",
       "      <td>ERS16703054</td>\n",
       "      <td>SAMEA114591031</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 1</td>\n",
       "      <td>TB18.0655</td>\n",
       "      <td>73</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P1_B1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-features.tsv.gz</td>\n",
       "      <td>non-small cell lung cancer</td>\n",
       "      <td>MDSC</td>\n",
       "      <td>normal tissue adjacent to tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P1_B1</td>\n",
       "      <td>ERS16703054</td>\n",
       "      <td>SAMEA114591031</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 1</td>\n",
       "      <td>TB18.0655</td>\n",
       "      <td>73</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P1_B1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-features.tsv.gz</td>\n",
       "      <td>non-small cell lung cancer</td>\n",
       "      <td>MDSC</td>\n",
       "      <td>normal tissue adjacent to tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P1_B1</td>\n",
       "      <td>ERS16703054</td>\n",
       "      <td>SAMEA114591031</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 1</td>\n",
       "      <td>TB18.0655</td>\n",
       "      <td>73</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P1_B1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-features.tsv.gz</td>\n",
       "      <td>non-small cell lung cancer</td>\n",
       "      <td>MDSC</td>\n",
       "      <td>normal tissue adjacent to tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P1_B1</td>\n",
       "      <td>ERS16703054</td>\n",
       "      <td>SAMEA114591031</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 1</td>\n",
       "      <td>TB18.0655</td>\n",
       "      <td>73</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P1_B1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-features.tsv.gz</td>\n",
       "      <td>non-small cell lung cancer</td>\n",
       "      <td>MDSC</td>\n",
       "      <td>normal tissue adjacent to tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P1_B1</td>\n",
       "      <td>ERS16703054</td>\n",
       "      <td>SAMEA114591031</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 1</td>\n",
       "      <td>TB18.0655</td>\n",
       "      <td>73</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P1_B1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-features.tsv.gz</td>\n",
       "      <td>non-small cell lung cancer</td>\n",
       "      <td>MDSC</td>\n",
       "      <td>normal tissue adjacent to tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1562</th>\n",
       "      <td>P18_T1</td>\n",
       "      <td>ERS16703118</td>\n",
       "      <td>SAMEA114591092</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 18</td>\n",
       "      <td>TB21.0006</td>\n",
       "      <td>77</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P18_T1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-features.tsv.gz</td>\n",
       "      <td>lung squamous cell carcinoma</td>\n",
       "      <td>CD45+</td>\n",
       "      <td>tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1563</th>\n",
       "      <td>P18_T1</td>\n",
       "      <td>ERS16703118</td>\n",
       "      <td>SAMEA114591092</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 18</td>\n",
       "      <td>TB21.0006</td>\n",
       "      <td>77</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P18_T1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-features.tsv.gz</td>\n",
       "      <td>lung squamous cell carcinoma</td>\n",
       "      <td>CD45+</td>\n",
       "      <td>tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1564</th>\n",
       "      <td>P18_T1</td>\n",
       "      <td>ERS16703118</td>\n",
       "      <td>SAMEA114591092</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 18</td>\n",
       "      <td>TB21.0006</td>\n",
       "      <td>77</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P18_T1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-features.tsv.gz</td>\n",
       "      <td>lung squamous cell carcinoma</td>\n",
       "      <td>CD45+</td>\n",
       "      <td>tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1565</th>\n",
       "      <td>P18_T1</td>\n",
       "      <td>ERS16703118</td>\n",
       "      <td>SAMEA114591092</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 18</td>\n",
       "      <td>TB21.0006</td>\n",
       "      <td>77</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P18_T1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-features.tsv.gz</td>\n",
       "      <td>lung squamous cell carcinoma</td>\n",
       "      <td>CD45+</td>\n",
       "      <td>tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1566</th>\n",
       "      <td>P18_T1</td>\n",
       "      <td>ERS16703118</td>\n",
       "      <td>SAMEA114591092</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 18</td>\n",
       "      <td>TB21.0006</td>\n",
       "      <td>77</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P18_T1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-features.tsv.gz</td>\n",
       "      <td>lung squamous cell carcinoma</td>\n",
       "      <td>CD45+</td>\n",
       "      <td>tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1567 rows × 69 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Source Name Comment[ENA_SAMPLE] Comment[BioSD_SAMPLE]  \\\n",
       "0          P1_B1         ERS16703054        SAMEA114591031   \n",
       "1          P1_B1         ERS16703054        SAMEA114591031   \n",
       "2          P1_B1         ERS16703054        SAMEA114591031   \n",
       "3          P1_B1         ERS16703054        SAMEA114591031   \n",
       "4          P1_B1         ERS16703054        SAMEA114591031   \n",
       "...          ...                 ...                   ...   \n",
       "1562      P18_T1         ERS16703118        SAMEA114591092   \n",
       "1563      P18_T1         ERS16703118        SAMEA114591092   \n",
       "1564      P18_T1         ERS16703118        SAMEA114591092   \n",
       "1565      P18_T1         ERS16703118        SAMEA114591092   \n",
       "1566      P18_T1         ERS16703118        SAMEA114591092   \n",
       "\n",
       "     Characteristics[organism] Characteristics[individual]  \\\n",
       "0                 Homo sapiens                   Patient 1   \n",
       "1                 Homo sapiens                   Patient 1   \n",
       "2                 Homo sapiens                   Patient 1   \n",
       "3                 Homo sapiens                   Patient 1   \n",
       "4                 Homo sapiens                   Patient 1   \n",
       "...                        ...                         ...   \n",
       "1562              Homo sapiens                  Patient 18   \n",
       "1563              Homo sapiens                  Patient 18   \n",
       "1564              Homo sapiens                  Patient 18   \n",
       "1565              Homo sapiens                  Patient 18   \n",
       "1566              Homo sapiens                  Patient 18   \n",
       "\n",
       "     Characteristics[original source name]  Characteristics[age]  \\\n",
       "0                                TB18.0655                    73   \n",
       "1                                TB18.0655                    73   \n",
       "2                                TB18.0655                    73   \n",
       "3                                TB18.0655                    73   \n",
       "4                                TB18.0655                    73   \n",
       "...                                    ...                   ...   \n",
       "1562                             TB21.0006                    77   \n",
       "1563                             TB21.0006                    77   \n",
       "1564                             TB21.0006                    77   \n",
       "1565                             TB21.0006                    77   \n",
       "1566                             TB21.0006                    77   \n",
       "\n",
       "     Unit[time unit] Term Source REF Term Accession Number  ...  \\\n",
       "0               year             EFO            UO_0000036  ...   \n",
       "1               year             EFO            UO_0000036  ...   \n",
       "2               year             EFO            UO_0000036  ...   \n",
       "3               year             EFO            UO_0000036  ...   \n",
       "4               year             EFO            UO_0000036  ...   \n",
       "...              ...             ...                   ...  ...   \n",
       "1562            year             EFO            UO_0000036  ...   \n",
       "1563            year             EFO            UO_0000036  ...   \n",
       "1564            year             EFO            UO_0000036  ...   \n",
       "1565            year             EFO            UO_0000036  ...   \n",
       "1566            year             EFO            UO_0000036  ...   \n",
       "\n",
       "     Derived Array Data File Protocol REF.6 Protocol REF.7  \\\n",
       "0        P1_B1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1        P1_B1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "2        P1_B1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "3        P1_B1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "4        P1_B1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "...                      ...            ...            ...   \n",
       "1562    P18_T1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1563    P18_T1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1564    P18_T1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1565    P18_T1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1566    P18_T1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "\n",
       "     Derived Array Data File.1 Protocol REF.8 Protocol REF.9  \\\n",
       "0        P1_B1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1        P1_B1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "2        P1_B1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "3        P1_B1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "4        P1_B1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "...                        ...            ...            ...   \n",
       "1562    P18_T1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1563    P18_T1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1564    P18_T1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1565    P18_T1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1566    P18_T1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "\n",
       "     Derived Array Data File.2         Factor Value[disease]  \\\n",
       "0        P1_B1-features.tsv.gz    non-small cell lung cancer   \n",
       "1        P1_B1-features.tsv.gz    non-small cell lung cancer   \n",
       "2        P1_B1-features.tsv.gz    non-small cell lung cancer   \n",
       "3        P1_B1-features.tsv.gz    non-small cell lung cancer   \n",
       "4        P1_B1-features.tsv.gz    non-small cell lung cancer   \n",
       "...                        ...                           ...   \n",
       "1562    P18_T1-features.tsv.gz  lung squamous cell carcinoma   \n",
       "1563    P18_T1-features.tsv.gz  lung squamous cell carcinoma   \n",
       "1564    P18_T1-features.tsv.gz  lung squamous cell carcinoma   \n",
       "1565    P18_T1-features.tsv.gz  lung squamous cell carcinoma   \n",
       "1566    P18_T1-features.tsv.gz  lung squamous cell carcinoma   \n",
       "\n",
       "     Factor Value[FACS]      Factor Value[sampling site]  \n",
       "0                  MDSC  normal tissue adjacent to tumor  \n",
       "1                  MDSC  normal tissue adjacent to tumor  \n",
       "2                  MDSC  normal tissue adjacent to tumor  \n",
       "3                  MDSC  normal tissue adjacent to tumor  \n",
       "4                  MDSC  normal tissue adjacent to tumor  \n",
       "...                 ...                              ...  \n",
       "1562              CD45+                            tumor  \n",
       "1563              CD45+                            tumor  \n",
       "1564              CD45+                            tumor  \n",
       "1565              CD45+                            tumor  \n",
       "1566              CD45+                            tumor  \n",
       "\n",
       "[1567 rows x 69 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zuani_dir = f'{data_dir}/Zuani2024'\n",
    "zuani_samples = pd.read_csv(f'{zuani_dir}/E-MTAB-13526.sdrf.txt', sep='\\t')\n",
    "\n",
    "# Filter for non enriched MDSC = Myeloid derived supressor cells, CD235a- filters for erythrocites\n",
    "zuani_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "546d01dd-5a3c-4eb4-9a3a-e2f5447ce1e5",
   "metadata": {},
   "source": [
    "We filter for non-immune enriched samples and only tumor tissue. All the non-tumor enirhced ones are filtered for no plasma cells (erythrocites CD235a-)\n",
    "that is also the case in many of the other samples... we can consider the eryhrocites enriched depleted, for they are only one cell type that is not very present. But we should impute some abundance values in these samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4fb1b2dd-900d-4ca5-adf7-e0c6c8ba8d3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "facs_filter = ['CD235a-']\n",
    "zuani_samples = zuani_samples[zuani_samples['Factor Value[FACS]'].isin(facs_filter)]\n",
    "\n",
    "zuani_samples = zuani_samples[zuani_samples['Factor Value[sampling site]']=='tumor']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3b25edd8-8c34-42df-b8bd-9e0d9bf9695d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Source Name', 'Comment[ENA_SAMPLE]', 'Comment[BioSD_SAMPLE]',\n",
       "       'Characteristics[organism]', 'Characteristics[individual]',\n",
       "       'Characteristics[original source name]', 'Characteristics[age]',\n",
       "       'Unit[time unit]', 'Term Source REF', 'Term Accession Number',\n",
       "       'Characteristics[developmental stage]', 'Characteristics[sex]',\n",
       "       'Characteristics[organism part]', 'Characteristics[disease]',\n",
       "       'Characteristics[FACS]', 'Characteristics[sampling site]',\n",
       "       'Characteristics[tumor grading]', 'Description', 'Material Type',\n",
       "       'Protocol REF', 'Protocol REF.1', 'Protocol REF.2', 'Extract Name',\n",
       "       'Comment[LIBRARY_LAYOUT]', 'Comment[LIBRARY_SELECTION]',\n",
       "       'Comment[LIBRARY_SOURCE]', 'Comment[LIBRARY_STRATEGY]',\n",
       "       'Comment[cdna read]', 'Comment[cdna read offset]',\n",
       "       'Comment[cdna read size]', 'Comment[cell barcode offset]',\n",
       "       'Comment[cell barcode read]', 'Comment[cell barcode size]',\n",
       "       'Comment[end bias]', 'Comment[input molecule]',\n",
       "       'Comment[library construction]', 'Comment[primer]',\n",
       "       'Comment[sample barcode offset]', 'Comment[sample barcode read]',\n",
       "       'Comment[sample barcode size]', 'Comment[single cell isolation]',\n",
       "       'Comment[spike in]', 'Comment[umi barcode offset]',\n",
       "       'Comment[umi barcode read]', 'Comment[umi barcode size]',\n",
       "       'Protocol REF.3', 'Performer', 'Assay Name',\n",
       "       'Comment[technical replicate group]', 'Technology Type',\n",
       "       'Comment[ENA_EXPERIMENT]', 'Scan Name', 'Comment[SUBMITTED_FILE_NAME]',\n",
       "       'Comment[ENA_RUN]', 'Comment[FASTQ_URI]', 'Comment[READ_TYPE]',\n",
       "       'Comment[READ_INDEX]', 'Protocol REF.4', 'Protocol REF.5',\n",
       "       'Derived Array Data File', 'Protocol REF.6', 'Protocol REF.7',\n",
       "       'Derived Array Data File.1', 'Protocol REF.8', 'Protocol REF.9',\n",
       "       'Derived Array Data File.2', 'Factor Value[disease]',\n",
       "       'Factor Value[FACS]', 'Factor Value[sampling site]'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zuani_samples.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1240c3c0-076c-40ae-95a5-2a3538ad5c44",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['T3N0M0', 'T2bN0M0', nan, 'T2aN0M0', 'T4N2M0', 'T2aN1M0', 'T3M0N0',\n",
       "       'T2bN1M0', 'T1cN0M0', 'T4N0'], dtype=object)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zuani_samples['Characteristics[tumor grading]'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "6f188cc0-e8b4-433d-bc20-13054845bb2a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPUS:  8\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P1_T1-barcodes.tsv.gz\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P1_T1-features.tsv.gz\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P1_T1-matrix.mtx.gz\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P2_T1-barcodes.tsv.gz\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P2_T1-features.tsv.gz\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P2_T1-matrix.mtx.gz\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P15_T2-barcodes.tsv.gz\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P15_T2-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P2_T1-features.tsv.gz time (s): 1.5921189785003662\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P15_T2-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P1_T1-features.tsv.gz time (s): 1.5972809791564941\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P16_T2-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P15_T2-features.tsv.gz time (s): 1.3004364967346191\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P16_T2-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P1_T1-barcodes.tsv.gz time (s): 6.099670171737671\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P16_T2-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P16_T2-features.tsv.gz time (s): 1.1689391136169434\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T2-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P2_T1-matrix.mtx.gz time (s): 25.24450135231018\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T2-features.tsv.gzurl: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P2_T1-barcodes.tsv.gz time (s): 43.2615282535553\n",
      "\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T2-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P15_T2-barcodes.tsv.gz time (s): 43.600990295410156\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T3-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T2-features.tsv.gz time (s): 1.2455544471740723\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T3-features.tsv.gzurl: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P16_T2-barcodes.tsv.gz time (s): 42.82427263259888\n",
      "\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T3-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T3-features.tsv.gz time (s): 1.2398083209991455\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T2-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P1_T1-matrix.mtx.gz time (s): 52.44864463806152\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T2-features.tsv.gzurl: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T2-barcodes.tsv.gz time (s): 42.84986734390259\n",
      "\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T2-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T2-features.tsv.gz time (s): 1.2251718044281006\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T3-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T3-barcodes.tsv.gz time (s): 42.83008527755737\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T3-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T2-barcodes.tsv.gz time (s): 42.71306538581848\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T3-matrix.mtx.gzurl: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T3-features.tsv.gz time (s): 1.2472662925720215\n",
      "\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P20_T2-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T2-matrix.mtx.gz time (s): 76.8129460811615\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P20_T2-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T3-barcodes.tsv.gz time (s): 42.92999482154846\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P20_T2-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P20_T2-features.tsv.gz time (s): 1.2489855289459229\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P18_T2-barcodes.tsv.gzurl: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T3-matrix.mtx.gz time (s): 95.85074806213379\n",
      "\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P18_T2-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P16_T2-matrix.mtx.gz time (s): 142.04853892326355\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P18_T2-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P18_T2-features.tsv.gz time (s): 1.438136339187622\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P19_T2-barcodes.tsv.gzurl: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P20_T2-barcodes.tsv.gz time (s): 42.95699501037598\n",
      "\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P19_T2-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P18_T2-barcodes.tsv.gz time (s): 42.92334604263306\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P19_T2-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P19_T2-features.tsv.gz time (s): 1.2489831447601318\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P24_T1-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P19_T2-barcodes.tsv.gz time (s): 42.852856397628784\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P24_T1-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P15_T2-matrix.mtx.gz time (s): 210.2394998073578\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P24_T1-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P24_T1-features.tsv.gz time (s): 1.2681214809417725\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P22_T1-barcodes.tsv.gzurl: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P24_T1-barcodes.tsv.gz time (s): 42.93040895462036\n",
      "\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P22_T1-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P22_T1-barcodes.tsv.gz time (s): 42.95659422874451\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P22_T1-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P22_T1-features.tsv.gz time (s): 1.2519481182098389\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P8_T2-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P22_T1-matrix.mtx.gz time (s): 83.76424765586853\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P8_T2-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P8_T2-barcodes.tsv.gz time (s): 43.164130449295044\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P8_T2-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P8_T2-features.tsv.gz time (s): 1.2520792484283447\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T1-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P20_T2-matrix.mtx.gz time (s): 340.315548658371\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T1-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P19_T2-matrix.mtx.gz time (s): 318.12532663345337\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T1-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T1-features.tsv.gz time (s): 1.2496387958526611\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P23_T1-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T1-barcodes.tsv.gz time (s): 43.02663064002991\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P23_T1-features.tsv.gzurl: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P18_T2-matrix.mtx.gz time (s): 405.18058800697327\n",
      "\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P23_T1-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P23_T1-features.tsv.gz time (s): 1.375727891921997\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T2-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P23_T1-barcodes.tsv.gz time (s): 42.89959907531738\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T2-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T2-matrix.mtx.gz time (s): 505.055460691452\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T2-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T2-features.tsv.gz time (s): 1.2500417232513428\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T2-matrix.mtx.gz time (s): 23.810145139694214\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T2-barcodes.tsv.gz time (s): 43.953601360321045\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T1-matrix.mtx.gz time (s): 133.2835156917572\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T3-matrix.mtx.gz time (s): 639.7812283039093\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P24_T1-matrix.mtx.gz time (s): 703.8568208217621\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P8_T2-matrix.mtx.gz time (s): 500.3829493522644\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P23_T1-matrix.mtx.gz time (s): 617.2215695381165\n"
     ]
    }
   ],
   "source": [
    "base_url = 'https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files'\n",
    "zuani_patients = zuani_samples['Source Name'].unique()\n",
    "\n",
    "urls_paths = []\n",
    "\n",
    "for pat in zuani_patients:\n",
    "    filename = f'{pat}-barcodes.tsv.gz'\n",
    "    urls_paths.append((f'{base_url}/{filename}',f'{zuani_dir}/{filename}'))\n",
    "\n",
    "    filename = f'{pat}-features.tsv.gz'\n",
    "    urls_paths.append((f'{base_url}/{filename}',f'{zuani_dir}/{filename}'))\n",
    "\n",
    "    filename = f'{pat}-matrix.mtx.gz'\n",
    "    urls_paths.append((f'{base_url}/{filename}',f'{zuani_dir}/{filename}'))\n",
    "\n",
    "ut.download_parallel(urls_paths, cpus=8)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d26fa5c-b686-4ce3-8afd-bf8b7c7d363a",
   "metadata": {},
   "source": [
    "The authors from Zuani alos offer an nanotated dataset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "224811f2-c3d6-4373-928a-8c2e6d5533c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "bash: line 8: warning: here-document at line 3 delimited by end-of-file (wanted `EOF')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Connected to hl-xfer-public.ebi.ac.uk.\n",
      "220-Welcome to ftp.ebi.ac.uk\n",
      "220 \n",
      "230 Login successful.\n",
      "Remote system type is UNIX.\n",
      "Using binary mode to transfer files.\n",
      "250 Directory successfully changed.\n",
      "200 Switching to Binary mode.\n",
      "local: 10X_Lung_Healthy_Background_Annotated_v2.h5ad remote: 10X_Lung_Healthy_Background_Annotated_v2.h5ad\n",
      "229 Entering Extended Passive Mode (|||7072|)\n",
      "150 Opening BINARY mode data connection for 10X_Lung_Healthy_Background_Annotated_v2.h5ad (45505872062 bytes).\n",
      "\n",
      "receive aborted. Waiting for remote to finish abort.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd /root/datos/maestria/netopaas/Zuani2024/\n",
    "wget https://www.ebi.ac.uk/biostudies/files/E-MTAB-13526/10X_Lung_Tumour_Annotated_v2.h5ad\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c81a40fc-32fd-431e-bced-f64fb48c7561",
   "metadata": {},
   "source": [
    "### Remove empty drops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b8300901-e215-4229-9d48-815a53481a4a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /root/datos/maestria/netopaas/Zuani2024//EmptyDrop_P16_T2-.sh\n",
      "[1] \"/root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P16_T2-\"\n",
      "[1] \"Creating folder  /root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P16_T2-\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /root/datos/maestria/netopaas/Zuani2024//EmptyDrop_P17_T2-.sh\n",
      "[1] \"/root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P17_T2-\"\n",
      "[1] \"Creating folder  /root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P17_T2-\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /root/datos/maestria/netopaas/Zuani2024//EmptyDrop_P17_T3-.sh\n",
      "[1] \"/root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P17_T3-\"\n",
      "[1] \"Creating folder  /root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P17_T3-\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /root/datos/maestria/netopaas/Zuani2024//EmptyDrop_P18_T2-.sh\n",
      "[1] \"/root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P18_T2-\"\n",
      "[1] \"Creating folder  /root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P18_T2-\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /root/datos/maestria/netopaas/Zuani2024//EmptyDrop_P19_T2-.sh\n",
      "[1] \"/root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P19_T2-\"\n",
      "[1] \"Creating folder  /root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P19_T2-\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /root/datos/maestria/netopaas/Zuani2024//EmptyDrop_P20_T2-.sh\n",
      "[1] \"/root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P20_T2-\"\n",
      "[1] \"Creating folder  /root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P20_T2-\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /root/datos/maestria/netopaas/Zuani2024//EmptyDrop_P21_T1-.sh\n",
      "[1] \"/root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P21_T1-\"\n",
      "[1] \"Creating folder  /root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P21_T1-\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /root/datos/maestria/netopaas/Zuani2024//EmptyDrop_P21_T2-.sh\n",
      "[1] \"/root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P21_T2-\"\n",
      "[1] \"Creating folder  /root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P21_T2-\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /root/datos/maestria/netopaas/Zuani2024//EmptyDrop_P22_T1-.sh\n",
      "[1] \"/root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P22_T1-\"\n",
      "[1] \"Creating folder  /root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P22_T1-\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /root/datos/maestria/netopaas/Zuani2024//EmptyDrop_P23_T1-.sh\n",
      "[1] \"/root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P23_T1-\"\n",
      "[1] \"Creating folder  /root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P23_T1-\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /root/datos/maestria/netopaas/Zuani2024//EmptyDrop_P24_T1-.sh\n",
      "[1] \"/root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P24_T1-\"\n",
      "[1] \"Creating folder  /root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P24_T1-\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /root/datos/maestria/netopaas/Zuani2024//EmptyDrop_P4_T2-.sh\n",
      "[1] \"/root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P4_T2-\"\n",
      "[1] \"Creating folder  /root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P4_T2-\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /root/datos/maestria/netopaas/Zuani2024//EmptyDrop_P4_T3-.sh\n",
      "[1] \"/root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P4_T3-\"\n",
      "[1] \"Creating folder  /root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P4_T3-\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /root/datos/maestria/netopaas/Zuani2024//EmptyDrop_P8_T2-.sh\n",
      "[1] \"/root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P8_T2-\"\n",
      "[1] \"Creating folder  /root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P8_T2-\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from os import system as sys\n",
    "import os\n",
    "\n",
    "codeDir = '/root/host_home/luca/utils/' # CHANGE ME TO THE DIRECTORY WHERE run_EmptyDrop.R IS CLONED\n",
    "directory = '/root/datos/maestria/netopaas/Zuani2024/' # change me to the dir where the matrix files are stored\n",
    "\n",
    "# All CellRanger samples for the lung data\n",
    "samples = []\n",
    "\n",
    "file_pattern = sorted([f for f in os.listdir(directory) if 'matrix' in f])\n",
    "for mtx_file in file_pattern:\n",
    "    idx = mtx_file.replace('matrix.mtx.gz', '')\n",
    "    samples.append(idx)\n",
    "\n",
    "for sample in samples:\n",
    "    fname = directory+\"/EmptyDrop_\"+sample+\".sh\"\n",
    "    f = open(f'{fname}', 'w')\n",
    "    f.write(\"cd /root/datos/maestria/netopaas/Zuani2024/\\n\") # CHANGE ME TO A PLACE YOU WOULD LIKE TO WRITE OUTPUTS TO\n",
    "    # f.write(\"conda activate minimal_env\\n\" )\n",
    "    f.write(\"Rscript \" + codeDir + \"run_EmptyDrop.R {0}\\n\".format(str(sample)))\n",
    "    f.close()\n",
    "\n",
    "    # Example batch calling on the codon LSF cluster (this will vary depending on the system, bsub vs. qsub etc)\n",
    "    print(f'bash {fname}')\n",
    "    sys(f'bash {fname}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c69ad9fb-439e-4150-87ee-5e2dfd53bfa9",
   "metadata": {},
   "source": [
    "### Convert to adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "2c1b051d-478a-4a99-8c16-71d810eaedd9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 318313 × 33538\n",
      "    obs: 'sample'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:1897: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import io\n",
    "import csv\n",
    "import gzip\n",
    "\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "\n",
    "def concatenate_datasets(directory):\n",
    "    \"\"\"\n",
    "    Concatenate multiple .mtx datasets from a single directory into a single AnnData object.\n",
    "    \n",
    "    Parameters:\n",
    "    - directory (str): The directory containing .mtx, features.tsv, and barcodes.tsv files.\n",
    "    \n",
    "    Returns:\n",
    "    - AnnData: The concatenated AnnData object.\n",
    "    \"\"\"\n",
    "    adatas = {}\n",
    "    features_file = ''\n",
    "    \n",
    "    # List all subdirectories in the main directory\n",
    "    subdirectories = [d for d in os.listdir(directory) if os.path.isdir(os.path.join(directory, d))]\n",
    "    \n",
    "    for subdirectory in subdirectories:\n",
    "        # Construct the path to the subdirectory\n",
    "        subdirectory_path = os.path.join(directory, subdirectory)\n",
    "        \n",
    "        # Read the .mtx file and associated files from the subdirectory\n",
    "        adata = sc.read_10x_mtx(subdirectory_path, var_names='gene_symbols', cache=True)\n",
    "        \n",
    "        # Use the subdirectory name as the identifier\n",
    "        idx = subdirectory\n",
    "        \n",
    "        # Store the AnnData object in the dictionary\n",
    "        adatas[idx] = adata\n",
    "    \n",
    "    # Concatenate all AnnData objects, labeling them by the subdirectory name which is used as the sample identifier\n",
    "    concatenated_adata = ad.concat(adatas, label='sample')\n",
    "\n",
    "    return concatenated_adata\n",
    "\n",
    "# Example usage:\n",
    "adata = concatenate_datasets(f'{zuani_dir}/outputEmptyDrops')\n",
    "print(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9c3bc81f-38f9-43a8-aa35-3ca50f24a8a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata as ad\n",
    "adata2 = ad.read_h5ad('/root/datos/maestria/netopaas/lung_scRNA/LUCA_model/full_atlas_hvg_integrated_scvi_integrated_scanvi.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "554b896e-6147-48c4-8462-72941b6c192d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Alveolar cell type 1', 'Alveolar cell type 2', 'B cell',\n",
       "       'B cell dividing', 'Ciliated', 'Club', 'DC mature',\n",
       "       'Endothelial cell arterial', 'Endothelial cell capillary',\n",
       "       'Endothelial cell lymphatic', 'Endothelial cell venous',\n",
       "       'Fibroblast adventitial', 'Fibroblast alveolar',\n",
       "       'Fibroblast peribronchial', 'Macrophage', 'Macrophage alveolar',\n",
       "       'Mast cell', 'Mesothelial', 'Monocyte classical',\n",
       "       'Monocyte non-classical', 'NK cell', 'NK cell dividing', 'Neutrophils',\n",
       "       'Pericyte', 'Plasma cell', 'Plasma cell dividing',\n",
       "       'ROS1+ healthy epithelial', 'Smooth muscle cell', 'T cell CD4',\n",
       "       'T cell CD4 dividing', 'T cell CD8 activated', 'T cell CD8 dividing',\n",
       "       'T cell CD8 effector memory', 'T cell CD8 naive',\n",
       "       'T cell CD8 terminally exhausted', 'T cell NK-like',\n",
       "       'T cell regulatory', 'Tumor cells LUAD', 'Tumor cells LUAD EMT',\n",
       "       'Tumor cells LUAD MSLN', 'Tumor cells LUAD NE',\n",
       "       'Tumor cells LUAD mitotic', 'Tumor cells LUSC',\n",
       "       'Tumor cells LUSC mitotic', 'Tumor cells NSCLC mixed', 'cDC1', 'cDC2',\n",
       "       'myeloid dividing', 'pDC', 'stromal dividing', 'transitional club/AT2'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata2.obs.cell_type_tumor.cat.categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "dd8a6b40-c97d-4f8e-a844-4bae0cabf068",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write_h5ad(f'{zuani_dir}/Zuani2024.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "c2be4836-f9cd-4974-8062-3ece65db64d5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:1897: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    }
   ],
   "source": [
    "adata = sc.read_h5ad('/root/datos/maestria/netopaas/Zuani2024/Zuani.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8658d30-c55c-4dfe-9393-309590c3a618",
   "metadata": {},
   "source": [
    "Before emptydrops with FDR 0.001"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "68d711ff-e6ee-4694-b1f0-6dd6e821019c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 101923200 × 33538\n",
       "    obs: 'sample'\n",
       "    var: 'gene_symbols'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad9ea544-821f-4b5c-9307-47900d3d8c87",
   "metadata": {},
   "source": [
    "after EmptyDrops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "c2af13e1-e9be-462e-bfbd-1a0d9de00133",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 312502 × 33538\n",
       "    obs: 'sample', 'barcode', 'Source Name', 'Characteristics[tumor grading]'"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "be47ea48-abc2-4c6c-a3a9-89a859d9df63",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:850: UserWarning: \n",
      "AnnData expects .obs.index to contain strings, but got values like:\n",
      "    [0, 1, 2, 3, 4]\n",
      "\n",
      "    Inferred to be: integer\n",
      "\n",
      "  value_idx = self._prep_dim_index(value.index, attr)\n"
     ]
    }
   ],
   "source": [
    "adata.obs['sample'] = adata.obs['sample'].str[:-1]\n",
    "zuani_samples_unique = zuani_samples.drop_duplicates(subset=['Source Name'], keep='first')\n",
    "adata.obs = pd.merge(adata.obs, zuani_samples_unique.loc[:, ['Source Name', 'Characteristics[tumor grading]']],\n",
    "                            left_on='sample', right_on='Source Name', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "7bb01c65-bb61-4320-8f17-763ad9a254fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "def map_tnm_to_roman(tnm_stage):\n",
    "    stage_mappings = {\n",
    "        r\"^TisN0M0$\": \"0\",\n",
    "        r\"^T1[ab]?N0M0$\": \"I\",\n",
    "        r\"^T2[ab]?N0M0$\": \"II\",\n",
    "        r\"^T2[ab]?N0/1$\": \"II\",\n",
    "        r\"^T1[ab]?N1M0$\": \"II\",\n",
    "        r\"^T2[ab]?N1M0$\": \"II\",\n",
    "        r\"^T3N[01]M0$\": \"II\",\n",
    "        r\"^T1[ab]?N2M0$\": \"III\",\n",
    "        r\"^T2[ab]?N2M0$\": \"III\",\n",
    "        r\"^T3N2M0$\": \"III\",\n",
    "        r\"^T4N[012]M0$\": \"III\",\n",
    "        r\"^T4N0$\": \"III\",\n",
    "        r\"^T[1-4][ab]?N3M0$\": \"III\",\n",
    "        r\"^T[1-4][ab]?N[0-3]M1[abc]$\": \"IV\",\n",
    "    }\n",
    "    \n",
    "    for pattern, stage in stage_mappings.items():\n",
    "        if re.match(pattern, tnm_stage):\n",
    "            return stage\n",
    "    return \"Unknown stage\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c604c393-8970-4674-984e-fd5ef239fb86",
   "metadata": {},
   "source": [
    "Remove the patients with no tumor grading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "5a37a31f-f664-464d-aaa1-2d3635a8f260",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[~adata.obs['Characteristics[tumor grading]'].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72831dc7-6b78-4ce8-ba1b-8b5a3b94225e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "adata.obs['stage'] = adata.obs['Characteristics[tumor grading]'].apply(map_tnm_to_roman)\n",
    "# adata.obs.stage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "7c619493-37e6-410f-8d1a-f2a9b89c7ef5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:1292: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:1897: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:1897: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:1292: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:1897: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:1897: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:1292: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:1897: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:1897: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:1292: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:1897: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:1897: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    }
   ],
   "source": [
    "adata.write_h5ad(f'{zuani_dir}/Zuani2024.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8999abd6-a111-4106-be8f-7ab6d1a81968",
   "metadata": {},
   "source": [
    "## Deng 2024"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93f4066a-b712-4504-afad-319d866e1e65",
   "metadata": {},
   "outputs": [],
   "source": [
    "# urllib.request.urlretrieve('https://figshare.com/ndownloader/files/44695465',f'{data_dir}/Deng2024/dengluad.rds')\n",
    "urllib.request.urlretrieve('https://figshare.com/ndownloader/files/46617667',f'{data_dir}/Deng2024/dengEcoTyper.rds')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "adfd4c7f-5ede-43c0-b5c3-fe57c8446b81",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ! rm /root/datos/maestria/netopaas/Deng2024/deng.h5Seurat\n",
    "# ! rm /root/datos/maestria/netopaas/Deng2024/deng.h5ad\n",
    "# ! rm /root/datos/maestria/netopaas/Deng2024/dengRNA.h5ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "2710b1b5-0cf7-4701-99f1-366384940f58",
   "metadata": {
    "editable": true,
    "scrolled": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Validating object structure\n",
       "Updating object slots\n",
       "Ensuring keys are in the proper structure\n",
       "Updating matrix keys for DimReduc ‘pca’\n",
       "Updating matrix keys for DimReduc ‘umap’\n",
       "Updating matrix keys for DimReduc ‘tsne’\n",
       "Warning: Assay RNA changing from Assay to Assay\n",
       "Warning: Assay integrated changing from Assay to Assay\n",
       "Warning: Graph integrated_nn changing from Graph to Graph\n",
       "Warning: Graph integrated_snn changing from Graph to Graph\n",
       "Warning: DimReduc pca changing from DimReduc to DimReduc\n",
       "Warning: DimReduc umap changing from DimReduc to DimReduc\n",
       "Warning: DimReduc tsne changing from DimReduc to DimReduc\n",
       "Ensuring keys are in the proper structure\n",
       "Ensuring feature names don't have underscores or pipes\n",
       "Updating slots in RNA\n",
       "Updating slots in integrated\n",
       "Updating slots in integrated_nn\n",
       "Setting default assay of integrated_nn to integrated\n",
       "Updating slots in integrated_snn\n",
       "Setting default assay of integrated_snn to integrated\n",
       "Updating slots in pca\n",
       "Updating slots in umap\n",
       "Setting umap DimReduc to global\n",
       "Updating slots in tsne\n",
       "Setting tsne DimReduc to global\n",
       "No assay information could be found for FindIntegrationAnchors\n",
       "No assay information could be found for IntegrateData\n",
       "Setting assay used for ScaleData.integrated to integrated\n",
       "Setting assay used for RunPCA.integrated to integrated\n",
       "Setting assay used for RunUMAP.integrated.pca to integrated\n",
       "No assay information could be found for RunTSNE\n",
       "Setting assay used for FindNeighbors.integrated.pca to integrated\n",
       "No assay information could be found for FindClusters\n",
       "Validating object structure for Assay ‘RNA’\n",
       "Validating object structure for Assay ‘integrated’\n",
       "Validating object structure for Graph ‘integrated_nn’\n",
       "Validating object structure for Graph ‘integrated_snn’\n",
       "Validating object structure for DimReduc ‘pca’\n",
       "Validating object structure for DimReduc ‘umap’\n",
       "Validating object structure for DimReduc ‘tsne’\n",
       "Object representation is consistent with the most current Seurat version\n",
       "Creating h5Seurat file for version 3.1.5.9900\n",
       "Adding counts for RNA\n",
       "Adding data for RNA\n",
       "No variable features found for RNA\n",
       "No feature-level metadata found for RNA\n",
       "Adding data for integrated\n",
       "Adding scale.data for integrated\n",
       "Adding variable features for integrated\n",
       "No feature-level metadata found for integrated\n",
       "Adding cell embeddings for pca\n",
       "Adding loadings for pca\n",
       "No projected loadings for pca\n",
       "Adding standard deviations for pca\n",
       "No JackStraw data for pca\n",
       "Adding cell embeddings for umap\n",
       "No loadings for umap\n",
       "No projected loadings for umap\n",
       "No standard deviations for umap\n",
       "No JackStraw data for umap\n",
       "Adding cell embeddings for tsne\n",
       "No loadings for tsne\n",
       "No projected loadings for tsne\n",
       "No standard deviations for tsne\n",
       "No JackStraw data for tsne\n",
       "Validating h5Seurat file\n",
       "Adding scale.data from integrated as X\n",
       "Adding data from integrated as raw\n",
       "Transfering meta.data to obs\n",
       "Adding dimensional reduction information for pca\n",
       "Adding feature loadings for pca\n",
       "Adding dimensional reduction information for tsne\n",
       "Adding dimensional reduction information for umap\n",
       "Validating h5Seurat file\n",
       "Adding data from RNA as X\n",
       "Adding counts from RNA as raw\n",
       "Transfering meta.data to obs\n",
       "Adding dimensional reduction information for tsne (global)\n",
       "Adding dimensional reduction information for umap (global)\n",
       "In addition: Warning messages:\n",
       "1: Adding a command log without an assay associated with it \n",
       "2: Adding a command log without an assay associated with it \n",
       "3: Adding a command log without an assay associated with it \n",
       "4: Adding a command log without an assay associated with it \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i data_dir\n",
    "library(Seurat)\n",
    "library(SeuratDisk)\n",
    "\n",
    "deng <- readRDS(paste0(data_dir, '/Deng2024/dengEcoTyper.rds'))\n",
    "deng = UpdateSeuratObject(deng)\n",
    "\n",
    "# After P49 there is no metadata so better delete. Also the ones with N as they are healthy controls,\n",
    "cells_to_exclude <- Cells(deng)[grepl(\"P05|N\", deng@meta.data$orig.ident)]\n",
    "deng <- subset(deng, cells = setdiff(Cells(deng), cells_to_exclude))\n",
    "\n",
    "\n",
    "SaveH5Seurat(deng, filename=paste0(data_dir, '/Deng2024/deng.h5Seurat'))\n",
    "Convert(paste0(data_dir, '/Deng2024/deng.h5Seurat'), dest = \"h5ad\")\n",
    "Convert(paste0(data_dir, '/Deng2024/deng.h5Seurat'), dest = paste0(data_dir, '/Deng2024/dengRNA.h5ad'), assay = 'RNA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "3c6cd572-3a8a-403b-9328-fac4b9fa0386",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " [1] \"P001\" \"P002\" \"P003\" \"P004\" \"P005\" \"P006\" \"P007\" \"P008\" \"P009\" \"P010\"\n",
      "[11] \"P011\" \"P012\" \"P013\" \"P014\" \"P015\" \"P016\" \"P017\" \"P018\" \"P019\" \"P020\"\n",
      "[21] \"P021\" \"P022\" \"P023\" \"P024\" \"P025\" \"P026\" \"P027\" \"P028\" \"P029\" \"P030\"\n",
      "[31] \"P031\" \"P032\" \"P033\" \"P034\" \"P035\" \"P036\" \"P037\" \"P038\" \"P039\" \"P040\"\n",
      "[41] \"P041\" \"P042\" \"P043\" \"P044\" \"P045\" \"P046\" \"P047\" \"P048\" \"P049\"\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "unique(deng@meta.data$orig.ident)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "6ff2ee1a-a284-48bd-8238-0a0c899b3542",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:767: UserWarning: \n",
      "AnnData expects .obs.index to contain strings, but got values like:\n",
      "    [0, 1, 2, 3, 4]\n",
      "\n",
      "    Inferred to be: integer\n",
      "\n",
      "  value_idx = self._prep_dim_index(value.index, attr)\n"
     ]
    }
   ],
   "source": [
    "adata_integ = sc.read_h5ad(f'{data_dir}/Deng2024/deng.h5ad')\n",
    "adata = sc.read_h5ad(f'{data_dir}/Deng2024/dengRNA.h5ad')\n",
    "\n",
    "# The assignment fo slots and assays is not very good with Convert from SeuratDisk so we make some arrangements\n",
    "# We also arrange the X so that it is like the other filtered datasets in the repo\n",
    "adata.layers['data'] = adata.X.copy()\n",
    "adata.X = adata.raw.X\n",
    "\n",
    "adata.raw = adata_integ\n",
    "\n",
    "path_metadata = f'{data_dir}/Deng2024/metadata.xlsx'\n",
    "if not os.path.exists(path_metadata):\n",
    "    urllib.request.urlretrieve('https://www.ncbi.nlm.nih.gov/pmc/articles/PMC11031428/bin/mmc2.xlsx', path_metadata)\n",
    "xl_deng = pd.ExcelFile(path_metadata)\n",
    "metada_deng = xl_deng.parse(' lung cancer in scRNAseq')\n",
    "adata.obs = pd.merge(adata.obs, metada_deng, how='left', left_on='orig.ident', right_on='Patient Number')\n",
    "\n",
    "\n",
    "adata.write_h5ad(f'{backup_dir}/filtered_Deng_Liu_LUAD_2024.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "d25e8766-c554-4299-8fd1-94e2a0a9bd1e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>percent.mt</th>\n",
       "      <th>assigned_cell_type</th>\n",
       "      <th>stage</th>\n",
       "      <th>Ecotype</th>\n",
       "      <th>ecoRecovery</th>\n",
       "      <th>Patient Number</th>\n",
       "      <th>Group</th>\n",
       "      <th>...</th>\n",
       "      <th>Current or former smoking</th>\n",
       "      <th>Cancer history</th>\n",
       "      <th>Tumor location</th>\n",
       "      <th>Tumor size（cm）</th>\n",
       "      <th>Pathology</th>\n",
       "      <th>Histological subtype</th>\n",
       "      <th>T stage</th>\n",
       "      <th>N stage</th>\n",
       "      <th>M stage</th>\n",
       "      <th>Pathological stage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P001</td>\n",
       "      <td>4348.0</td>\n",
       "      <td>2099</td>\n",
       "      <td>12.531686</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>P001</td>\n",
       "      <td>pGGO</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>LUL</td>\n",
       "      <td>0.9</td>\n",
       "      <td>ADC</td>\n",
       "      <td>MIA</td>\n",
       "      <td>T1a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P001</td>\n",
       "      <td>5701.0</td>\n",
       "      <td>2614</td>\n",
       "      <td>4.821479</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>CD8.T.cells_S04</td>\n",
       "      <td>P001</td>\n",
       "      <td>pGGO</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>LUL</td>\n",
       "      <td>0.9</td>\n",
       "      <td>ADC</td>\n",
       "      <td>MIA</td>\n",
       "      <td>T1a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P001</td>\n",
       "      <td>5636.0</td>\n",
       "      <td>2489</td>\n",
       "      <td>2.784740</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>CD4.T.cells_S04</td>\n",
       "      <td>P001</td>\n",
       "      <td>pGGO</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>LUL</td>\n",
       "      <td>0.9</td>\n",
       "      <td>ADC</td>\n",
       "      <td>MIA</td>\n",
       "      <td>T1a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P001</td>\n",
       "      <td>5991.0</td>\n",
       "      <td>2591</td>\n",
       "      <td>6.061365</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>CD4.T.cells_S05</td>\n",
       "      <td>P001</td>\n",
       "      <td>pGGO</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>LUL</td>\n",
       "      <td>0.9</td>\n",
       "      <td>ADC</td>\n",
       "      <td>MIA</td>\n",
       "      <td>T1a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P001</td>\n",
       "      <td>3464.0</td>\n",
       "      <td>1625</td>\n",
       "      <td>9.162717</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Mast.cells_S01</td>\n",
       "      <td>P001</td>\n",
       "      <td>pGGO</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>LUL</td>\n",
       "      <td>0.9</td>\n",
       "      <td>ADC</td>\n",
       "      <td>MIA</td>\n",
       "      <td>T1a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421706</th>\n",
       "      <td>P049</td>\n",
       "      <td>1497.0</td>\n",
       "      <td>723</td>\n",
       "      <td>0.407106</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>P049</td>\n",
       "      <td>SolidN</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RUL</td>\n",
       "      <td>1.6</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Acinar predominant</td>\n",
       "      <td>T1b</td>\n",
       "      <td>N2</td>\n",
       "      <td>M0</td>\n",
       "      <td>IIIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421707</th>\n",
       "      <td>P049</td>\n",
       "      <td>7093.0</td>\n",
       "      <td>541</td>\n",
       "      <td>0.027082</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>P049</td>\n",
       "      <td>SolidN</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RUL</td>\n",
       "      <td>1.6</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Acinar predominant</td>\n",
       "      <td>T1b</td>\n",
       "      <td>N2</td>\n",
       "      <td>M0</td>\n",
       "      <td>IIIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421708</th>\n",
       "      <td>P049</td>\n",
       "      <td>5584.0</td>\n",
       "      <td>2466</td>\n",
       "      <td>5.115492</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>P049</td>\n",
       "      <td>SolidN</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RUL</td>\n",
       "      <td>1.6</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Acinar predominant</td>\n",
       "      <td>T1b</td>\n",
       "      <td>N2</td>\n",
       "      <td>M0</td>\n",
       "      <td>IIIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421709</th>\n",
       "      <td>P049</td>\n",
       "      <td>5451.0</td>\n",
       "      <td>2348</td>\n",
       "      <td>9.361350</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>CD8.T.cells_S04</td>\n",
       "      <td>CD8.T.cells_S04</td>\n",
       "      <td>P049</td>\n",
       "      <td>SolidN</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RUL</td>\n",
       "      <td>1.6</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Acinar predominant</td>\n",
       "      <td>T1b</td>\n",
       "      <td>N2</td>\n",
       "      <td>M0</td>\n",
       "      <td>IIIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421710</th>\n",
       "      <td>P049</td>\n",
       "      <td>4361.0</td>\n",
       "      <td>2007</td>\n",
       "      <td>6.144832</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>P049</td>\n",
       "      <td>SolidN</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RUL</td>\n",
       "      <td>1.6</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Acinar predominant</td>\n",
       "      <td>T1b</td>\n",
       "      <td>N2</td>\n",
       "      <td>M0</td>\n",
       "      <td>IIIA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>421711 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       orig.ident  nCount_RNA  nFeature_RNA  percent.mt  assigned_cell_type  \\\n",
       "0            P001      4348.0          2099   12.531686                   5   \n",
       "1            P001      5701.0          2614    4.821479                   0   \n",
       "2            P001      5636.0          2489    2.784740                   0   \n",
       "3            P001      5991.0          2591    6.061365                   0   \n",
       "4            P001      3464.0          1625    9.162717                   5   \n",
       "...           ...         ...           ...         ...                 ...   \n",
       "421706       P049      1497.0           723    0.407106                   1   \n",
       "421707       P049      7093.0           541    0.027082                   4   \n",
       "421708       P049      5584.0          2466    5.115492                   4   \n",
       "421709       P049      5451.0          2348    9.361350                   0   \n",
       "421710       P049      4361.0          2007    6.144832                   4   \n",
       "\n",
       "        stage          Ecotype      ecoRecovery Patient Number   Group  ...  \\\n",
       "0           1       unassigned       unassigned           P001    pGGO  ...   \n",
       "1           1       unassigned  CD8.T.cells_S04           P001    pGGO  ...   \n",
       "2           1       unassigned  CD4.T.cells_S04           P001    pGGO  ...   \n",
       "3           1       unassigned  CD4.T.cells_S05           P001    pGGO  ...   \n",
       "4           1       unassigned   Mast.cells_S01           P001    pGGO  ...   \n",
       "...       ...              ...              ...            ...     ...  ...   \n",
       "421706      9       unassigned       unassigned           P049  SolidN  ...   \n",
       "421707      9       unassigned       unassigned           P049  SolidN  ...   \n",
       "421708      9       unassigned       unassigned           P049  SolidN  ...   \n",
       "421709      9  CD8.T.cells_S04  CD8.T.cells_S04           P049  SolidN  ...   \n",
       "421710      9       unassigned       unassigned           P049  SolidN  ...   \n",
       "\n",
       "       Current or former smoking  Cancer history Tumor location  \\\n",
       "0                             No              No            LUL   \n",
       "1                             No              No            LUL   \n",
       "2                             No              No            LUL   \n",
       "3                             No              No            LUL   \n",
       "4                             No              No            LUL   \n",
       "...                          ...             ...            ...   \n",
       "421706                        No              No            RUL   \n",
       "421707                        No              No            RUL   \n",
       "421708                        No              No            RUL   \n",
       "421709                        No              No            RUL   \n",
       "421710                        No              No            RUL   \n",
       "\n",
       "       Tumor size（cm） Pathology  Histological subtype T stage N stage M stage  \\\n",
       "0                 0.9       ADC                   MIA     T1a      N0      M0   \n",
       "1                 0.9       ADC                   MIA     T1a      N0      M0   \n",
       "2                 0.9       ADC                   MIA     T1a      N0      M0   \n",
       "3                 0.9       ADC                   MIA     T1a      N0      M0   \n",
       "4                 0.9       ADC                   MIA     T1a      N0      M0   \n",
       "...               ...       ...                   ...     ...     ...     ...   \n",
       "421706            1.6       ADC    Acinar predominant     T1b      N2      M0   \n",
       "421707            1.6       ADC    Acinar predominant     T1b      N2      M0   \n",
       "421708            1.6       ADC    Acinar predominant     T1b      N2      M0   \n",
       "421709            1.6       ADC    Acinar predominant     T1b      N2      M0   \n",
       "421710            1.6       ADC    Acinar predominant     T1b      N2      M0   \n",
       "\n",
       "       Pathological stage  \n",
       "0                     IA1  \n",
       "1                     IA1  \n",
       "2                     IA1  \n",
       "3                     IA1  \n",
       "4                     IA1  \n",
       "...                   ...  \n",
       "421706               IIIA  \n",
       "421707               IIIA  \n",
       "421708               IIIA  \n",
       "421709               IIIA  \n",
       "421710               IIIA  \n",
       "\n",
       "[421711 rows x 22 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "bc20928f-c6c3-4ac5-b697-d0c9caeff4df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "139394.0"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.max(adata.obs['nCount_RNA'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9ec1d3bc-a5e9-491c-95e1-31842566691c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient Number</th>\n",
       "      <th>Group</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age （years）</th>\n",
       "      <th>Current or former smoking</th>\n",
       "      <th>Cancer history</th>\n",
       "      <th>Tumor location</th>\n",
       "      <th>Tumor size（cm）</th>\n",
       "      <th>Pathology</th>\n",
       "      <th>Histological subtype</th>\n",
       "      <th>T stage</th>\n",
       "      <th>N stage</th>\n",
       "      <th>M stage</th>\n",
       "      <th>Pathological stage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P001</td>\n",
       "      <td>pGGO</td>\n",
       "      <td>Female</td>\n",
       "      <td>61.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>LUL</td>\n",
       "      <td>0.9</td>\n",
       "      <td>ADC</td>\n",
       "      <td>MIA</td>\n",
       "      <td>T1a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P002</td>\n",
       "      <td>pGGO</td>\n",
       "      <td>Female</td>\n",
       "      <td>51.0</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>LUL</td>\n",
       "      <td>3.8</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Lepidic+Acinar</td>\n",
       "      <td>T2a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P003</td>\n",
       "      <td>pGGO</td>\n",
       "      <td>Female</td>\n",
       "      <td>55.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RUL</td>\n",
       "      <td>1.1</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Unkown</td>\n",
       "      <td>T1a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P004</td>\n",
       "      <td>dGGO</td>\n",
       "      <td>Female</td>\n",
       "      <td>63.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RUL</td>\n",
       "      <td>1.1</td>\n",
       "      <td>ADC</td>\n",
       "      <td>MIA</td>\n",
       "      <td>T1a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P005</td>\n",
       "      <td>pGGO</td>\n",
       "      <td>Female</td>\n",
       "      <td>42.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RLL</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Lepidic+Acinar</td>\n",
       "      <td>T1b</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>P006</td>\n",
       "      <td>pGGO</td>\n",
       "      <td>Female</td>\n",
       "      <td>75.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RUL</td>\n",
       "      <td>1.5</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Lepidic predominant</td>\n",
       "      <td>T1c</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>P007</td>\n",
       "      <td>GGO25</td>\n",
       "      <td>Female</td>\n",
       "      <td>63.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RUL</td>\n",
       "      <td>1.4</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Lepidic predominant</td>\n",
       "      <td>T1b</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>P008</td>\n",
       "      <td>GGO25</td>\n",
       "      <td>Female</td>\n",
       "      <td>60.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RUL</td>\n",
       "      <td>1.7</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Lepidic+Acinar</td>\n",
       "      <td>T1b</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>P009</td>\n",
       "      <td>GGO25</td>\n",
       "      <td>Female</td>\n",
       "      <td>77.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RUL</td>\n",
       "      <td>2.5</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Lepidic predominant</td>\n",
       "      <td>T2a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>P010</td>\n",
       "      <td>GGO25</td>\n",
       "      <td>Male</td>\n",
       "      <td>49.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RUL</td>\n",
       "      <td>1.4</td>\n",
       "      <td>ADC</td>\n",
       "      <td>MIA</td>\n",
       "      <td>T1b</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>P011</td>\n",
       "      <td>GGO50</td>\n",
       "      <td>Male</td>\n",
       "      <td>69.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RUL</td>\n",
       "      <td>1.9</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Lepidic</td>\n",
       "      <td>T1b</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>P012</td>\n",
       "      <td>GGO50</td>\n",
       "      <td>Female</td>\n",
       "      <td>67.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>LUL</td>\n",
       "      <td>1.8</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Lepidic+Acinar</td>\n",
       "      <td>T2a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>P013</td>\n",
       "      <td>GGO50</td>\n",
       "      <td>Male</td>\n",
       "      <td>61.0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>RUL</td>\n",
       "      <td>2.1</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Lepidic+Acinar+Papillary</td>\n",
       "      <td>T1c</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>P014</td>\n",
       "      <td>GGO50</td>\n",
       "      <td>Female</td>\n",
       "      <td>61.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>LUL</td>\n",
       "      <td>3.2</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Lepidic+Acinar</td>\n",
       "      <td>T2a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>P015</td>\n",
       "      <td>GGO50</td>\n",
       "      <td>Female</td>\n",
       "      <td>64.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RLL</td>\n",
       "      <td>2.2</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Lepidic predominant</td>\n",
       "      <td>T2a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>P016</td>\n",
       "      <td>GGO50</td>\n",
       "      <td>Male</td>\n",
       "      <td>56.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RLL</td>\n",
       "      <td>2.5</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Lepidic+Acinar+Papillary</td>\n",
       "      <td>T1c</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>P017</td>\n",
       "      <td>GGO75</td>\n",
       "      <td>Male</td>\n",
       "      <td>54.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>LLL</td>\n",
       "      <td>2.1</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Lepidic+Acinar+Papillary</td>\n",
       "      <td>T1c</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>P018</td>\n",
       "      <td>GGO100</td>\n",
       "      <td>Male</td>\n",
       "      <td>47.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>LLL</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Acinar+Papillary</td>\n",
       "      <td>T2a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>P019</td>\n",
       "      <td>GGO75</td>\n",
       "      <td>Female</td>\n",
       "      <td>63.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>LUL</td>\n",
       "      <td>1.3</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Lepidic+Acinar+Papillary</td>\n",
       "      <td>T1b</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>P020</td>\n",
       "      <td>GGO75</td>\n",
       "      <td>Female</td>\n",
       "      <td>59.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>LUL</td>\n",
       "      <td>1.4</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Acinar predominant</td>\n",
       "      <td>T1b</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>P021</td>\n",
       "      <td>GGO75</td>\n",
       "      <td>Male</td>\n",
       "      <td>59.0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>RLL</td>\n",
       "      <td>1.1</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Lepidic+Acinar</td>\n",
       "      <td>T1a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>P022</td>\n",
       "      <td>GGO75</td>\n",
       "      <td>Female</td>\n",
       "      <td>67.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RLL</td>\n",
       "      <td>1.4</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Lepidic+Acinar</td>\n",
       "      <td>T1b</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>P023</td>\n",
       "      <td>GGO75</td>\n",
       "      <td>Female</td>\n",
       "      <td>64.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RUL</td>\n",
       "      <td>2.1</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Lepidic+Acinar</td>\n",
       "      <td>T1c</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>P024</td>\n",
       "      <td>GGO75</td>\n",
       "      <td>Male</td>\n",
       "      <td>73.0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>LUL</td>\n",
       "      <td>4.0</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Lepidic+Acinar+Papillary</td>\n",
       "      <td>T2a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>P025</td>\n",
       "      <td>dGGO</td>\n",
       "      <td>Male</td>\n",
       "      <td>68.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RLL</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Lepidic+Acinar</td>\n",
       "      <td>T1b</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>P026</td>\n",
       "      <td>pGGO</td>\n",
       "      <td>Male</td>\n",
       "      <td>51.0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>RUL</td>\n",
       "      <td>1.5</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Lepidic+Acinar+Papillary</td>\n",
       "      <td>T1b</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>P027</td>\n",
       "      <td>dGGO</td>\n",
       "      <td>Male</td>\n",
       "      <td>56.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>LLL</td>\n",
       "      <td>1.2</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Lepidic+Acinar</td>\n",
       "      <td>T1b</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>P028</td>\n",
       "      <td>dGGO</td>\n",
       "      <td>Female</td>\n",
       "      <td>44.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RUL</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Lepidic+Acinar+Papillary</td>\n",
       "      <td>T1b</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>P029</td>\n",
       "      <td>dGGO</td>\n",
       "      <td>Female</td>\n",
       "      <td>64.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RUL</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Lepidic predominant</td>\n",
       "      <td>T1b</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>P030</td>\n",
       "      <td>dGGO</td>\n",
       "      <td>Female</td>\n",
       "      <td>55.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RUL</td>\n",
       "      <td>1.5</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Lepidic predominant</td>\n",
       "      <td>T1a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>P031</td>\n",
       "      <td>Solid1</td>\n",
       "      <td>Male</td>\n",
       "      <td>66.0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>RLL</td>\n",
       "      <td>1.5</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Acinar+Papillary</td>\n",
       "      <td>T1b</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>P032</td>\n",
       "      <td>Solid1</td>\n",
       "      <td>Male</td>\n",
       "      <td>62.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RLL</td>\n",
       "      <td>1.4</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Acinar predominant</td>\n",
       "      <td>T1b</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>P033</td>\n",
       "      <td>Solid1</td>\n",
       "      <td>Female</td>\n",
       "      <td>56.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RUL</td>\n",
       "      <td>1.6</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Acinar predominant</td>\n",
       "      <td>T1b</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>P034</td>\n",
       "      <td>GGO100</td>\n",
       "      <td>Male</td>\n",
       "      <td>46.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>LUL</td>\n",
       "      <td>2.1</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Lepidic+Acinar</td>\n",
       "      <td>T2a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>P035</td>\n",
       "      <td>GGO100</td>\n",
       "      <td>Female</td>\n",
       "      <td>51.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>LUL</td>\n",
       "      <td>1.5</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Lepidic+Acinar</td>\n",
       "      <td>T1b</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>P036</td>\n",
       "      <td>Solid1</td>\n",
       "      <td>Female</td>\n",
       "      <td>72.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RLL</td>\n",
       "      <td>1.1</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Mucinous</td>\n",
       "      <td>T1a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>P037</td>\n",
       "      <td>GGO100</td>\n",
       "      <td>Male</td>\n",
       "      <td>79.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RUL</td>\n",
       "      <td>3.4</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Acinar+Papillary</td>\n",
       "      <td>T2a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>P038</td>\n",
       "      <td>GGO100</td>\n",
       "      <td>Female</td>\n",
       "      <td>72.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>LUL</td>\n",
       "      <td>4.4</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Unkown</td>\n",
       "      <td>T2a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>P039</td>\n",
       "      <td>Solid3</td>\n",
       "      <td>Male</td>\n",
       "      <td>57.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>LLL</td>\n",
       "      <td>3.3</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Soild</td>\n",
       "      <td>T2a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>P040</td>\n",
       "      <td>Solid3</td>\n",
       "      <td>Female</td>\n",
       "      <td>49.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>LUL</td>\n",
       "      <td>3.1</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Acinar predominant</td>\n",
       "      <td>T2a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>P041</td>\n",
       "      <td>GGO100</td>\n",
       "      <td>Male</td>\n",
       "      <td>64.0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>LLL</td>\n",
       "      <td>4.0</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Lepidic+Acinar+Papillary+ Micropapillary</td>\n",
       "      <td>T2a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>P042</td>\n",
       "      <td>Solid3</td>\n",
       "      <td>Female</td>\n",
       "      <td>57.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RUL</td>\n",
       "      <td>4.0</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Acinar+Lepidic+ Micropapillary</td>\n",
       "      <td>T2a</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>IB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>P043</td>\n",
       "      <td>SolidN</td>\n",
       "      <td>Female</td>\n",
       "      <td>64.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>LUL</td>\n",
       "      <td>3.5</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Acinar+Lepidic+ Micropapillary</td>\n",
       "      <td>T2a</td>\n",
       "      <td>N1</td>\n",
       "      <td>M0</td>\n",
       "      <td>IIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>P044</td>\n",
       "      <td>SolidN</td>\n",
       "      <td>Male</td>\n",
       "      <td>63.0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>LLL</td>\n",
       "      <td>5.5</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Solid+Acinar</td>\n",
       "      <td>T3</td>\n",
       "      <td>N2</td>\n",
       "      <td>M0</td>\n",
       "      <td>IIIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>P045</td>\n",
       "      <td>SolidN</td>\n",
       "      <td>Female</td>\n",
       "      <td>66.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RLL</td>\n",
       "      <td>6.5</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Papillary+ Micropapillary</td>\n",
       "      <td>T3</td>\n",
       "      <td>N2</td>\n",
       "      <td>M0</td>\n",
       "      <td>IIIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>P046</td>\n",
       "      <td>SolidN</td>\n",
       "      <td>Female</td>\n",
       "      <td>73.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RUL</td>\n",
       "      <td>2.8</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Acinar+Lepidic+ Micropapillary</td>\n",
       "      <td>T2a</td>\n",
       "      <td>N2</td>\n",
       "      <td>M0</td>\n",
       "      <td>IIIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>P047</td>\n",
       "      <td>SolidN</td>\n",
       "      <td>Male</td>\n",
       "      <td>55.0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>RLL</td>\n",
       "      <td>3.1</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Solid+Acinar+Papillary</td>\n",
       "      <td>T2a</td>\n",
       "      <td>N2</td>\n",
       "      <td>M0</td>\n",
       "      <td>IIIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>P048</td>\n",
       "      <td>SolidN</td>\n",
       "      <td>Male</td>\n",
       "      <td>62.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>LUL</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Unkown</td>\n",
       "      <td>T1a</td>\n",
       "      <td>N2</td>\n",
       "      <td>M0</td>\n",
       "      <td>IIIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>P049</td>\n",
       "      <td>SolidN</td>\n",
       "      <td>Female</td>\n",
       "      <td>75.0</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>RUL</td>\n",
       "      <td>1.6</td>\n",
       "      <td>ADC</td>\n",
       "      <td>Acinar predominant</td>\n",
       "      <td>T1b</td>\n",
       "      <td>N2</td>\n",
       "      <td>M0</td>\n",
       "      <td>IIIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>Abbreviations: GGO,ground-glass opacity; ADC, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       Patient Number   Group     Sex  \\\n",
       "0                                                P001    pGGO  Female   \n",
       "1                                                P002    pGGO  Female   \n",
       "2                                                P003    pGGO  Female   \n",
       "3                                                P004    dGGO  Female   \n",
       "4                                                P005    pGGO  Female   \n",
       "5                                                P006    pGGO  Female   \n",
       "6                                                P007   GGO25  Female   \n",
       "7                                                P008   GGO25  Female   \n",
       "8                                                P009   GGO25  Female   \n",
       "9                                                P010   GGO25    Male   \n",
       "10                                               P011   GGO50    Male   \n",
       "11                                               P012   GGO50  Female   \n",
       "12                                               P013   GGO50    Male   \n",
       "13                                               P014   GGO50  Female   \n",
       "14                                               P015   GGO50  Female   \n",
       "15                                               P016   GGO50    Male   \n",
       "16                                               P017   GGO75    Male   \n",
       "17                                               P018  GGO100    Male   \n",
       "18                                               P019   GGO75  Female   \n",
       "19                                               P020   GGO75  Female   \n",
       "20                                               P021   GGO75    Male   \n",
       "21                                               P022   GGO75  Female   \n",
       "22                                               P023   GGO75  Female   \n",
       "23                                               P024   GGO75    Male   \n",
       "24                                               P025    dGGO    Male   \n",
       "25                                               P026    pGGO    Male   \n",
       "26                                               P027    dGGO    Male   \n",
       "27                                               P028    dGGO  Female   \n",
       "28                                               P029    dGGO  Female   \n",
       "29                                               P030    dGGO  Female   \n",
       "30                                               P031  Solid1    Male   \n",
       "31                                               P032  Solid1    Male   \n",
       "32                                               P033  Solid1  Female   \n",
       "33                                               P034  GGO100    Male   \n",
       "34                                               P035  GGO100  Female   \n",
       "35                                               P036  Solid1  Female   \n",
       "36                                               P037  GGO100    Male   \n",
       "37                                               P038  GGO100  Female   \n",
       "38                                               P039  Solid3    Male   \n",
       "39                                               P040  Solid3  Female   \n",
       "40                                               P041  GGO100    Male   \n",
       "41                                               P042  Solid3  Female   \n",
       "42                                               P043  SolidN  Female   \n",
       "43                                               P044  SolidN    Male   \n",
       "44                                               P045  SolidN  Female   \n",
       "45                                               P046  SolidN  Female   \n",
       "46                                               P047  SolidN    Male   \n",
       "47                                               P048  SolidN    Male   \n",
       "48                                               P049  SolidN  Female   \n",
       "49  Abbreviations: GGO,ground-glass opacity; ADC, ...     NaN     NaN   \n",
       "\n",
       "    Age （years） Current or former smoking Cancer history Tumor location  \\\n",
       "0          61.0                        No             No            LUL   \n",
       "1          51.0                        No            Yes            LUL   \n",
       "2          55.0                        No             No            RUL   \n",
       "3          63.0                        No             No            RUL   \n",
       "4          42.0                        No             No            RLL   \n",
       "5          75.0                        No             No            RUL   \n",
       "6          63.0                        No             No            RUL   \n",
       "7          60.0                        No             No            RUL   \n",
       "8          77.0                        No             No            RUL   \n",
       "9          49.0                        No             No            RUL   \n",
       "10         69.0                        No             No            RUL   \n",
       "11         67.0                        No             No            LUL   \n",
       "12         61.0                       Yes             No            RUL   \n",
       "13         61.0                        No             No            LUL   \n",
       "14         64.0                        No             No            RLL   \n",
       "15         56.0                        No             No            RLL   \n",
       "16         54.0                        No             No            LLL   \n",
       "17         47.0                        No             No            LLL   \n",
       "18         63.0                        No             No            LUL   \n",
       "19         59.0                        No             No            LUL   \n",
       "20         59.0                       Yes             No            RLL   \n",
       "21         67.0                        No             No            RLL   \n",
       "22         64.0                        No             No            RUL   \n",
       "23         73.0                       Yes             No            LUL   \n",
       "24         68.0                        No             No            RLL   \n",
       "25         51.0                       Yes             No            RUL   \n",
       "26         56.0                        No             No            LLL   \n",
       "27         44.0                        No             No            RUL   \n",
       "28         64.0                        No             No            RUL   \n",
       "29         55.0                        No             No            RUL   \n",
       "30         66.0                       Yes             No            RLL   \n",
       "31         62.0                        No             No            RLL   \n",
       "32         56.0                        No             No            RUL   \n",
       "33         46.0                        No             No            LUL   \n",
       "34         51.0                        No             No            LUL   \n",
       "35         72.0                        No             No            RLL   \n",
       "36         79.0                        No             No            RUL   \n",
       "37         72.0                        No             No            LUL   \n",
       "38         57.0                        No             No            LLL   \n",
       "39         49.0                        No             No            LUL   \n",
       "40         64.0                       Yes             No            LLL   \n",
       "41         57.0                        No             No            RUL   \n",
       "42         64.0                        No             No            LUL   \n",
       "43         63.0                       Yes             No            LLL   \n",
       "44         66.0                        No             No            RLL   \n",
       "45         73.0                        No             No            RUL   \n",
       "46         55.0                       Yes             No            RLL   \n",
       "47         62.0                        No             No            LUL   \n",
       "48         75.0                        No             No            RUL   \n",
       "49          NaN                       NaN            NaN            NaN   \n",
       "\n",
       "    Tumor size（cm） Pathology                      Histological subtype  \\\n",
       "0              0.9       ADC                                       MIA   \n",
       "1              3.8       ADC                            Lepidic+Acinar   \n",
       "2              1.1       ADC                                    Unkown   \n",
       "3              1.1       ADC                                       MIA   \n",
       "4              1.0       ADC                            Lepidic+Acinar   \n",
       "5              1.5       ADC                       Lepidic predominant   \n",
       "6              1.4       ADC                       Lepidic predominant   \n",
       "7              1.7       ADC                            Lepidic+Acinar   \n",
       "8              2.5       ADC                       Lepidic predominant   \n",
       "9              1.4       ADC                                       MIA   \n",
       "10             1.9       ADC                                   Lepidic   \n",
       "11             1.8       ADC                            Lepidic+Acinar   \n",
       "12             2.1       ADC                  Lepidic+Acinar+Papillary   \n",
       "13             3.2       ADC                            Lepidic+Acinar   \n",
       "14             2.2       ADC                       Lepidic predominant   \n",
       "15             2.5       ADC                  Lepidic+Acinar+Papillary   \n",
       "16             2.1       ADC                  Lepidic+Acinar+Papillary   \n",
       "17             1.0       ADC                          Acinar+Papillary   \n",
       "18             1.3       ADC                  Lepidic+Acinar+Papillary   \n",
       "19             1.4       ADC                        Acinar predominant   \n",
       "20             1.1       ADC                            Lepidic+Acinar   \n",
       "21             1.4       ADC                            Lepidic+Acinar   \n",
       "22             2.1       ADC                            Lepidic+Acinar   \n",
       "23             4.0       ADC                  Lepidic+Acinar+Papillary   \n",
       "24             2.0       ADC                            Lepidic+Acinar   \n",
       "25             1.5       ADC                  Lepidic+Acinar+Papillary   \n",
       "26             1.2       ADC                            Lepidic+Acinar   \n",
       "27             2.0       ADC                  Lepidic+Acinar+Papillary   \n",
       "28             2.0       ADC                       Lepidic predominant   \n",
       "29             1.5       ADC                       Lepidic predominant   \n",
       "30             1.5       ADC                          Acinar+Papillary   \n",
       "31             1.4       ADC                        Acinar predominant   \n",
       "32             1.6       ADC                        Acinar predominant   \n",
       "33             2.1       ADC                            Lepidic+Acinar   \n",
       "34             1.5       ADC                            Lepidic+Acinar   \n",
       "35             1.1       ADC                                  Mucinous   \n",
       "36             3.4       ADC                          Acinar+Papillary   \n",
       "37             4.4       ADC                                    Unkown   \n",
       "38             3.3       ADC                                     Soild   \n",
       "39             3.1       ADC                        Acinar predominant   \n",
       "40             4.0       ADC  Lepidic+Acinar+Papillary+ Micropapillary   \n",
       "41             4.0       ADC            Acinar+Lepidic+ Micropapillary   \n",
       "42             3.5       ADC            Acinar+Lepidic+ Micropapillary   \n",
       "43             5.5       ADC                              Solid+Acinar   \n",
       "44             6.5       ADC                 Papillary+ Micropapillary   \n",
       "45             2.8       ADC            Acinar+Lepidic+ Micropapillary   \n",
       "46             3.1       ADC                    Solid+Acinar+Papillary   \n",
       "47             2.0       ADC                                    Unkown   \n",
       "48             1.6       ADC                        Acinar predominant   \n",
       "49             NaN       NaN                                       NaN   \n",
       "\n",
       "   T stage N stage M stage Pathological stage  \n",
       "0      T1a      N0      M0                IA1  \n",
       "1      T2a      N0      M0                 IB  \n",
       "2      T1a      N0      M0                IA2  \n",
       "3      T1a      N0      M0                IA1  \n",
       "4      T1b      N0      M0                IA2  \n",
       "5      T1c      N0      M0                IA3  \n",
       "6      T1b      N0      M0                IA2  \n",
       "7      T1b      N0      M0                IA2  \n",
       "8      T2a      N0      M0                 IB  \n",
       "9      T1b      N0      M0                IA2  \n",
       "10     T1b      N0      M0                IA2  \n",
       "11     T2a      N0      M0                 IB  \n",
       "12     T1c      N0      M0                IA3  \n",
       "13     T2a      N0      M0                 IB  \n",
       "14     T2a      N0      M0                 IB  \n",
       "15     T1c      N0      M0                IA3  \n",
       "16     T1c      N0      M0                IA3  \n",
       "17     T2a      N0      M0                 IB  \n",
       "18     T1b      N0      M0                IA2  \n",
       "19     T1b      N0      M0                IA2  \n",
       "20     T1a      N0      M0                IA1  \n",
       "21     T1b      N0      M0                IA2  \n",
       "22     T1c      N0      M0                IA3  \n",
       "23     T2a      N0      M0                 IB  \n",
       "24     T1b      N0      M0                IA2  \n",
       "25     T1b      N0      M0                IA2  \n",
       "26     T1b      N0      M0                IA2  \n",
       "27     T1b      N0      M0                IA2  \n",
       "28     T1b      N0      M0                IA2  \n",
       "29     T1a      N0      M0                IA1  \n",
       "30     T1b      N0      M0                IA2  \n",
       "31     T1b      N0      M0                IA2  \n",
       "32     T1b      N0      M0                IA2  \n",
       "33     T2a      N0      M0                 IB  \n",
       "34     T1b      N0      M0                IA2  \n",
       "35     T1a      N0      M0                IA1  \n",
       "36     T2a      N0      M0                 IB  \n",
       "37     T2a      N0      M0                 IB  \n",
       "38     T2a      N0      M0                 IB  \n",
       "39     T2a      N0      M0                 IB  \n",
       "40     T2a      N0      M0                 IB  \n",
       "41     T2a      N0      M0                 IB  \n",
       "42     T2a      N1      M0                IIB  \n",
       "43      T3      N2      M0               IIIB  \n",
       "44      T3      N2      M0               IIIB  \n",
       "45     T2a      N2      M0               IIIA  \n",
       "46     T2a      N2      M0               IIIA  \n",
       "47     T1a      N2      M0               IIIA  \n",
       "48     T1b      N2      M0               IIIA  \n",
       "49     NaN     NaN     NaN                NaN  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "deng_meta = pd.read_excel('/root/datos/maestria/netopaas/Deng2024/metadata.xlsx')\n",
    "deng_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "1e1dcce6-f282-4926-8d4c-c8747e7bff66",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['unassigned', 'CD8.T.cells_S04', 'CD4.T.cells_S04',\n",
       "       'CD4.T.cells_S05', 'Mast.cells_S01', 'CD8.T.cells_S03',\n",
       "       'NK.cells_S04', 'CD4.T.cells_S03', 'CD8.T.cells_S01',\n",
       "       'B.cells_S02', 'CD4.T.cells_S01', 'Monocytes.and.Macrophages_S05',\n",
       "       'NK.cells_S02', 'NK.cells_S01', 'CD4.T.cells_S02',\n",
       "       'CD8.T.cells_S02', 'NK.cells_S03', 'Dendritic.cells_S01',\n",
       "       'PMNs_S01', 'Monocytes.and.Macrophages_S04', 'B.cells_S01',\n",
       "       'Monocytes.and.Macrophages_S02', 'B.cells_S03',\n",
       "       'Monocytes.and.Macrophages_S03', 'PMNs_S03', 'Dendritic.cells_S02',\n",
       "       'Mast.cells_S02', 'Monocytes.and.Macrophages_S01',\n",
       "       'Epithelial.cells_S02', 'PMNs_S02', 'Fibroblasts_S03',\n",
       "       'Endothelial.cells_S02', 'Epithelial.cells_S03', 'Fibroblasts_S01',\n",
       "       'Endothelial.cells_S01', 'Endothelial.cells_S04',\n",
       "       'Fibroblasts_S04', 'Epithelial.cells_S01', 'Endothelial.cells_S05',\n",
       "       'Mast.cells_S03', 'Mast.cells_S04', 'Fibroblasts_S02',\n",
       "       'Endothelial.cells_S03'], dtype=object)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.ecoRecovery.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b60d5673-27f4-4ab8-b551-6b84c3ba0aa9",
   "metadata": {},
   "source": [
    "## Alexandra 2023"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "1e7ee161-c989-4081-a668-f06834bef03d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/root/datos/maestria/netopaas//alexandra2023/data.zip',\n",
       " <http.client.HTTPMessage at 0x7f7325a5cf10>)"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "urllib.request.urlretrieve('https://zenodo.org/records/7852154/files/data_code.zip?download=1',f'{data_dir}/Alexandra2023/data.zip')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5628b30-65fc-40ab-abc2-7a9d1a8190b8",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Zhong 2024\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "cc7923b6-6f52-44f5-9135-377b38a3aa69",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>percent.mt</th>\n",
       "      <th>sampleID</th>\n",
       "      <th>cellID</th>\n",
       "      <th>RNA_snn_res.0.8</th>\n",
       "      <th>seurat_clusters</th>\n",
       "      <th>major.cell.type</th>\n",
       "      <th>RNA_snn_res.1</th>\n",
       "      <th>...</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Age</th>\n",
       "      <th>Histology</th>\n",
       "      <th>Cycles</th>\n",
       "      <th>PD1</th>\n",
       "      <th>Pathological Response</th>\n",
       "      <th>EGFR</th>\n",
       "      <th>Pathological Response Rate</th>\n",
       "      <th>Smoking_History</th>\n",
       "      <th>PD-L1 TPS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P343</td>\n",
       "      <td>17606</td>\n",
       "      <td>4574</td>\n",
       "      <td>2.891060</td>\n",
       "      <td>P343</td>\n",
       "      <td>P343_AAACCTGAGCTATGCT-1</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>Fibro</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>F</td>\n",
       "      <td>60</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>L858R</td>\n",
       "      <td>0.85</td>\n",
       "      <td>N</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P343</td>\n",
       "      <td>9542</td>\n",
       "      <td>3230</td>\n",
       "      <td>4.003354</td>\n",
       "      <td>P343</td>\n",
       "      <td>P343_AAACCTGCATGATCCA-1</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>Fibro</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>F</td>\n",
       "      <td>60</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>L858R</td>\n",
       "      <td>0.85</td>\n",
       "      <td>N</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P343</td>\n",
       "      <td>10729</td>\n",
       "      <td>3832</td>\n",
       "      <td>2.302172</td>\n",
       "      <td>P343</td>\n",
       "      <td>P343_AAACCTGGTGAGTATA-1</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>Endo</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>F</td>\n",
       "      <td>60</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>L858R</td>\n",
       "      <td>0.85</td>\n",
       "      <td>N</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P343</td>\n",
       "      <td>7788</td>\n",
       "      <td>2859</td>\n",
       "      <td>5.225989</td>\n",
       "      <td>P343</td>\n",
       "      <td>P343_AAACCTGTCCAGTATG-1</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>Endo</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>F</td>\n",
       "      <td>60</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>L858R</td>\n",
       "      <td>0.85</td>\n",
       "      <td>N</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P343</td>\n",
       "      <td>15449</td>\n",
       "      <td>4074</td>\n",
       "      <td>3.301185</td>\n",
       "      <td>P343</td>\n",
       "      <td>P343_AAACGGGCACGACGAA-1</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>Endo</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>F</td>\n",
       "      <td>60</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>L858R</td>\n",
       "      <td>0.85</td>\n",
       "      <td>N</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78686</th>\n",
       "      <td>P591</td>\n",
       "      <td>6608</td>\n",
       "      <td>2528</td>\n",
       "      <td>3.344431</td>\n",
       "      <td>P591</td>\n",
       "      <td>P591_TGCCCTAGTCAGATAA-1</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>Epi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>51</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>KDD</td>\n",
       "      <td>0.58</td>\n",
       "      <td>Y</td>\n",
       "      <td>&lt;1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78687</th>\n",
       "      <td>P591</td>\n",
       "      <td>1064</td>\n",
       "      <td>707</td>\n",
       "      <td>7.518797</td>\n",
       "      <td>P591</td>\n",
       "      <td>P591_TGGCTGGAGCGTTGCC-1</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>Epi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>51</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>KDD</td>\n",
       "      <td>0.58</td>\n",
       "      <td>Y</td>\n",
       "      <td>&lt;1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78688</th>\n",
       "      <td>P591</td>\n",
       "      <td>13226</td>\n",
       "      <td>3318</td>\n",
       "      <td>1.126569</td>\n",
       "      <td>P591</td>\n",
       "      <td>P591_TTCTCCTAGTGCGTGA-1</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>Epi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>51</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>KDD</td>\n",
       "      <td>0.58</td>\n",
       "      <td>Y</td>\n",
       "      <td>&lt;1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78689</th>\n",
       "      <td>P591</td>\n",
       "      <td>7335</td>\n",
       "      <td>2912</td>\n",
       "      <td>6.693933</td>\n",
       "      <td>P591</td>\n",
       "      <td>P591_TTTCCTCAGCGTAATA-1</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>Epi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>51</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>KDD</td>\n",
       "      <td>0.58</td>\n",
       "      <td>Y</td>\n",
       "      <td>&lt;1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78690</th>\n",
       "      <td>P591</td>\n",
       "      <td>14498</td>\n",
       "      <td>4608</td>\n",
       "      <td>1.448476</td>\n",
       "      <td>P591</td>\n",
       "      <td>P591_TTTGGTTCATGGGACA-1</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>Epi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>51</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>KDD</td>\n",
       "      <td>0.58</td>\n",
       "      <td>Y</td>\n",
       "      <td>&lt;1%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78691 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      orig.ident  nCount_RNA  nFeature_RNA  percent.mt sampleID  \\\n",
       "0           P343       17606          4574    2.891060     P343   \n",
       "1           P343        9542          3230    4.003354     P343   \n",
       "2           P343       10729          3832    2.302172     P343   \n",
       "3           P343        7788          2859    5.225989     P343   \n",
       "4           P343       15449          4074    3.301185     P343   \n",
       "...          ...         ...           ...         ...      ...   \n",
       "78686       P591        6608          2528    3.344431     P591   \n",
       "78687       P591        1064           707    7.518797     P591   \n",
       "78688       P591       13226          3318    1.126569     P591   \n",
       "78689       P591        7335          2912    6.693933     P591   \n",
       "78690       P591       14498          4608    1.448476     P591   \n",
       "\n",
       "                        cellID  RNA_snn_res.0.8  seurat_clusters  \\\n",
       "0      P343_AAACCTGAGCTATGCT-1                6                6   \n",
       "1      P343_AAACCTGCATGATCCA-1                8                8   \n",
       "2      P343_AAACCTGGTGAGTATA-1               10               10   \n",
       "3      P343_AAACCTGTCCAGTATG-1               10               10   \n",
       "4      P343_AAACGGGCACGACGAA-1               10               10   \n",
       "...                        ...              ...              ...   \n",
       "78686  P591_TGCCCTAGTCAGATAA-1                7                7   \n",
       "78687  P591_TGGCTGGAGCGTTGCC-1                7                7   \n",
       "78688  P591_TTCTCCTAGTGCGTGA-1                7                7   \n",
       "78689  P591_TTTCCTCAGCGTAATA-1                7                7   \n",
       "78690  P591_TTTGGTTCATGGGACA-1               20               20   \n",
       "\n",
       "      major.cell.type  RNA_snn_res.1  ... Gender  Age Histology Cycles  \\\n",
       "0               Fibro            NaN  ...      F   60      LUAD      3   \n",
       "1               Fibro            NaN  ...      F   60      LUAD      3   \n",
       "2                Endo            NaN  ...      F   60      LUAD      3   \n",
       "3                Endo            NaN  ...      F   60      LUAD      3   \n",
       "4                Endo            NaN  ...      F   60      LUAD      3   \n",
       "...               ...            ...  ...    ...  ...       ...    ...   \n",
       "78686             Epi            NaN  ...      M   51      LUAD      3   \n",
       "78687             Epi            NaN  ...      M   51      LUAD      3   \n",
       "78688             Epi            NaN  ...      M   51      LUAD      3   \n",
       "78689             Epi            NaN  ...      M   51      LUAD      3   \n",
       "78690             Epi            NaN  ...      M   51      LUAD      3   \n",
       "\n",
       "              PD1 Pathological Response   EGFR Pathological Response Rate  \\\n",
       "0      Sintilimab               non-MPR  L858R                       0.85   \n",
       "1      Sintilimab               non-MPR  L858R                       0.85   \n",
       "2      Sintilimab               non-MPR  L858R                       0.85   \n",
       "3      Sintilimab               non-MPR  L858R                       0.85   \n",
       "4      Sintilimab               non-MPR  L858R                       0.85   \n",
       "...           ...                   ...    ...                        ...   \n",
       "78686  Sintilimab               non-MPR    KDD                       0.58   \n",
       "78687  Sintilimab               non-MPR    KDD                       0.58   \n",
       "78688  Sintilimab               non-MPR    KDD                       0.58   \n",
       "78689  Sintilimab               non-MPR    KDD                       0.58   \n",
       "78690  Sintilimab               non-MPR    KDD                       0.58   \n",
       "\n",
       "      Smoking_History PD-L1 TPS  \n",
       "0                   N      0.01  \n",
       "1                   N      0.01  \n",
       "2                   N      0.01  \n",
       "3                   N      0.01  \n",
       "4                   N      0.01  \n",
       "...               ...       ...  \n",
       "78686               Y       <1%  \n",
       "78687               Y       <1%  \n",
       "78688               Y       <1%  \n",
       "78689               Y       <1%  \n",
       "78690               Y       <1%  \n",
       "\n",
       "[78691 rows x 23 columns]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zhong_samples = pd.read_csv(f'{data_dir}/Zhong2024/GSE241934_IIT_Meta.txt', sep='\\t')\n",
    "zhong_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63681388-f7cf-4b89-bd43-2715e4050437",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Thomas 2023"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "febe513c-7440-4a9c-8d8e-67ceecbdd5c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "! cd /root/datos/maestria/netopaas/Thomas2023 & wget https://ftp.ncbi.nlm.nih.gov/geo/series/GSE153nnn/GSE153935/suppl/GSE153935%5FTLDS%5FAllCells%2Etxt%2Egz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3c7a2df8-b60e-48bc-a06a-60421c015fe2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>D2T_GTTCAGCTTCTT</th>\n",
       "      <th>D2T_GCGCCCCATCAC</th>\n",
       "      <th>D2T_TGTCGGGGATCA</th>\n",
       "      <th>D2T_CCAATAATCTTA</th>\n",
       "      <th>D2T_CCTTTCCGTTTG</th>\n",
       "      <th>D2T_TCGCTGAATGTA</th>\n",
       "      <th>D2T_CAATACAACGAT</th>\n",
       "      <th>D2T_CTCTCCCTGTGG</th>\n",
       "      <th>D2T_GCTCGTGCTCTA</th>\n",
       "      <th>D2T_GATCCTTCTAGA</th>\n",
       "      <th>...</th>\n",
       "      <th>D4T_GCGCGGGATTTT</th>\n",
       "      <th>D12T_CACGCGTTTGTG</th>\n",
       "      <th>D12T_CGCCAACCACTG</th>\n",
       "      <th>D12T_GCTGACATATTT</th>\n",
       "      <th>D12T_TCCCTAAAGTGT</th>\n",
       "      <th>D12T_ACATGGGTAAGC</th>\n",
       "      <th>D12T_TGGTGATGGCCA</th>\n",
       "      <th>D12T_CACTATTCCTGA</th>\n",
       "      <th>D12T_TGGAGTATTGCA</th>\n",
       "      <th>D12T_TGTTCGAAGCTT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M-AS1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-mir-5195</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-mir-6080</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-mir-7515</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-mir-8072</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snoU2-30</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25389 rows × 8158 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              D2T_GTTCAGCTTCTT  D2T_GCGCCCCATCAC  D2T_TGTCGGGGATCA  \\\n",
       "A1BG                         0                 0                 0   \n",
       "A1BG-AS1                     0                 0                 0   \n",
       "A1CF                         0                 0                 0   \n",
       "A2M                          0                 7                 9   \n",
       "A2M-AS1                      0                 0                 0   \n",
       "...                        ...               ...               ...   \n",
       "hsa-mir-5195                 0                 0                 0   \n",
       "hsa-mir-6080                 0                 0                 0   \n",
       "hsa-mir-7515                 0                 0                 0   \n",
       "hsa-mir-8072                 0                 0                 0   \n",
       "snoU2-30                     0                 0                 0   \n",
       "\n",
       "              D2T_CCAATAATCTTA  D2T_CCTTTCCGTTTG  D2T_TCGCTGAATGTA  \\\n",
       "A1BG                         0                 0                 0   \n",
       "A1BG-AS1                     0                 0                 0   \n",
       "A1CF                         0                 0                 0   \n",
       "A2M                          0                 2                 3   \n",
       "A2M-AS1                      0                 0                 0   \n",
       "...                        ...               ...               ...   \n",
       "hsa-mir-5195                 0                 0                 0   \n",
       "hsa-mir-6080                 0                 0                 0   \n",
       "hsa-mir-7515                 0                 0                 0   \n",
       "hsa-mir-8072                 0                 0                 0   \n",
       "snoU2-30                     0                 0                 0   \n",
       "\n",
       "              D2T_CAATACAACGAT  D2T_CTCTCCCTGTGG  D2T_GCTCGTGCTCTA  \\\n",
       "A1BG                         0                 0                 0   \n",
       "A1BG-AS1                     0                 0                 0   \n",
       "A1CF                         0                 0                 0   \n",
       "A2M                          0                 0                 1   \n",
       "A2M-AS1                      0                 0                 0   \n",
       "...                        ...               ...               ...   \n",
       "hsa-mir-5195                 0                 0                 0   \n",
       "hsa-mir-6080                 0                 0                 0   \n",
       "hsa-mir-7515                 0                 0                 0   \n",
       "hsa-mir-8072                 0                 0                 0   \n",
       "snoU2-30                     0                 0                 0   \n",
       "\n",
       "              D2T_GATCCTTCTAGA  ...  D4T_GCGCGGGATTTT  D12T_CACGCGTTTGTG  \\\n",
       "A1BG                         0  ...                 0                  0   \n",
       "A1BG-AS1                     0  ...                 0                  0   \n",
       "A1CF                         0  ...                 0                  0   \n",
       "A2M                          0  ...                 0                  0   \n",
       "A2M-AS1                      0  ...                 0                  0   \n",
       "...                        ...  ...               ...                ...   \n",
       "hsa-mir-5195                 0  ...                 0                  0   \n",
       "hsa-mir-6080                 0  ...                 0                  0   \n",
       "hsa-mir-7515                 0  ...                 0                  0   \n",
       "hsa-mir-8072                 1  ...                 0                  0   \n",
       "snoU2-30                     0  ...                 0                  0   \n",
       "\n",
       "              D12T_CGCCAACCACTG  D12T_GCTGACATATTT  D12T_TCCCTAAAGTGT  \\\n",
       "A1BG                          0                  0                  0   \n",
       "A1BG-AS1                      0                  0                  0   \n",
       "A1CF                          0                  0                  0   \n",
       "A2M                           0                  0                  0   \n",
       "A2M-AS1                       0                  0                  0   \n",
       "...                         ...                ...                ...   \n",
       "hsa-mir-5195                  0                  0                  0   \n",
       "hsa-mir-6080                  0                  0                  0   \n",
       "hsa-mir-7515                  0                  0                  0   \n",
       "hsa-mir-8072                  0                  0                  0   \n",
       "snoU2-30                      0                  0                  0   \n",
       "\n",
       "              D12T_ACATGGGTAAGC  D12T_TGGTGATGGCCA  D12T_CACTATTCCTGA  \\\n",
       "A1BG                          0                  0                  0   \n",
       "A1BG-AS1                      0                  0                  0   \n",
       "A1CF                          0                  0                  0   \n",
       "A2M                           0                  0                  0   \n",
       "A2M-AS1                       0                  0                  0   \n",
       "...                         ...                ...                ...   \n",
       "hsa-mir-5195                  0                  0                  0   \n",
       "hsa-mir-6080                  0                  0                  0   \n",
       "hsa-mir-7515                  0                  0                  0   \n",
       "hsa-mir-8072                  0                  0                  0   \n",
       "snoU2-30                      0                  0                  0   \n",
       "\n",
       "              D12T_TGGAGTATTGCA  D12T_TGTTCGAAGCTT  \n",
       "A1BG                          0                  0  \n",
       "A1BG-AS1                      0                  0  \n",
       "A1CF                          0                  0  \n",
       "A2M                           0                  0  \n",
       "A2M-AS1                       0                  0  \n",
       "...                         ...                ...  \n",
       "hsa-mir-5195                  0                  0  \n",
       "hsa-mir-6080                  0                  0  \n",
       "hsa-mir-7515                  0                  0  \n",
       "hsa-mir-8072                  0                  0  \n",
       "snoU2-30                      0                  0  \n",
       "\n",
       "[25389 rows x 8158 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "thomas_mtx = pd.read_csv(f'{data_dir}/Thomas2023/GSE153935_TLDS_AllCells.txt.gz', sep='\\t')\n",
    "thomas_mtx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b48fa5cf-e458-4347-a087-64f25fee8e42",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
