{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a80560b2-00f4-4d4f-9d13-2b5c48e057d0",
   "metadata": {},
   "source": [
    "# Download the scRNA data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "32151548-2c85-4ced-887e-2523ab68a156",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = '/root/datos/maestria/netopaas/'\n",
    "backup_dir = '/root/datos/maestria/netopaas/luca_explore/surgeries'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "id": "8a4a47b6-e24d-4170-9d4a-e3566be33eae",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_383908/1655992530.py:20: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  anndata2ri.activate()\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import utils.functions as ut\n",
    "import urllib\n",
    "\n",
    "import rpy2.rinterface_lib.callbacks\n",
    "import logging\n",
    "from rpy2.robjects import pandas2ri\n",
    "import rpy2.robjects as robjects\n",
    "\n",
    "# Set the graphics device\n",
    "robjects.r('options(bitmapType=\"cairo\")')\n",
    "import anndata2ri\n",
    "\n",
    "# Ignore R warning messages\n",
    "#Note: this can be commented out to get more verbose R output\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)\n",
    "\n",
    "# Automatically convert rpy2 outputs to pandas dataframes\n",
    "pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68ea9660-e216-414c-aca4-11143ca39a33",
   "metadata": {},
   "source": [
    "## Zuani 2024"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a0630a4-3ef6-40a0-aa9d-f6f4abd79cbf",
   "metadata": {},
   "source": [
    "### Download"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "77a1cbc9-e3cb-4cdf-9e76-28d3802cc92f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Source Name</th>\n",
       "      <th>Comment[ENA_SAMPLE]</th>\n",
       "      <th>Comment[BioSD_SAMPLE]</th>\n",
       "      <th>Characteristics[organism]</th>\n",
       "      <th>Characteristics[individual]</th>\n",
       "      <th>Characteristics[original source name]</th>\n",
       "      <th>Characteristics[age]</th>\n",
       "      <th>Unit[time unit]</th>\n",
       "      <th>Term Source REF</th>\n",
       "      <th>Term Accession Number</th>\n",
       "      <th>...</th>\n",
       "      <th>Derived Array Data File</th>\n",
       "      <th>Protocol REF.6</th>\n",
       "      <th>Protocol REF.7</th>\n",
       "      <th>Derived Array Data File.1</th>\n",
       "      <th>Protocol REF.8</th>\n",
       "      <th>Protocol REF.9</th>\n",
       "      <th>Derived Array Data File.2</th>\n",
       "      <th>Factor Value[disease]</th>\n",
       "      <th>Factor Value[FACS]</th>\n",
       "      <th>Factor Value[sampling site]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P1_B1</td>\n",
       "      <td>ERS16703054</td>\n",
       "      <td>SAMEA114591031</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 1</td>\n",
       "      <td>TB18.0655</td>\n",
       "      <td>73</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P1_B1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-features.tsv.gz</td>\n",
       "      <td>non-small cell lung cancer</td>\n",
       "      <td>MDSC</td>\n",
       "      <td>normal tissue adjacent to tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P1_B1</td>\n",
       "      <td>ERS16703054</td>\n",
       "      <td>SAMEA114591031</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 1</td>\n",
       "      <td>TB18.0655</td>\n",
       "      <td>73</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P1_B1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-features.tsv.gz</td>\n",
       "      <td>non-small cell lung cancer</td>\n",
       "      <td>MDSC</td>\n",
       "      <td>normal tissue adjacent to tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P1_B1</td>\n",
       "      <td>ERS16703054</td>\n",
       "      <td>SAMEA114591031</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 1</td>\n",
       "      <td>TB18.0655</td>\n",
       "      <td>73</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P1_B1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-features.tsv.gz</td>\n",
       "      <td>non-small cell lung cancer</td>\n",
       "      <td>MDSC</td>\n",
       "      <td>normal tissue adjacent to tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P1_B1</td>\n",
       "      <td>ERS16703054</td>\n",
       "      <td>SAMEA114591031</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 1</td>\n",
       "      <td>TB18.0655</td>\n",
       "      <td>73</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P1_B1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-features.tsv.gz</td>\n",
       "      <td>non-small cell lung cancer</td>\n",
       "      <td>MDSC</td>\n",
       "      <td>normal tissue adjacent to tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P1_B1</td>\n",
       "      <td>ERS16703054</td>\n",
       "      <td>SAMEA114591031</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 1</td>\n",
       "      <td>TB18.0655</td>\n",
       "      <td>73</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P1_B1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-features.tsv.gz</td>\n",
       "      <td>non-small cell lung cancer</td>\n",
       "      <td>MDSC</td>\n",
       "      <td>normal tissue adjacent to tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1562</th>\n",
       "      <td>P18_T1</td>\n",
       "      <td>ERS16703118</td>\n",
       "      <td>SAMEA114591092</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 18</td>\n",
       "      <td>TB21.0006</td>\n",
       "      <td>77</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P18_T1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-features.tsv.gz</td>\n",
       "      <td>lung squamous cell carcinoma</td>\n",
       "      <td>CD45+</td>\n",
       "      <td>tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1563</th>\n",
       "      <td>P18_T1</td>\n",
       "      <td>ERS16703118</td>\n",
       "      <td>SAMEA114591092</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 18</td>\n",
       "      <td>TB21.0006</td>\n",
       "      <td>77</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P18_T1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-features.tsv.gz</td>\n",
       "      <td>lung squamous cell carcinoma</td>\n",
       "      <td>CD45+</td>\n",
       "      <td>tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1564</th>\n",
       "      <td>P18_T1</td>\n",
       "      <td>ERS16703118</td>\n",
       "      <td>SAMEA114591092</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 18</td>\n",
       "      <td>TB21.0006</td>\n",
       "      <td>77</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P18_T1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-features.tsv.gz</td>\n",
       "      <td>lung squamous cell carcinoma</td>\n",
       "      <td>CD45+</td>\n",
       "      <td>tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1565</th>\n",
       "      <td>P18_T1</td>\n",
       "      <td>ERS16703118</td>\n",
       "      <td>SAMEA114591092</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 18</td>\n",
       "      <td>TB21.0006</td>\n",
       "      <td>77</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P18_T1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-features.tsv.gz</td>\n",
       "      <td>lung squamous cell carcinoma</td>\n",
       "      <td>CD45+</td>\n",
       "      <td>tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1566</th>\n",
       "      <td>P18_T1</td>\n",
       "      <td>ERS16703118</td>\n",
       "      <td>SAMEA114591092</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 18</td>\n",
       "      <td>TB21.0006</td>\n",
       "      <td>77</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P18_T1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-features.tsv.gz</td>\n",
       "      <td>lung squamous cell carcinoma</td>\n",
       "      <td>CD45+</td>\n",
       "      <td>tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1567 rows Ã— 69 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Source Name Comment[ENA_SAMPLE] Comment[BioSD_SAMPLE]  \\\n",
       "0          P1_B1         ERS16703054        SAMEA114591031   \n",
       "1          P1_B1         ERS16703054        SAMEA114591031   \n",
       "2          P1_B1         ERS16703054        SAMEA114591031   \n",
       "3          P1_B1         ERS16703054        SAMEA114591031   \n",
       "4          P1_B1         ERS16703054        SAMEA114591031   \n",
       "...          ...                 ...                   ...   \n",
       "1562      P18_T1         ERS16703118        SAMEA114591092   \n",
       "1563      P18_T1         ERS16703118        SAMEA114591092   \n",
       "1564      P18_T1         ERS16703118        SAMEA114591092   \n",
       "1565      P18_T1         ERS16703118        SAMEA114591092   \n",
       "1566      P18_T1         ERS16703118        SAMEA114591092   \n",
       "\n",
       "     Characteristics[organism] Characteristics[individual]  \\\n",
       "0                 Homo sapiens                   Patient 1   \n",
       "1                 Homo sapiens                   Patient 1   \n",
       "2                 Homo sapiens                   Patient 1   \n",
       "3                 Homo sapiens                   Patient 1   \n",
       "4                 Homo sapiens                   Patient 1   \n",
       "...                        ...                         ...   \n",
       "1562              Homo sapiens                  Patient 18   \n",
       "1563              Homo sapiens                  Patient 18   \n",
       "1564              Homo sapiens                  Patient 18   \n",
       "1565              Homo sapiens                  Patient 18   \n",
       "1566              Homo sapiens                  Patient 18   \n",
       "\n",
       "     Characteristics[original source name]  Characteristics[age]  \\\n",
       "0                                TB18.0655                    73   \n",
       "1                                TB18.0655                    73   \n",
       "2                                TB18.0655                    73   \n",
       "3                                TB18.0655                    73   \n",
       "4                                TB18.0655                    73   \n",
       "...                                    ...                   ...   \n",
       "1562                             TB21.0006                    77   \n",
       "1563                             TB21.0006                    77   \n",
       "1564                             TB21.0006                    77   \n",
       "1565                             TB21.0006                    77   \n",
       "1566                             TB21.0006                    77   \n",
       "\n",
       "     Unit[time unit] Term Source REF Term Accession Number  ...  \\\n",
       "0               year             EFO            UO_0000036  ...   \n",
       "1               year             EFO            UO_0000036  ...   \n",
       "2               year             EFO            UO_0000036  ...   \n",
       "3               year             EFO            UO_0000036  ...   \n",
       "4               year             EFO            UO_0000036  ...   \n",
       "...              ...             ...                   ...  ...   \n",
       "1562            year             EFO            UO_0000036  ...   \n",
       "1563            year             EFO            UO_0000036  ...   \n",
       "1564            year             EFO            UO_0000036  ...   \n",
       "1565            year             EFO            UO_0000036  ...   \n",
       "1566            year             EFO            UO_0000036  ...   \n",
       "\n",
       "     Derived Array Data File Protocol REF.6 Protocol REF.7  \\\n",
       "0        P1_B1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1        P1_B1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "2        P1_B1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "3        P1_B1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "4        P1_B1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "...                      ...            ...            ...   \n",
       "1562    P18_T1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1563    P18_T1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1564    P18_T1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1565    P18_T1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1566    P18_T1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "\n",
       "     Derived Array Data File.1 Protocol REF.8 Protocol REF.9  \\\n",
       "0        P1_B1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1        P1_B1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "2        P1_B1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "3        P1_B1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "4        P1_B1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "...                        ...            ...            ...   \n",
       "1562    P18_T1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1563    P18_T1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1564    P18_T1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1565    P18_T1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1566    P18_T1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "\n",
       "     Derived Array Data File.2         Factor Value[disease]  \\\n",
       "0        P1_B1-features.tsv.gz    non-small cell lung cancer   \n",
       "1        P1_B1-features.tsv.gz    non-small cell lung cancer   \n",
       "2        P1_B1-features.tsv.gz    non-small cell lung cancer   \n",
       "3        P1_B1-features.tsv.gz    non-small cell lung cancer   \n",
       "4        P1_B1-features.tsv.gz    non-small cell lung cancer   \n",
       "...                        ...                           ...   \n",
       "1562    P18_T1-features.tsv.gz  lung squamous cell carcinoma   \n",
       "1563    P18_T1-features.tsv.gz  lung squamous cell carcinoma   \n",
       "1564    P18_T1-features.tsv.gz  lung squamous cell carcinoma   \n",
       "1565    P18_T1-features.tsv.gz  lung squamous cell carcinoma   \n",
       "1566    P18_T1-features.tsv.gz  lung squamous cell carcinoma   \n",
       "\n",
       "     Factor Value[FACS]      Factor Value[sampling site]  \n",
       "0                  MDSC  normal tissue adjacent to tumor  \n",
       "1                  MDSC  normal tissue adjacent to tumor  \n",
       "2                  MDSC  normal tissue adjacent to tumor  \n",
       "3                  MDSC  normal tissue adjacent to tumor  \n",
       "4                  MDSC  normal tissue adjacent to tumor  \n",
       "...                 ...                              ...  \n",
       "1562              CD45+                            tumor  \n",
       "1563              CD45+                            tumor  \n",
       "1564              CD45+                            tumor  \n",
       "1565              CD45+                            tumor  \n",
       "1566              CD45+                            tumor  \n",
       "\n",
       "[1567 rows x 69 columns]"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zuani_dir = f'{data_dir}/Zuani2024'\n",
    "zuani_samples = pd.read_csv(f'{zuani_dir}/E-MTAB-13526.sdrf.txt', sep='\\t')\n",
    "\n",
    "# Filter for non enriched MDSC = Myeloid derived supressor cells, CD235a- filters for erythrocites\n",
    "zuani_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "546d01dd-5a3c-4eb4-9a3a-e2f5447ce1e5",
   "metadata": {},
   "source": [
    "We filter for non-immune enriched samples and only tumor tissue. All the non-tumor enirhced ones are filtered for no plasma cells (erythrocites CD235a-)\n",
    "that is also the case in many of the other samples... and we can also enrich for imput ethe as they are only 1 of the many cell types."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "4fb1b2dd-900d-4ca5-adf7-e0c6c8ba8d3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "facs_filter = ['CD235a-']\n",
    "zuani_samples = zuani_samples[zuani_samples['Factor Value[FACS]'].isin(facs_filter)]\n",
    "\n",
    "zuani_samples = zuani_samples[zuani_samples['Factor Value[sampling site]']=='tumor']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "6f188cc0-e8b4-433d-bc20-13054845bb2a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPUS:  8\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P1_T1-barcodes.tsv.gz\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P1_T1-features.tsv.gz\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P1_T1-matrix.mtx.gz\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P2_T1-barcodes.tsv.gz\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P2_T1-features.tsv.gz\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P2_T1-matrix.mtx.gz\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P15_T2-barcodes.tsv.gz\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P15_T2-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P2_T1-features.tsv.gz time (s): 1.5921189785003662\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P15_T2-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P1_T1-features.tsv.gz time (s): 1.5972809791564941\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P16_T2-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P15_T2-features.tsv.gz time (s): 1.3004364967346191\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P16_T2-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P1_T1-barcodes.tsv.gz time (s): 6.099670171737671\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P16_T2-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P16_T2-features.tsv.gz time (s): 1.1689391136169434\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T2-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P2_T1-matrix.mtx.gz time (s): 25.24450135231018\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T2-features.tsv.gzurl: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P2_T1-barcodes.tsv.gz time (s): 43.2615282535553\n",
      "\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T2-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P15_T2-barcodes.tsv.gz time (s): 43.600990295410156\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T3-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T2-features.tsv.gz time (s): 1.2455544471740723\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T3-features.tsv.gzurl: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P16_T2-barcodes.tsv.gz time (s): 42.82427263259888\n",
      "\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T3-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T3-features.tsv.gz time (s): 1.2398083209991455\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T2-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P1_T1-matrix.mtx.gz time (s): 52.44864463806152\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T2-features.tsv.gzurl: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T2-barcodes.tsv.gz time (s): 42.84986734390259\n",
      "\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T2-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T2-features.tsv.gz time (s): 1.2251718044281006\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T3-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T3-barcodes.tsv.gz time (s): 42.83008527755737\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T3-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T2-barcodes.tsv.gz time (s): 42.71306538581848\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T3-matrix.mtx.gzurl: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T3-features.tsv.gz time (s): 1.2472662925720215\n",
      "\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P20_T2-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T2-matrix.mtx.gz time (s): 76.8129460811615\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P20_T2-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T3-barcodes.tsv.gz time (s): 42.92999482154846\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P20_T2-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P20_T2-features.tsv.gz time (s): 1.2489855289459229\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P18_T2-barcodes.tsv.gzurl: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T3-matrix.mtx.gz time (s): 95.85074806213379\n",
      "\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P18_T2-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P16_T2-matrix.mtx.gz time (s): 142.04853892326355\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P18_T2-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P18_T2-features.tsv.gz time (s): 1.438136339187622\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P19_T2-barcodes.tsv.gzurl: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P20_T2-barcodes.tsv.gz time (s): 42.95699501037598\n",
      "\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P19_T2-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P18_T2-barcodes.tsv.gz time (s): 42.92334604263306\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P19_T2-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P19_T2-features.tsv.gz time (s): 1.2489831447601318\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P24_T1-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P19_T2-barcodes.tsv.gz time (s): 42.852856397628784\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P24_T1-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P15_T2-matrix.mtx.gz time (s): 210.2394998073578\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P24_T1-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P24_T1-features.tsv.gz time (s): 1.2681214809417725\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P22_T1-barcodes.tsv.gzurl: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P24_T1-barcodes.tsv.gz time (s): 42.93040895462036\n",
      "\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P22_T1-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P22_T1-barcodes.tsv.gz time (s): 42.95659422874451\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P22_T1-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P22_T1-features.tsv.gz time (s): 1.2519481182098389\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P8_T2-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P22_T1-matrix.mtx.gz time (s): 83.76424765586853\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P8_T2-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P8_T2-barcodes.tsv.gz time (s): 43.164130449295044\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P8_T2-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P8_T2-features.tsv.gz time (s): 1.2520792484283447\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T1-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P20_T2-matrix.mtx.gz time (s): 340.315548658371\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T1-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P19_T2-matrix.mtx.gz time (s): 318.12532663345337\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T1-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T1-features.tsv.gz time (s): 1.2496387958526611\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P23_T1-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T1-barcodes.tsv.gz time (s): 43.02663064002991\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P23_T1-features.tsv.gzurl: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P18_T2-matrix.mtx.gz time (s): 405.18058800697327\n",
      "\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P23_T1-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P23_T1-features.tsv.gz time (s): 1.375727891921997\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T2-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P23_T1-barcodes.tsv.gz time (s): 42.89959907531738\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T2-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T2-matrix.mtx.gz time (s): 505.055460691452\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T2-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T2-features.tsv.gz time (s): 1.2500417232513428\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T2-matrix.mtx.gz time (s): 23.810145139694214\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T2-barcodes.tsv.gz time (s): 43.953601360321045\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T1-matrix.mtx.gz time (s): 133.2835156917572\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T3-matrix.mtx.gz time (s): 639.7812283039093\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P24_T1-matrix.mtx.gz time (s): 703.8568208217621\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P8_T2-matrix.mtx.gz time (s): 500.3829493522644\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P23_T1-matrix.mtx.gz time (s): 617.2215695381165\n"
     ]
    }
   ],
   "source": [
    "base_url = 'https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files'\n",
    "zuani_patients = zuani_samples['Source Name'].unique()\n",
    "\n",
    "urls_paths = []\n",
    "\n",
    "for pat in zuani_patients:\n",
    "    filename = f'{pat}-barcodes.tsv.gz'\n",
    "    urls_paths.append((f'{base_url}/{filename}',f'{zuani_dir}/{filename}'))\n",
    "\n",
    "    filename = f'{pat}-features.tsv.gz'\n",
    "    urls_paths.append((f'{base_url}/{filename}',f'{zuani_dir}/{filename}'))\n",
    "\n",
    "    filename = f'{pat}-matrix.mtx.gz'\n",
    "    urls_paths.append((f'{base_url}/{filename}',f'{zuani_dir}/{filename}'))\n",
    "\n",
    "ut.download_parallel(urls_paths, cpus=8)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c69ad9fb-439e-4150-87ee-5e2dfd53bfa9",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Convert to adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "id": "2c1b051d-478a-4a99-8c16-71d810eaedd9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P15_T2-\n",
      "P16_T2-\n",
      "P17_T2-\n",
      "P17_T3-\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[187], line 45\u001b[0m\n\u001b[1;32m     42\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m concatenated_adata\n\u001b[1;32m     44\u001b[0m \u001b[38;5;66;03m# Example usage:\u001b[39;00m\n\u001b[0;32m---> 45\u001b[0m concatenated_adata2 \u001b[38;5;241m=\u001b[39m \u001b[43mconcatenate_datasets\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43mf\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43mzuani_dir\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m     46\u001b[0m \u001b[38;5;28mprint\u001b[39m(concatenated_adata2)\n",
      "Cell \u001b[0;32mIn[187], line 29\u001b[0m, in \u001b[0;36mconcatenate_datasets\u001b[0;34m(directory)\u001b[0m\n\u001b[1;32m     26\u001b[0m     features_file \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00midx\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124mfeatures.tsv\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m     28\u001b[0m     \u001b[38;5;66;03m# Load data and append to list\u001b[39;00m\n\u001b[0;32m---> 29\u001b[0m     adata \u001b[38;5;241m=\u001b[39m \u001b[43msc\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_10x_mtx\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdirectory\u001b[49m\u001b[43m,\u001b[49m\u001b[43mprefix\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43midx\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     30\u001b[0m     adatas[idx[:\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m]] \u001b[38;5;241m=\u001b[39m adata\n\u001b[1;32m     32\u001b[0m \u001b[38;5;66;03m# Concatenate all AnnData objects\u001b[39;00m\n",
      "File \u001b[0;32m/usr/local/lib/python3.11/site-packages/scanpy/readwrite.py:489\u001b[0m, in \u001b[0;36mread_10x_mtx\u001b[0;34m(path, var_names, make_unique, cache, cache_compression, gex_only, prefix)\u001b[0m\n\u001b[1;32m    487\u001b[0m genefile_exists \u001b[38;5;241m=\u001b[39m (path \u001b[38;5;241m/\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mprefix\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124mgenes.tsv\u001b[39m\u001b[38;5;124m'\u001b[39m)\u001b[38;5;241m.\u001b[39mis_file()\n\u001b[1;32m    488\u001b[0m read \u001b[38;5;241m=\u001b[39m _read_legacy_10x_mtx \u001b[38;5;28;01mif\u001b[39;00m genefile_exists \u001b[38;5;28;01melse\u001b[39;00m _read_v3_10x_mtx\n\u001b[0;32m--> 489\u001b[0m adata \u001b[38;5;241m=\u001b[39m \u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    490\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mstr\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mpath\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    491\u001b[0m \u001b[43m    \u001b[49m\u001b[43mvar_names\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mvar_names\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    492\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmake_unique\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmake_unique\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    493\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcache\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcache\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    494\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcache_compression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcache_compression\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    495\u001b[0m \u001b[43m    \u001b[49m\u001b[43mprefix\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mprefix\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    496\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    497\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m genefile_exists \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m gex_only:\n\u001b[1;32m    498\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m adata\n",
      "File \u001b[0;32m/usr/local/lib/python3.11/site-packages/scanpy/readwrite.py:553\u001b[0m, in \u001b[0;36m_read_v3_10x_mtx\u001b[0;34m(path, var_names, make_unique, cache, cache_compression, prefix)\u001b[0m\n\u001b[1;32m    549\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    550\u001b[0m \u001b[38;5;124;03mRead mtx from output from Cell Ranger v3 or later versions\u001b[39;00m\n\u001b[1;32m    551\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    552\u001b[0m path \u001b[38;5;241m=\u001b[39m Path(path)\n\u001b[0;32m--> 553\u001b[0m adata \u001b[38;5;241m=\u001b[39m \u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    554\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpath\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m/\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43mf\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43mprefix\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43mmatrix.mtx.gz\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    555\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcache\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcache\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    556\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcache_compression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcache_compression\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    557\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39mT  \u001b[38;5;66;03m# transpose the data\u001b[39;00m\n\u001b[1;32m    558\u001b[0m genes \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(path \u001b[38;5;241m/\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mprefix\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124mfeatures.tsv.gz\u001b[39m\u001b[38;5;124m'\u001b[39m, header\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, sep\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m    559\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m var_names \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgene_symbols\u001b[39m\u001b[38;5;124m'\u001b[39m:\n",
      "File \u001b[0;32m/usr/local/lib/python3.11/site-packages/scanpy/readwrite.py:111\u001b[0m, in \u001b[0;36mread\u001b[0;34m(filename, backed, sheet, ext, delimiter, first_column_names, backup_url, cache, cache_compression, **kwargs)\u001b[0m\n\u001b[1;32m    109\u001b[0m filename \u001b[38;5;241m=\u001b[39m Path(filename)  \u001b[38;5;66;03m# allow passing strings\u001b[39;00m\n\u001b[1;32m    110\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_valid_filename(filename):\n\u001b[0;32m--> 111\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    112\u001b[0m \u001b[43m        \u001b[49m\u001b[43mfilename\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    113\u001b[0m \u001b[43m        \u001b[49m\u001b[43mbacked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbacked\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    114\u001b[0m \u001b[43m        \u001b[49m\u001b[43msheet\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msheet\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    115\u001b[0m \u001b[43m        \u001b[49m\u001b[43mext\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mext\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    116\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdelimiter\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdelimiter\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    117\u001b[0m \u001b[43m        \u001b[49m\u001b[43mfirst_column_names\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfirst_column_names\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    118\u001b[0m \u001b[43m        \u001b[49m\u001b[43mbackup_url\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbackup_url\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    119\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcache\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcache\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    120\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcache_compression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcache_compression\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    121\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    122\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    123\u001b[0m \u001b[38;5;66;03m# generate filename and read to dict\u001b[39;00m\n\u001b[1;32m    124\u001b[0m filekey \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mstr\u001b[39m(filename)\n",
      "File \u001b[0;32m/usr/local/lib/python3.11/site-packages/scanpy/readwrite.py:750\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filename, backed, sheet, ext, delimiter, first_column_names, backup_url, cache, cache_compression, suppress_cache_warning, **kwargs)\u001b[0m\n\u001b[1;32m    748\u001b[0m         adata \u001b[38;5;241m=\u001b[39m read_excel(filename, sheet)\n\u001b[1;32m    749\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m ext \u001b[38;5;129;01min\u001b[39;00m {\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmtx\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmtx.gz\u001b[39m\u001b[38;5;124m'\u001b[39m}:\n\u001b[0;32m--> 750\u001b[0m     adata \u001b[38;5;241m=\u001b[39m \u001b[43mread_mtx\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilename\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    751\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m ext \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcsv\u001b[39m\u001b[38;5;124m'\u001b[39m:\n\u001b[1;32m    752\u001b[0m     adata \u001b[38;5;241m=\u001b[39m read_csv(filename, first_column_names\u001b[38;5;241m=\u001b[39mfirst_column_names)\n",
      "File \u001b[0;32m/usr/local/lib/python3.11/site-packages/anndata/_io/read.py:317\u001b[0m, in \u001b[0;36mread_mtx\u001b[0;34m(filename, dtype)\u001b[0m\n\u001b[1;32m    314\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mscipy\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mio\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m mmread\n\u001b[1;32m    316\u001b[0m \u001b[38;5;66;03m# could be rewritten accounting for dtype to be more performant\u001b[39;00m\n\u001b[0;32m--> 317\u001b[0m X \u001b[38;5;241m=\u001b[39m \u001b[43mmmread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfspath\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilename\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39mastype(dtype)\n\u001b[1;32m    318\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mscipy\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01msparse\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m csr_matrix\n\u001b[1;32m    320\u001b[0m X \u001b[38;5;241m=\u001b[39m csr_matrix(X)\n",
      "File \u001b[0;32m/usr/local/lib/python3.11/site-packages/scipy/io/_mmio.py:129\u001b[0m, in \u001b[0;36mmmread\u001b[0;34m(source)\u001b[0m\n\u001b[1;32m     84\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mmmread\u001b[39m(source):\n\u001b[1;32m     85\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m     86\u001b[0m \u001b[38;5;124;03m    Reads the contents of a Matrix Market file-like 'source' into a matrix.\u001b[39;00m\n\u001b[1;32m     87\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    127\u001b[0m \u001b[38;5;124;03m           [0., 0., 0., 0., 0.]])\u001b[39;00m\n\u001b[1;32m    128\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 129\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mMMFile\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43msource\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/local/lib/python3.11/site-packages/scipy/io/_mmio.py:582\u001b[0m, in \u001b[0;36mMMFile.read\u001b[0;34m(self, source)\u001b[0m\n\u001b[1;32m    580\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    581\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_parse_header(stream)\n\u001b[0;32m--> 582\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_parse_body\u001b[49m\u001b[43m(\u001b[49m\u001b[43mstream\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    584\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m    585\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m close_it:\n",
      "File \u001b[0;32m/usr/local/lib/python3.11/site-packages/scipy/io/_mmio.py:748\u001b[0m, in \u001b[0;36mMMFile._parse_body\u001b[0;34m(self, stream)\u001b[0m\n\u001b[1;32m    746\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m is_pattern:\n\u001b[1;32m    747\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m is_integer:\n\u001b[0;32m--> 748\u001b[0m         V[entry_number] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mint\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43ml\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m2\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    749\u001b[0m     \u001b[38;5;28;01melif\u001b[39;00m is_unsigned_integer:\n\u001b[1;32m    750\u001b[0m         V[entry_number] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mint\u001b[39m(l[\u001b[38;5;241m2\u001b[39m])\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import os\n",
    "import io\n",
    "import csv\n",
    "import gzip\n",
    "\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "\n",
    "def concatenate_datasets(directory):\n",
    "    \"\"\"\n",
    "    Concatenate multiple .mtx datasets from a single directory into a single AnnData object.\n",
    "    \n",
    "    Parameters:\n",
    "    - directory (str): The directory containing .mtx, features.tsv, and barcodes.tsv files.\n",
    "    \n",
    "    Returns:\n",
    "    - AnnData: The concatenated AnnData object.\n",
    "    \"\"\"\n",
    "    adatas = {}\n",
    "    features_file = ''\n",
    "    \n",
    "    file_pattern = sorted([f for f in os.listdir(directory) if 'matrix' in f])\n",
    "    for mtx_file in file_pattern:\n",
    "        idx = mtx_file.replace('matrix.mtx.gz', '')\n",
    "        print(idx)\n",
    "        features_file = f'{idx}features.tsv'\n",
    "        \n",
    "        # Load data and append to list\n",
    "        adata = sc.read_10x_mtx(directory,prefix=idx)\n",
    "        adatas[idx[:-1]] = adata\n",
    "    \n",
    "    # Concatenate all AnnData objects\n",
    "    concatenated_adata = ad.concat(adatas, label='sample')\n",
    "\n",
    "    gene_names = []\n",
    "    with gzip.open(features_file, 'rb') as f:\n",
    "        reader = csv.reader(io.TextIOWrapper(f, encoding='utf-8'), delimiter='\\t')\n",
    "        for row in reader:\n",
    "            gene_names.append(row[1])\n",
    "    concatenated_adata.var['gene_symbols'] = gene_names\n",
    "\n",
    "    return concatenated_adata\n",
    "\n",
    "# Example usage:\n",
    "adata = concatenate_datasets(f'{zuani_dir}')\n",
    "print(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "dd8a6b40-c97d-4f8e-a844-4bae0cabf068",
   "metadata": {},
   "outputs": [],
   "source": [
    "concatenated_adata.write_h5ad('Zuani.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8999abd6-a111-4106-be8f-7ab6d1a81968",
   "metadata": {},
   "source": [
    "## Deng 2024"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "id": "93f4066a-b712-4504-afad-319d866e1e65",
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'curl'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[193], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mcurl\u001b[39;00m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'curl'"
     ]
    }
   ],
   "source": [
    "urllib.request.urlretrieve('https://figshare.com/ndownloader/files/44695465',f'{data_dir}/Deng2024/dengluad.rds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "id": "2710b1b5-0cf7-4701-99f1-366384940f58",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Validating h5Seurat file\n",
       "Adding data from RNA as X\n",
       "Adding counts from RNA as raw\n",
       "Transfering meta.data to obs\n",
       "Adding dimensional reduction information for tsne (global)\n",
       "Adding dimensional reduction information for umap (global)\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i data_dir\n",
    "library(Seurat)\n",
    "library(SeuratDisk)\n",
    "\n",
    "deng <- readRDS(paste0(data_dir, '/Deng2024/dengluad.rds'))\n",
    "deng = UpdateSeuratObject(deng)\n",
    "\n",
    "# After P49 there is no metadata so better delete. Also the ones with N as they are healthy controls,\n",
    "cells_to_exclude <- which(grepl(\"P05|N\", deng@meta.data$orig.ident))\n",
    "deng <- subset(deng, cells = setdiff(Cells(deng), cells_to_exclude))\n",
    "\n",
    "\n",
    "SaveH5Seurat(deng, filename=paste0(data_dir, '/Deng2024/deng.h5Seurat'))\n",
    "Convert(paste0(data_dir, '/Deng2024/deng.h5Seurat'), dest = \"h5ad\")\n",
    "Convert(paste0(data_dir, '/Deng2024/deng.h5Seurat'), dest = paste0(data_dir, '/Deng2024/dengRNA.h5ad'), assay = 'RNA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf03290c-48e7-4835-9e6c-8b9d3193e9c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "! rm {data_dir}/Deng2024/deng.h5Seurat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ff2ee1a-a284-48bd-8238-0a0c899b3542",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# adata = sc.read_h5ad(f'{data_dir}/Deng2024/deng.h5ad')\n",
    "adata_raw = sc.read_h5ad(f'{data_dir}/Deng2024/dengRNA.h5ad')\n",
    "\n",
    "# After P49 there is no metadata so better delete. Also the ones with N as they are healthy controls,\n",
    "# Should be done before in R to be faster\n",
    "# adata = adata[np.logical_not(adata.obs['orig.ident'].str.contains('P05'))]\n",
    "# adata = adata[np.logical_not(adata.obs['orig.ident'].str.contains('N'))]\n",
    "\n",
    "# The assignment fo slots and assays is not very good with Convert from SeuratDisk so we make some arrangements\n",
    "adata.layers['data'] = adata.raw.X.copy()\n",
    "del adata.raw\n",
    "adata.layers['scale.data'] = adata.X\n",
    "\n",
    "adata_raw.layers['counts'] = adata_raw.raw.X.copy()\n",
    "del adata_raw.raw\n",
    "adata_raw.layers['data'] = adata_raw.X\n",
    "\n",
    "adata.raw = adata_raw.copy()\n",
    "\n",
    "del adata_raw\n",
    "\n",
    "path_metadata = '/root/datos/maestria/netopaas/Zuani2024/metadata.xlsx'\n",
    "if not os.path.exists(path_metadata):\n",
    "    urllib.request.urlretrieve('https://www.ncbi.nlm.nih.gov/pmc/articles/PMC11031428/bin/mmc2.xlsx', path_metadata)\n",
    "xl_zuani = pd.ExcelFile(path_metadata)\n",
    "metada_zuani = xl_zuani.parse(' lung cancer in scRNAseq')\n",
    "adata.obs = pd.merge(adata.obs, metada_zuani, how='left', left_on='orig.ident', right_on='Patient Number')\n",
    "\n",
    "\n",
    "adata.write_h5ad(f'{backup_dir}/filtered_deng.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b60d5673-27f4-4ab8-b551-6b84c3ba0aa9",
   "metadata": {},
   "source": [
    "## Alexandra 2023"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "1e7ee161-c989-4081-a668-f06834bef03d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/root/datos/maestria/netopaas//alexandra2023/data.zip',\n",
       " <http.client.HTTPMessage at 0x7f7325a5cf10>)"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "urllib.request.urlretrieve('https://zenodo.org/records/7852154/files/data_code.zip?download=1',f'{data_dir}/Alexandra2023/data.zip')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5628b30-65fc-40ab-abc2-7a9d1a8190b8",
   "metadata": {},
   "source": [
    "## Zhong 2024\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "cc7923b6-6f52-44f5-9135-377b38a3aa69",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>percent.mt</th>\n",
       "      <th>sampleID</th>\n",
       "      <th>cellID</th>\n",
       "      <th>RNA_snn_res.0.8</th>\n",
       "      <th>seurat_clusters</th>\n",
       "      <th>major.cell.type</th>\n",
       "      <th>RNA_snn_res.1</th>\n",
       "      <th>...</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Age</th>\n",
       "      <th>Histology</th>\n",
       "      <th>Cycles</th>\n",
       "      <th>PD1</th>\n",
       "      <th>Pathological Response</th>\n",
       "      <th>EGFR</th>\n",
       "      <th>Pathological Response Rate</th>\n",
       "      <th>Smoking_History</th>\n",
       "      <th>PD-L1 TPS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P343</td>\n",
       "      <td>17606</td>\n",
       "      <td>4574</td>\n",
       "      <td>2.891060</td>\n",
       "      <td>P343</td>\n",
       "      <td>P343_AAACCTGAGCTATGCT-1</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>Fibro</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>F</td>\n",
       "      <td>60</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>L858R</td>\n",
       "      <td>0.85</td>\n",
       "      <td>N</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P343</td>\n",
       "      <td>9542</td>\n",
       "      <td>3230</td>\n",
       "      <td>4.003354</td>\n",
       "      <td>P343</td>\n",
       "      <td>P343_AAACCTGCATGATCCA-1</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>Fibro</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>F</td>\n",
       "      <td>60</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>L858R</td>\n",
       "      <td>0.85</td>\n",
       "      <td>N</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P343</td>\n",
       "      <td>10729</td>\n",
       "      <td>3832</td>\n",
       "      <td>2.302172</td>\n",
       "      <td>P343</td>\n",
       "      <td>P343_AAACCTGGTGAGTATA-1</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>Endo</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>F</td>\n",
       "      <td>60</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>L858R</td>\n",
       "      <td>0.85</td>\n",
       "      <td>N</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P343</td>\n",
       "      <td>7788</td>\n",
       "      <td>2859</td>\n",
       "      <td>5.225989</td>\n",
       "      <td>P343</td>\n",
       "      <td>P343_AAACCTGTCCAGTATG-1</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>Endo</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>F</td>\n",
       "      <td>60</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>L858R</td>\n",
       "      <td>0.85</td>\n",
       "      <td>N</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P343</td>\n",
       "      <td>15449</td>\n",
       "      <td>4074</td>\n",
       "      <td>3.301185</td>\n",
       "      <td>P343</td>\n",
       "      <td>P343_AAACGGGCACGACGAA-1</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>Endo</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>F</td>\n",
       "      <td>60</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>L858R</td>\n",
       "      <td>0.85</td>\n",
       "      <td>N</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78686</th>\n",
       "      <td>P591</td>\n",
       "      <td>6608</td>\n",
       "      <td>2528</td>\n",
       "      <td>3.344431</td>\n",
       "      <td>P591</td>\n",
       "      <td>P591_TGCCCTAGTCAGATAA-1</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>Epi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>51</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>KDD</td>\n",
       "      <td>0.58</td>\n",
       "      <td>Y</td>\n",
       "      <td>&lt;1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78687</th>\n",
       "      <td>P591</td>\n",
       "      <td>1064</td>\n",
       "      <td>707</td>\n",
       "      <td>7.518797</td>\n",
       "      <td>P591</td>\n",
       "      <td>P591_TGGCTGGAGCGTTGCC-1</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>Epi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>51</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>KDD</td>\n",
       "      <td>0.58</td>\n",
       "      <td>Y</td>\n",
       "      <td>&lt;1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78688</th>\n",
       "      <td>P591</td>\n",
       "      <td>13226</td>\n",
       "      <td>3318</td>\n",
       "      <td>1.126569</td>\n",
       "      <td>P591</td>\n",
       "      <td>P591_TTCTCCTAGTGCGTGA-1</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>Epi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>51</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>KDD</td>\n",
       "      <td>0.58</td>\n",
       "      <td>Y</td>\n",
       "      <td>&lt;1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78689</th>\n",
       "      <td>P591</td>\n",
       "      <td>7335</td>\n",
       "      <td>2912</td>\n",
       "      <td>6.693933</td>\n",
       "      <td>P591</td>\n",
       "      <td>P591_TTTCCTCAGCGTAATA-1</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>Epi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>51</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>KDD</td>\n",
       "      <td>0.58</td>\n",
       "      <td>Y</td>\n",
       "      <td>&lt;1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78690</th>\n",
       "      <td>P591</td>\n",
       "      <td>14498</td>\n",
       "      <td>4608</td>\n",
       "      <td>1.448476</td>\n",
       "      <td>P591</td>\n",
       "      <td>P591_TTTGGTTCATGGGACA-1</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>Epi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>51</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>KDD</td>\n",
       "      <td>0.58</td>\n",
       "      <td>Y</td>\n",
       "      <td>&lt;1%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78691 rows Ã— 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      orig.ident  nCount_RNA  nFeature_RNA  percent.mt sampleID  \\\n",
       "0           P343       17606          4574    2.891060     P343   \n",
       "1           P343        9542          3230    4.003354     P343   \n",
       "2           P343       10729          3832    2.302172     P343   \n",
       "3           P343        7788          2859    5.225989     P343   \n",
       "4           P343       15449          4074    3.301185     P343   \n",
       "...          ...         ...           ...         ...      ...   \n",
       "78686       P591        6608          2528    3.344431     P591   \n",
       "78687       P591        1064           707    7.518797     P591   \n",
       "78688       P591       13226          3318    1.126569     P591   \n",
       "78689       P591        7335          2912    6.693933     P591   \n",
       "78690       P591       14498          4608    1.448476     P591   \n",
       "\n",
       "                        cellID  RNA_snn_res.0.8  seurat_clusters  \\\n",
       "0      P343_AAACCTGAGCTATGCT-1                6                6   \n",
       "1      P343_AAACCTGCATGATCCA-1                8                8   \n",
       "2      P343_AAACCTGGTGAGTATA-1               10               10   \n",
       "3      P343_AAACCTGTCCAGTATG-1               10               10   \n",
       "4      P343_AAACGGGCACGACGAA-1               10               10   \n",
       "...                        ...              ...              ...   \n",
       "78686  P591_TGCCCTAGTCAGATAA-1                7                7   \n",
       "78687  P591_TGGCTGGAGCGTTGCC-1                7                7   \n",
       "78688  P591_TTCTCCTAGTGCGTGA-1                7                7   \n",
       "78689  P591_TTTCCTCAGCGTAATA-1                7                7   \n",
       "78690  P591_TTTGGTTCATGGGACA-1               20               20   \n",
       "\n",
       "      major.cell.type  RNA_snn_res.1  ... Gender  Age Histology Cycles  \\\n",
       "0               Fibro            NaN  ...      F   60      LUAD      3   \n",
       "1               Fibro            NaN  ...      F   60      LUAD      3   \n",
       "2                Endo            NaN  ...      F   60      LUAD      3   \n",
       "3                Endo            NaN  ...      F   60      LUAD      3   \n",
       "4                Endo            NaN  ...      F   60      LUAD      3   \n",
       "...               ...            ...  ...    ...  ...       ...    ...   \n",
       "78686             Epi            NaN  ...      M   51      LUAD      3   \n",
       "78687             Epi            NaN  ...      M   51      LUAD      3   \n",
       "78688             Epi            NaN  ...      M   51      LUAD      3   \n",
       "78689             Epi            NaN  ...      M   51      LUAD      3   \n",
       "78690             Epi            NaN  ...      M   51      LUAD      3   \n",
       "\n",
       "              PD1 Pathological Response   EGFR Pathological Response Rate  \\\n",
       "0      Sintilimab               non-MPR  L858R                       0.85   \n",
       "1      Sintilimab               non-MPR  L858R                       0.85   \n",
       "2      Sintilimab               non-MPR  L858R                       0.85   \n",
       "3      Sintilimab               non-MPR  L858R                       0.85   \n",
       "4      Sintilimab               non-MPR  L858R                       0.85   \n",
       "...           ...                   ...    ...                        ...   \n",
       "78686  Sintilimab               non-MPR    KDD                       0.58   \n",
       "78687  Sintilimab               non-MPR    KDD                       0.58   \n",
       "78688  Sintilimab               non-MPR    KDD                       0.58   \n",
       "78689  Sintilimab               non-MPR    KDD                       0.58   \n",
       "78690  Sintilimab               non-MPR    KDD                       0.58   \n",
       "\n",
       "      Smoking_History PD-L1 TPS  \n",
       "0                   N      0.01  \n",
       "1                   N      0.01  \n",
       "2                   N      0.01  \n",
       "3                   N      0.01  \n",
       "4                   N      0.01  \n",
       "...               ...       ...  \n",
       "78686               Y       <1%  \n",
       "78687               Y       <1%  \n",
       "78688               Y       <1%  \n",
       "78689               Y       <1%  \n",
       "78690               Y       <1%  \n",
       "\n",
       "[78691 rows x 23 columns]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zhong_samples = pd.read_csv(f'{data_dir}/Zhong2024/GSE241934_IIT_Meta.txt', sep='\\t')\n",
    "zhong_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c04af7d-8683-4fbb-992c-25da5ea97805",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
