{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a80560b2-00f4-4d4f-9d13-2b5c48e057d0",
   "metadata": {},
   "source": [
    "# Download the scRNA data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "32151548-2c85-4ced-887e-2523ab68a156",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = '/root/datos/maestria/netopaas/'\n",
    "backup_dir = '/root/datos/maestria/netopaas/luca_explore/surgeries'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "8a4a47b6-e24d-4170-9d4a-e3566be33eae",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_33074/1855495475.py:23: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  anndata2ri.activate()\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import utils.functions as ut\n",
    "import urllib\n",
    "\n",
    "import scanpy as sc\n",
    "\n",
    "import rpy2.rinterface_lib.callbacks\n",
    "import logging\n",
    "from rpy2.robjects import pandas2ri\n",
    "import rpy2.robjects as robjects\n",
    "\n",
    "# Set the graphics device\n",
    "robjects.r('options(bitmapType=\"cairo\")')\n",
    "import anndata2ri\n",
    "\n",
    "# Ignore R warning messages\n",
    "#Note: this can be commented out to get more verbose R output\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)\n",
    "\n",
    "# Automatically convert rpy2 outputs to pandas dataframes\n",
    "pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68ea9660-e216-414c-aca4-11143ca39a33",
   "metadata": {},
   "source": [
    "## Zuani 2024"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a0630a4-3ef6-40a0-aa9d-f6f4abd79cbf",
   "metadata": {},
   "source": [
    "### Download"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "77a1cbc9-e3cb-4cdf-9e76-28d3802cc92f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Source Name</th>\n",
       "      <th>Comment[ENA_SAMPLE]</th>\n",
       "      <th>Comment[BioSD_SAMPLE]</th>\n",
       "      <th>Characteristics[organism]</th>\n",
       "      <th>Characteristics[individual]</th>\n",
       "      <th>Characteristics[original source name]</th>\n",
       "      <th>Characteristics[age]</th>\n",
       "      <th>Unit[time unit]</th>\n",
       "      <th>Term Source REF</th>\n",
       "      <th>Term Accession Number</th>\n",
       "      <th>...</th>\n",
       "      <th>Derived Array Data File</th>\n",
       "      <th>Protocol REF.6</th>\n",
       "      <th>Protocol REF.7</th>\n",
       "      <th>Derived Array Data File.1</th>\n",
       "      <th>Protocol REF.8</th>\n",
       "      <th>Protocol REF.9</th>\n",
       "      <th>Derived Array Data File.2</th>\n",
       "      <th>Factor Value[disease]</th>\n",
       "      <th>Factor Value[FACS]</th>\n",
       "      <th>Factor Value[sampling site]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P1_B1</td>\n",
       "      <td>ERS16703054</td>\n",
       "      <td>SAMEA114591031</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 1</td>\n",
       "      <td>TB18.0655</td>\n",
       "      <td>73</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P1_B1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-features.tsv.gz</td>\n",
       "      <td>non-small cell lung cancer</td>\n",
       "      <td>MDSC</td>\n",
       "      <td>normal tissue adjacent to tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P1_B1</td>\n",
       "      <td>ERS16703054</td>\n",
       "      <td>SAMEA114591031</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 1</td>\n",
       "      <td>TB18.0655</td>\n",
       "      <td>73</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P1_B1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-features.tsv.gz</td>\n",
       "      <td>non-small cell lung cancer</td>\n",
       "      <td>MDSC</td>\n",
       "      <td>normal tissue adjacent to tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P1_B1</td>\n",
       "      <td>ERS16703054</td>\n",
       "      <td>SAMEA114591031</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 1</td>\n",
       "      <td>TB18.0655</td>\n",
       "      <td>73</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P1_B1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-features.tsv.gz</td>\n",
       "      <td>non-small cell lung cancer</td>\n",
       "      <td>MDSC</td>\n",
       "      <td>normal tissue adjacent to tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P1_B1</td>\n",
       "      <td>ERS16703054</td>\n",
       "      <td>SAMEA114591031</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 1</td>\n",
       "      <td>TB18.0655</td>\n",
       "      <td>73</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P1_B1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-features.tsv.gz</td>\n",
       "      <td>non-small cell lung cancer</td>\n",
       "      <td>MDSC</td>\n",
       "      <td>normal tissue adjacent to tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P1_B1</td>\n",
       "      <td>ERS16703054</td>\n",
       "      <td>SAMEA114591031</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 1</td>\n",
       "      <td>TB18.0655</td>\n",
       "      <td>73</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P1_B1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P1_B1-features.tsv.gz</td>\n",
       "      <td>non-small cell lung cancer</td>\n",
       "      <td>MDSC</td>\n",
       "      <td>normal tissue adjacent to tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1562</th>\n",
       "      <td>P18_T1</td>\n",
       "      <td>ERS16703118</td>\n",
       "      <td>SAMEA114591092</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 18</td>\n",
       "      <td>TB21.0006</td>\n",
       "      <td>77</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P18_T1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-features.tsv.gz</td>\n",
       "      <td>lung squamous cell carcinoma</td>\n",
       "      <td>CD45+</td>\n",
       "      <td>tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1563</th>\n",
       "      <td>P18_T1</td>\n",
       "      <td>ERS16703118</td>\n",
       "      <td>SAMEA114591092</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 18</td>\n",
       "      <td>TB21.0006</td>\n",
       "      <td>77</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P18_T1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-features.tsv.gz</td>\n",
       "      <td>lung squamous cell carcinoma</td>\n",
       "      <td>CD45+</td>\n",
       "      <td>tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1564</th>\n",
       "      <td>P18_T1</td>\n",
       "      <td>ERS16703118</td>\n",
       "      <td>SAMEA114591092</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 18</td>\n",
       "      <td>TB21.0006</td>\n",
       "      <td>77</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P18_T1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-features.tsv.gz</td>\n",
       "      <td>lung squamous cell carcinoma</td>\n",
       "      <td>CD45+</td>\n",
       "      <td>tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1565</th>\n",
       "      <td>P18_T1</td>\n",
       "      <td>ERS16703118</td>\n",
       "      <td>SAMEA114591092</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 18</td>\n",
       "      <td>TB21.0006</td>\n",
       "      <td>77</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P18_T1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-features.tsv.gz</td>\n",
       "      <td>lung squamous cell carcinoma</td>\n",
       "      <td>CD45+</td>\n",
       "      <td>tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1566</th>\n",
       "      <td>P18_T1</td>\n",
       "      <td>ERS16703118</td>\n",
       "      <td>SAMEA114591092</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 18</td>\n",
       "      <td>TB21.0006</td>\n",
       "      <td>77</td>\n",
       "      <td>year</td>\n",
       "      <td>EFO</td>\n",
       "      <td>UO_0000036</td>\n",
       "      <td>...</td>\n",
       "      <td>P18_T1-matrix.mtx.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-barcodes.tsv.gz</td>\n",
       "      <td>P-MTAB-137970</td>\n",
       "      <td>P-MTAB-137971</td>\n",
       "      <td>P18_T1-features.tsv.gz</td>\n",
       "      <td>lung squamous cell carcinoma</td>\n",
       "      <td>CD45+</td>\n",
       "      <td>tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1567 rows × 69 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Source Name Comment[ENA_SAMPLE] Comment[BioSD_SAMPLE]  \\\n",
       "0          P1_B1         ERS16703054        SAMEA114591031   \n",
       "1          P1_B1         ERS16703054        SAMEA114591031   \n",
       "2          P1_B1         ERS16703054        SAMEA114591031   \n",
       "3          P1_B1         ERS16703054        SAMEA114591031   \n",
       "4          P1_B1         ERS16703054        SAMEA114591031   \n",
       "...          ...                 ...                   ...   \n",
       "1562      P18_T1         ERS16703118        SAMEA114591092   \n",
       "1563      P18_T1         ERS16703118        SAMEA114591092   \n",
       "1564      P18_T1         ERS16703118        SAMEA114591092   \n",
       "1565      P18_T1         ERS16703118        SAMEA114591092   \n",
       "1566      P18_T1         ERS16703118        SAMEA114591092   \n",
       "\n",
       "     Characteristics[organism] Characteristics[individual]  \\\n",
       "0                 Homo sapiens                   Patient 1   \n",
       "1                 Homo sapiens                   Patient 1   \n",
       "2                 Homo sapiens                   Patient 1   \n",
       "3                 Homo sapiens                   Patient 1   \n",
       "4                 Homo sapiens                   Patient 1   \n",
       "...                        ...                         ...   \n",
       "1562              Homo sapiens                  Patient 18   \n",
       "1563              Homo sapiens                  Patient 18   \n",
       "1564              Homo sapiens                  Patient 18   \n",
       "1565              Homo sapiens                  Patient 18   \n",
       "1566              Homo sapiens                  Patient 18   \n",
       "\n",
       "     Characteristics[original source name]  Characteristics[age]  \\\n",
       "0                                TB18.0655                    73   \n",
       "1                                TB18.0655                    73   \n",
       "2                                TB18.0655                    73   \n",
       "3                                TB18.0655                    73   \n",
       "4                                TB18.0655                    73   \n",
       "...                                    ...                   ...   \n",
       "1562                             TB21.0006                    77   \n",
       "1563                             TB21.0006                    77   \n",
       "1564                             TB21.0006                    77   \n",
       "1565                             TB21.0006                    77   \n",
       "1566                             TB21.0006                    77   \n",
       "\n",
       "     Unit[time unit] Term Source REF Term Accession Number  ...  \\\n",
       "0               year             EFO            UO_0000036  ...   \n",
       "1               year             EFO            UO_0000036  ...   \n",
       "2               year             EFO            UO_0000036  ...   \n",
       "3               year             EFO            UO_0000036  ...   \n",
       "4               year             EFO            UO_0000036  ...   \n",
       "...              ...             ...                   ...  ...   \n",
       "1562            year             EFO            UO_0000036  ...   \n",
       "1563            year             EFO            UO_0000036  ...   \n",
       "1564            year             EFO            UO_0000036  ...   \n",
       "1565            year             EFO            UO_0000036  ...   \n",
       "1566            year             EFO            UO_0000036  ...   \n",
       "\n",
       "     Derived Array Data File Protocol REF.6 Protocol REF.7  \\\n",
       "0        P1_B1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1        P1_B1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "2        P1_B1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "3        P1_B1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "4        P1_B1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "...                      ...            ...            ...   \n",
       "1562    P18_T1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1563    P18_T1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1564    P18_T1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1565    P18_T1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1566    P18_T1-matrix.mtx.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "\n",
       "     Derived Array Data File.1 Protocol REF.8 Protocol REF.9  \\\n",
       "0        P1_B1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1        P1_B1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "2        P1_B1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "3        P1_B1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "4        P1_B1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "...                        ...            ...            ...   \n",
       "1562    P18_T1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1563    P18_T1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1564    P18_T1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1565    P18_T1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "1566    P18_T1-barcodes.tsv.gz  P-MTAB-137970  P-MTAB-137971   \n",
       "\n",
       "     Derived Array Data File.2         Factor Value[disease]  \\\n",
       "0        P1_B1-features.tsv.gz    non-small cell lung cancer   \n",
       "1        P1_B1-features.tsv.gz    non-small cell lung cancer   \n",
       "2        P1_B1-features.tsv.gz    non-small cell lung cancer   \n",
       "3        P1_B1-features.tsv.gz    non-small cell lung cancer   \n",
       "4        P1_B1-features.tsv.gz    non-small cell lung cancer   \n",
       "...                        ...                           ...   \n",
       "1562    P18_T1-features.tsv.gz  lung squamous cell carcinoma   \n",
       "1563    P18_T1-features.tsv.gz  lung squamous cell carcinoma   \n",
       "1564    P18_T1-features.tsv.gz  lung squamous cell carcinoma   \n",
       "1565    P18_T1-features.tsv.gz  lung squamous cell carcinoma   \n",
       "1566    P18_T1-features.tsv.gz  lung squamous cell carcinoma   \n",
       "\n",
       "     Factor Value[FACS]      Factor Value[sampling site]  \n",
       "0                  MDSC  normal tissue adjacent to tumor  \n",
       "1                  MDSC  normal tissue adjacent to tumor  \n",
       "2                  MDSC  normal tissue adjacent to tumor  \n",
       "3                  MDSC  normal tissue adjacent to tumor  \n",
       "4                  MDSC  normal tissue adjacent to tumor  \n",
       "...                 ...                              ...  \n",
       "1562              CD45+                            tumor  \n",
       "1563              CD45+                            tumor  \n",
       "1564              CD45+                            tumor  \n",
       "1565              CD45+                            tumor  \n",
       "1566              CD45+                            tumor  \n",
       "\n",
       "[1567 rows x 69 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zuani_dir = f'{data_dir}/Zuani2024'\n",
    "zuani_samples = pd.read_csv(f'{zuani_dir}/E-MTAB-13526.sdrf.txt', sep='\\t')\n",
    "\n",
    "# Filter for non enriched MDSC = Myeloid derived supressor cells, CD235a- filters for erythrocites\n",
    "zuani_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "546d01dd-5a3c-4eb4-9a3a-e2f5447ce1e5",
   "metadata": {},
   "source": [
    "We filter for non-immune enriched samples and only tumor tissue. All the non-tumor enirhced ones are filtered for no plasma cells (erythrocites CD235a-)\n",
    "that is also the case in many of the other samples... we can consider the eryhrocites enriched depleted, for they are only one cell type that is not very present. But we should impute some abundance values in these samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4fb1b2dd-900d-4ca5-adf7-e0c6c8ba8d3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "facs_filter = ['CD235a-']\n",
    "zuani_samples = zuani_samples[zuani_samples['Factor Value[FACS]'].isin(facs_filter)]\n",
    "\n",
    "zuani_samples = zuani_samples[zuani_samples['Factor Value[sampling site]']=='tumor']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3b25edd8-8c34-42df-b8bd-9e0d9bf9695d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Source Name', 'Comment[ENA_SAMPLE]', 'Comment[BioSD_SAMPLE]',\n",
       "       'Characteristics[organism]', 'Characteristics[individual]',\n",
       "       'Characteristics[original source name]', 'Characteristics[age]',\n",
       "       'Unit[time unit]', 'Term Source REF', 'Term Accession Number',\n",
       "       'Characteristics[developmental stage]', 'Characteristics[sex]',\n",
       "       'Characteristics[organism part]', 'Characteristics[disease]',\n",
       "       'Characteristics[FACS]', 'Characteristics[sampling site]',\n",
       "       'Characteristics[tumor grading]', 'Description', 'Material Type',\n",
       "       'Protocol REF', 'Protocol REF.1', 'Protocol REF.2', 'Extract Name',\n",
       "       'Comment[LIBRARY_LAYOUT]', 'Comment[LIBRARY_SELECTION]',\n",
       "       'Comment[LIBRARY_SOURCE]', 'Comment[LIBRARY_STRATEGY]',\n",
       "       'Comment[cdna read]', 'Comment[cdna read offset]',\n",
       "       'Comment[cdna read size]', 'Comment[cell barcode offset]',\n",
       "       'Comment[cell barcode read]', 'Comment[cell barcode size]',\n",
       "       'Comment[end bias]', 'Comment[input molecule]',\n",
       "       'Comment[library construction]', 'Comment[primer]',\n",
       "       'Comment[sample barcode offset]', 'Comment[sample barcode read]',\n",
       "       'Comment[sample barcode size]', 'Comment[single cell isolation]',\n",
       "       'Comment[spike in]', 'Comment[umi barcode offset]',\n",
       "       'Comment[umi barcode read]', 'Comment[umi barcode size]',\n",
       "       'Protocol REF.3', 'Performer', 'Assay Name',\n",
       "       'Comment[technical replicate group]', 'Technology Type',\n",
       "       'Comment[ENA_EXPERIMENT]', 'Scan Name', 'Comment[SUBMITTED_FILE_NAME]',\n",
       "       'Comment[ENA_RUN]', 'Comment[FASTQ_URI]', 'Comment[READ_TYPE]',\n",
       "       'Comment[READ_INDEX]', 'Protocol REF.4', 'Protocol REF.5',\n",
       "       'Derived Array Data File', 'Protocol REF.6', 'Protocol REF.7',\n",
       "       'Derived Array Data File.1', 'Protocol REF.8', 'Protocol REF.9',\n",
       "       'Derived Array Data File.2', 'Factor Value[disease]',\n",
       "       'Factor Value[FACS]', 'Factor Value[sampling site]'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zuani_samples.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1240c3c0-076c-40ae-95a5-2a3538ad5c44",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['T3N0M0', 'T2bN0M0', nan, 'T2aN0M0', 'T4N2M0', 'T2aN1M0', 'T3M0N0',\n",
       "       'T2bN1M0', 'T1cN0M0', 'T4N0'], dtype=object)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zuani_samples['Characteristics[tumor grading]'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "6f188cc0-e8b4-433d-bc20-13054845bb2a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPUS:  8\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P1_T1-barcodes.tsv.gz\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P1_T1-features.tsv.gz\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P1_T1-matrix.mtx.gz\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P2_T1-barcodes.tsv.gz\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P2_T1-features.tsv.gz\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P2_T1-matrix.mtx.gz\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P15_T2-barcodes.tsv.gz\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P15_T2-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P2_T1-features.tsv.gz time (s): 1.5921189785003662\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P15_T2-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P1_T1-features.tsv.gz time (s): 1.5972809791564941\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P16_T2-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P15_T2-features.tsv.gz time (s): 1.3004364967346191\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P16_T2-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P1_T1-barcodes.tsv.gz time (s): 6.099670171737671\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P16_T2-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P16_T2-features.tsv.gz time (s): 1.1689391136169434\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T2-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P2_T1-matrix.mtx.gz time (s): 25.24450135231018\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T2-features.tsv.gzurl: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P2_T1-barcodes.tsv.gz time (s): 43.2615282535553\n",
      "\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T2-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P15_T2-barcodes.tsv.gz time (s): 43.600990295410156\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T3-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T2-features.tsv.gz time (s): 1.2455544471740723\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T3-features.tsv.gzurl: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P16_T2-barcodes.tsv.gz time (s): 42.82427263259888\n",
      "\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T3-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T3-features.tsv.gz time (s): 1.2398083209991455\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T2-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P1_T1-matrix.mtx.gz time (s): 52.44864463806152\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T2-features.tsv.gzurl: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T2-barcodes.tsv.gz time (s): 42.84986734390259\n",
      "\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T2-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T2-features.tsv.gz time (s): 1.2251718044281006\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T3-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T3-barcodes.tsv.gz time (s): 42.83008527755737\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T3-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T2-barcodes.tsv.gz time (s): 42.71306538581848\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T3-matrix.mtx.gzurl: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T3-features.tsv.gz time (s): 1.2472662925720215\n",
      "\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P20_T2-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T2-matrix.mtx.gz time (s): 76.8129460811615\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P20_T2-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T3-barcodes.tsv.gz time (s): 42.92999482154846\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P20_T2-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P20_T2-features.tsv.gz time (s): 1.2489855289459229\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P18_T2-barcodes.tsv.gzurl: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P4_T3-matrix.mtx.gz time (s): 95.85074806213379\n",
      "\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P18_T2-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P16_T2-matrix.mtx.gz time (s): 142.04853892326355\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P18_T2-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P18_T2-features.tsv.gz time (s): 1.438136339187622\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P19_T2-barcodes.tsv.gzurl: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P20_T2-barcodes.tsv.gz time (s): 42.95699501037598\n",
      "\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P19_T2-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P18_T2-barcodes.tsv.gz time (s): 42.92334604263306\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P19_T2-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P19_T2-features.tsv.gz time (s): 1.2489831447601318\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P24_T1-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P19_T2-barcodes.tsv.gz time (s): 42.852856397628784\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P24_T1-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P15_T2-matrix.mtx.gz time (s): 210.2394998073578\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P24_T1-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P24_T1-features.tsv.gz time (s): 1.2681214809417725\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P22_T1-barcodes.tsv.gzurl: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P24_T1-barcodes.tsv.gz time (s): 42.93040895462036\n",
      "\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P22_T1-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P22_T1-barcodes.tsv.gz time (s): 42.95659422874451\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P22_T1-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P22_T1-features.tsv.gz time (s): 1.2519481182098389\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P8_T2-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P22_T1-matrix.mtx.gz time (s): 83.76424765586853\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P8_T2-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P8_T2-barcodes.tsv.gz time (s): 43.164130449295044\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P8_T2-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P8_T2-features.tsv.gz time (s): 1.2520792484283447\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T1-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P20_T2-matrix.mtx.gz time (s): 340.315548658371\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T1-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P19_T2-matrix.mtx.gz time (s): 318.12532663345337\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T1-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T1-features.tsv.gz time (s): 1.2496387958526611\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P23_T1-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T1-barcodes.tsv.gz time (s): 43.02663064002991\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P23_T1-features.tsv.gzurl: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P18_T2-matrix.mtx.gz time (s): 405.18058800697327\n",
      "\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P23_T1-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P23_T1-features.tsv.gz time (s): 1.375727891921997\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T2-barcodes.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P23_T1-barcodes.tsv.gz time (s): 42.89959907531738\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T2-features.tsv.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T2-matrix.mtx.gz time (s): 505.055460691452\n",
      "getting https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T2-matrix.mtx.gz\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T2-features.tsv.gz time (s): 1.2500417232513428\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T2-matrix.mtx.gz time (s): 23.810145139694214\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T2-barcodes.tsv.gz time (s): 43.953601360321045\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P21_T1-matrix.mtx.gz time (s): 133.2835156917572\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P17_T3-matrix.mtx.gz time (s): 639.7812283039093\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P24_T1-matrix.mtx.gz time (s): 703.8568208217621\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P8_T2-matrix.mtx.gz time (s): 500.3829493522644\n",
      "url: https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files/P23_T1-matrix.mtx.gz time (s): 617.2215695381165\n"
     ]
    }
   ],
   "source": [
    "base_url = 'https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/526/E-MTAB-13526/Files'\n",
    "zuani_patients = zuani_samples['Source Name'].unique()\n",
    "\n",
    "urls_paths = []\n",
    "\n",
    "for pat in zuani_patients:\n",
    "    filename = f'{pat}-barcodes.tsv.gz'\n",
    "    urls_paths.append((f'{base_url}/{filename}',f'{zuani_dir}/{filename}'))\n",
    "\n",
    "    filename = f'{pat}-features.tsv.gz'\n",
    "    urls_paths.append((f'{base_url}/{filename}',f'{zuani_dir}/{filename}'))\n",
    "\n",
    "    filename = f'{pat}-matrix.mtx.gz'\n",
    "    urls_paths.append((f'{base_url}/{filename}',f'{zuani_dir}/{filename}'))\n",
    "\n",
    "ut.download_parallel(urls_paths, cpus=8)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c81a40fc-32fd-431e-bced-f64fb48c7561",
   "metadata": {},
   "source": [
    "### Remove empty drops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b8300901-e215-4229-9d48-815a53481a4a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /root/datos/maestria/netopaas/Zuani2024//EmptyDrop_P16_T2-.sh\n",
      "[1] \"/root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P16_T2-\"\n",
      "[1] \"Creating folder  /root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P16_T2-\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /root/datos/maestria/netopaas/Zuani2024//EmptyDrop_P17_T2-.sh\n",
      "[1] \"/root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P17_T2-\"\n",
      "[1] \"Creating folder  /root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P17_T2-\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /root/datos/maestria/netopaas/Zuani2024//EmptyDrop_P17_T3-.sh\n",
      "[1] \"/root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P17_T3-\"\n",
      "[1] \"Creating folder  /root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P17_T3-\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /root/datos/maestria/netopaas/Zuani2024//EmptyDrop_P18_T2-.sh\n",
      "[1] \"/root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P18_T2-\"\n",
      "[1] \"Creating folder  /root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P18_T2-\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /root/datos/maestria/netopaas/Zuani2024//EmptyDrop_P19_T2-.sh\n",
      "[1] \"/root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P19_T2-\"\n",
      "[1] \"Creating folder  /root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P19_T2-\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /root/datos/maestria/netopaas/Zuani2024//EmptyDrop_P20_T2-.sh\n",
      "[1] \"/root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P20_T2-\"\n",
      "[1] \"Creating folder  /root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P20_T2-\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /root/datos/maestria/netopaas/Zuani2024//EmptyDrop_P21_T1-.sh\n",
      "[1] \"/root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P21_T1-\"\n",
      "[1] \"Creating folder  /root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P21_T1-\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /root/datos/maestria/netopaas/Zuani2024//EmptyDrop_P21_T2-.sh\n",
      "[1] \"/root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P21_T2-\"\n",
      "[1] \"Creating folder  /root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P21_T2-\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /root/datos/maestria/netopaas/Zuani2024//EmptyDrop_P22_T1-.sh\n",
      "[1] \"/root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P22_T1-\"\n",
      "[1] \"Creating folder  /root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P22_T1-\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /root/datos/maestria/netopaas/Zuani2024//EmptyDrop_P23_T1-.sh\n",
      "[1] \"/root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P23_T1-\"\n",
      "[1] \"Creating folder  /root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P23_T1-\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /root/datos/maestria/netopaas/Zuani2024//EmptyDrop_P24_T1-.sh\n",
      "[1] \"/root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P24_T1-\"\n",
      "[1] \"Creating folder  /root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P24_T1-\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /root/datos/maestria/netopaas/Zuani2024//EmptyDrop_P4_T2-.sh\n",
      "[1] \"/root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P4_T2-\"\n",
      "[1] \"Creating folder  /root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P4_T2-\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /root/datos/maestria/netopaas/Zuani2024//EmptyDrop_P4_T3-.sh\n",
      "[1] \"/root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P4_T3-\"\n",
      "[1] \"Creating folder  /root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P4_T3-\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /root/datos/maestria/netopaas/Zuani2024//EmptyDrop_P8_T2-.sh\n",
      "[1] \"/root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P8_T2-\"\n",
      "[1] \"Creating folder  /root/datos/maestria/netopaas/Zuani2024//outputEmptyDrops/P8_T2-\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from os import system as sys\n",
    "import os\n",
    "\n",
    "codeDir = '/root/host_home/luca/utils/' # CHANGE ME TO THE DIRECTORY WHERE run_EmptyDrop.R IS CLONED\n",
    "directory = '/root/datos/maestria/netopaas/Zuani2024/' # change me to the dir where the matrix files are stored\n",
    "\n",
    "# All CellRanger samples for the lung data\n",
    "samples = []\n",
    "\n",
    "file_pattern = sorted([f for f in os.listdir(directory) if 'matrix' in f])\n",
    "for mtx_file in file_pattern:\n",
    "    idx = mtx_file.replace('matrix.mtx.gz', '')\n",
    "    samples.append(idx)\n",
    "\n",
    "for sample in samples:\n",
    "    fname = directory+\"/EmptyDrop_\"+sample+\".sh\"\n",
    "    f = open(f'{fname}', 'w')\n",
    "    f.write(\"cd /root/datos/maestria/netopaas/Zuani2024/\\n\") # CHANGE ME TO A PLACE YOU WOULD LIKE TO WRITE OUTPUTS TO\n",
    "    # f.write(\"conda activate minimal_env\\n\" )\n",
    "    f.write(\"Rscript \" + codeDir + \"run_EmptyDrop.R {0}\\n\".format(str(sample)))\n",
    "    f.close()\n",
    "\n",
    "    # Example batch calling on the codon LSF cluster (this will vary depending on the system, bsub vs. qsub etc)\n",
    "    print(f'bash {fname}')\n",
    "    sys(f'bash {fname}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c69ad9fb-439e-4150-87ee-5e2dfd53bfa9",
   "metadata": {},
   "source": [
    "### Convert to adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "2c1b051d-478a-4a99-8c16-71d810eaedd9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.11/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 318313 × 33538\n",
      "    obs: 'sample'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:1897: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import io\n",
    "import csv\n",
    "import gzip\n",
    "\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "\n",
    "def concatenate_datasets(directory):\n",
    "    \"\"\"\n",
    "    Concatenate multiple .mtx datasets from a single directory into a single AnnData object.\n",
    "    \n",
    "    Parameters:\n",
    "    - directory (str): The directory containing .mtx, features.tsv, and barcodes.tsv files.\n",
    "    \n",
    "    Returns:\n",
    "    - AnnData: The concatenated AnnData object.\n",
    "    \"\"\"\n",
    "    adatas = {}\n",
    "    features_file = ''\n",
    "    \n",
    "    # List all subdirectories in the main directory\n",
    "    subdirectories = [d for d in os.listdir(directory) if os.path.isdir(os.path.join(directory, d))]\n",
    "    \n",
    "    for subdirectory in subdirectories:\n",
    "        # Construct the path to the subdirectory\n",
    "        subdirectory_path = os.path.join(directory, subdirectory)\n",
    "        \n",
    "        # Read the .mtx file and associated files from the subdirectory\n",
    "        adata = sc.read_10x_mtx(subdirectory_path, var_names='gene_symbols', cache=True)\n",
    "        \n",
    "        # Use the subdirectory name as the identifier\n",
    "        idx = subdirectory\n",
    "        \n",
    "        # Store the AnnData object in the dictionary\n",
    "        adatas[idx] = adata\n",
    "    \n",
    "    # Concatenate all AnnData objects, labeling them by the subdirectory name which is used as the sample identifier\n",
    "    concatenated_adata = ad.concat(adatas, label='sample')\n",
    "\n",
    "    return concatenated_adata\n",
    "\n",
    "# Example usage:\n",
    "adata = concatenate_datasets(f'{zuani_dir}/outputEmptyDrops')\n",
    "print(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "dd8a6b40-c97d-4f8e-a844-4bae0cabf068",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write_h5ad(f'{zuani_dir}/Zuani2024.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "c2be4836-f9cd-4974-8062-3ece65db64d5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:1897: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    }
   ],
   "source": [
    "adata = sc.read_h5ad('/root/datos/maestria/netopaas/Zuani2024/Zuani.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8658d30-c55c-4dfe-9393-309590c3a618",
   "metadata": {},
   "source": [
    "Before emptydrops with FDR 0.001"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "68d711ff-e6ee-4694-b1f0-6dd6e821019c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 101923200 × 33538\n",
       "    obs: 'sample'\n",
       "    var: 'gene_symbols'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad9ea544-821f-4b5c-9307-47900d3d8c87",
   "metadata": {},
   "source": [
    "adter EmptyDrops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "c2af13e1-e9be-462e-bfbd-1a0d9de00133",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 312502 × 33538\n",
       "    obs: 'sample', 'barcode', 'Source Name', 'Characteristics[tumor grading]'"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "be47ea48-abc2-4c6c-a3a9-89a859d9df63",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:850: UserWarning: \n",
      "AnnData expects .obs.index to contain strings, but got values like:\n",
      "    [0, 1, 2, 3, 4]\n",
      "\n",
      "    Inferred to be: integer\n",
      "\n",
      "  value_idx = self._prep_dim_index(value.index, attr)\n"
     ]
    }
   ],
   "source": [
    "adata.obs['sample'] = adata.obs['sample'].str[:-1]\n",
    "zuani_samples_unique = zuani_samples.drop_duplicates(subset=['Source Name'], keep='first')\n",
    "adata.obs = pd.merge(adata.obs, zuani_samples_unique.loc[:, ['Source Name', 'Characteristics[tumor grading]']],\n",
    "                            left_on='sample', right_on='Source Name', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "7bb01c65-bb61-4320-8f17-763ad9a254fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "def map_tnm_to_roman(tnm_stage):\n",
    "    stage_mappings = {\n",
    "        r\"^Tis N0 M0$\": \"Stage 0\",\n",
    "        r\"^T1[ab]? N0 M0$\": \"Stage I\",\n",
    "        r\"^T2[ab]? N0 M0$\": \"Stage II\",\n",
    "        r\"^T2[ab]? N0/1$\": \"Stage II\",\n",
    "        r\"^T1[ab]? N1 M0$\": \"Stage II\",\n",
    "        r\"^T2[ab]? N1 M0$\": \"Stage II\",\n",
    "        r\"^T3 N[01] M0$\": \"Stage II\",\n",
    "        r\"^T1[ab]? N2 M0$\": \"Stage III\",\n",
    "        r\"^T2[ab]? N2 M0$\": \"Stage III\",\n",
    "        r\"^T3 N2 M0$\": \"Stage III\",\n",
    "        r\"^T4 N[012] M0$\": \"Stage III\",\n",
    "        r\"^T4 N0$\": \"Stage III\",\n",
    "        r\"^T[1-4][ab]? N3 M0$\": \"Stage III\",\n",
    "        r\"^T[1-4][ab]? N[0-3] M1[abc]$\": \"Stage IV\",\n",
    "    }\n",
    "    \n",
    "    for pattern, stage in stage_mappings.items():\n",
    "        if re.match(pattern, tnm_stage):\n",
    "            return stage\n",
    "    return \"Unknown stage\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c604c393-8970-4674-984e-fd5ef239fb86",
   "metadata": {},
   "source": [
    "Remove the patients with no tumor grading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "5a37a31f-f664-464d-aaa1-2d3635a8f260",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[~adata.obs['Characteristics[tumor grading]'].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72831dc7-6b78-4ce8-ba1b-8b5a3b94225e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "adata.obs['Characteristics[tumor grading]'].apply(map_tnm_to_roman)\n",
    "# adata.obs.stage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "7c619493-37e6-410f-8d1a-f2a9b89c7ef5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:1292: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:1897: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:1897: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:1292: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:1897: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:1897: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:1292: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:1897: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:1897: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:1292: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:1897: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/usr/local/lib/python3.11/site-packages/anndata/_core/anndata.py:1897: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    }
   ],
   "source": [
    "adata.write_h5ad(f'{zuani_dir}/Zuani2024.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8999abd6-a111-4106-be8f-7ab6d1a81968",
   "metadata": {},
   "source": [
    "## Deng 2024"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "93f4066a-b712-4504-afad-319d866e1e65",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'urllib' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43murllib\u001b[49m\u001b[38;5;241m.\u001b[39mrequest\u001b[38;5;241m.\u001b[39murlretrieve(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mhttps://figshare.com/ndownloader/files/44695465\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mdata_dir\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m/Deng2024/dengluad.rds\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'urllib' is not defined"
     ]
    }
   ],
   "source": [
    "# urllib.request.urlretrieve('https://figshare.com/ndownloader/files/44695465',f'{data_dir}/Deng2024/dengluad.rds')\n",
    "urllib.request.urlretrieve('https://figshare.com/ndownloader/files/46617667',f'{data_dir}/Deng2024/dengEcoTyper.rds')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fec32e4c-1b7a-4076-82db-382a7641c0b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "! rm /root/datos/maestria/netopaas/Deng2024/dengRNA.h5ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "05a71c46-c164-4201-b5e2-1f1c16c96e1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    },
    {
     "data": {
      "text/plain": [
       "Registered S3 method overwritten by 'SeuratDisk':\n",
       "  method            from  \n",
       "  as.sparse.H5Group Seurat\n",
       "Validating h5Seurat file\n",
       "Adding data from RNA as X\n",
       "Adding counts from RNA as raw\n",
       "Transfering meta.data to obs\n",
       "Adding dimensional reduction information for tsne (global)\n",
       "Adding dimensional reduction information for umap (global)\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i data_dir\n",
    "library(SeuratDisk)\n",
    "Convert(paste0(data_dir, '/Deng2024/deng.h5Seurat'), dest = paste0(data_dir, '/Deng2024/dengRNA.h5ad'), assay = 'RNA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2710b1b5-0cf7-4701-99f1-366384940f58",
   "metadata": {
    "editable": true,
    "scrolled": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    Loading required package: SeuratObject\n",
      "Loading required package: sp\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "Registered S3 method overwritten by 'SeuratDisk':\n",
      "  method            from  \n",
      "  as.sparse.H5Group Seurat\n",
      "Validating object structure\n",
      "Updating object slots\n",
      "Ensuring keys are in the proper structure\n",
      "Updating matrix keys for DimReduc ‘pca’\n",
      "Updating matrix keys for DimReduc ‘umap’\n",
      "Updating matrix keys for DimReduc ‘tsne’\n",
      "Warning: Assay RNA changing from Assay to Assay\n",
      "Warning: Assay integrated changing from Assay to Assay\n",
      "Warning: Graph integrated_nn changing from Graph to Graph\n",
      "Warning: Graph integrated_snn changing from Graph to Graph\n",
      "Warning: DimReduc pca changing from DimReduc to DimReduc\n",
      "Warning: DimReduc umap changing from DimReduc to DimReduc\n",
      "Warning: DimReduc tsne changing from DimReduc to DimReduc\n",
      "Ensuring keys are in the proper structure\n",
      "Ensuring feature names don't have underscores or pipes\n",
      "Updating slots in RNA\n",
      "Updating slots in integrated\n",
      "Updating slots in integrated_nn\n",
      "Setting default assay of integrated_nn to integrated\n",
      "Updating slots in integrated_snn\n",
      "Setting default assay of integrated_snn to integrated\n",
      "Updating slots in pca\n",
      "Updating slots in umap\n",
      "Setting umap DimReduc to global\n",
      "Updating slots in tsne\n",
      "Setting tsne DimReduc to global\n",
      "No assay information could be found for FindIntegrationAnchors\n",
      "No assay information could be found for IntegrateData\n",
      "Setting assay used for ScaleData.integrated to integrated\n",
      "Setting assay used for RunPCA.integrated to integrated\n",
      "Setting assay used for RunUMAP.integrated.pca to integrated\n",
      "No assay information could be found for RunTSNE\n",
      "Setting assay used for FindNeighbors.integrated.pca to integrated\n",
      "No assay information could be found for FindClusters\n",
      "Validating object structure for Assay ‘RNA’\n",
      "Validating object structure for Assay ‘integrated’\n",
      "Validating object structure for Graph ‘integrated_nn’\n",
      "Validating object structure for Graph ‘integrated_snn’\n",
      "Validating object structure for DimReduc ‘pca’\n",
      "Validating object structure for DimReduc ‘umap’\n",
      "Validating object structure for DimReduc ‘tsne’\n",
      "Object representation is consistent with the most current Seurat version\n",
      "Creating h5Seurat file for version 3.1.5.9900\n",
      "Adding counts for RNA\n",
      "Adding data for RNA\n",
      "No variable features found for RNA\n",
      "No feature-level metadata found for RNA\n",
      "Adding data for integrated\n",
      "Adding scale.data for integrated\n",
      "Adding variable features for integrated\n",
      "No feature-level metadata found for integrated\n",
      "Adding cell embeddings for pca\n",
      "Adding loadings for pca\n",
      "No projected loadings for pca\n",
      "Adding standard deviations for pca\n",
      "No JackStraw data for pca\n",
      "Adding cell embeddings for umap\n",
      "No loadings for umap\n",
      "No projected loadings for umap\n",
      "No standard deviations for umap\n",
      "No JackStraw data for umap\n",
      "Adding cell embeddings for tsne\n",
      "No loadings for tsne\n",
      "No projected loadings for tsne\n",
      "No standard deviations for tsne\n",
      "No JackStraw data for tsne\n",
      "Validating h5Seurat file\n",
      "Error: Destination H5AD file exists\n",
      "In addition: Warning messages:\n",
      "1: Adding a command log without an assay associated with it \n",
      "2: Adding a command log without an assay associated with it \n",
      "3: Adding a command log without an assay associated with it \n",
      "4: Adding a command log without an assay associated with it\n",
      "Error: Destination H5AD file exists\n"
     ]
    },
    {
     "ename": "RInterpreterError",
     "evalue": "Failed to parse and evaluate line 'library(Seurat)\\nlibrary(SeuratDisk)\\n\\ndeng <- readRDS(paste0(data_dir, \\'/Deng2024/dengEcoTyper.rds\\'))\\ndeng = UpdateSeuratObject(deng)\\n\\n# After P49 there is no metadata so better delete. Also the ones with N as they are healthy controls,\\ncells_to_exclude <- which(grepl(\"P05|N\", deng@meta.data$orig.ident))\\ndeng <- subset(deng, cells = setdiff(Cells(deng), cells_to_exclude))\\n\\n\\nSaveH5Seurat(deng, filename=paste0(data_dir, \\'/Deng2024/deng.h5Seurat\\'))\\n# Convert(paste0(data_dir, \\'/Deng2024/deng.h5Seurat\\'), dest = \"h5ad\")\\nConvert(paste0(data_dir, \\'/Deng2024/deng.h5Seurat\\'), dest = paste0(data_dir, \\'/Deng2024/dengRNA.h5ad\\'), assay = \\'RNA\\')\\n'.\nR error message: 'Error: Destination H5AD file exists'\nR stdout:\nLoading required package: SeuratObject\nLoading required package: sp\n\nAttaching package: ‘SeuratObject’\n\nThe following object is masked from ‘package:base’:\n\n    intersect\n\nRegistered S3 method overwritten by 'SeuratDisk':\n  method            from  \n  as.sparse.H5Group Seurat\nValidating object structure\nUpdating object slots\nEnsuring keys are in the proper structure\nUpdating matrix keys for DimReduc ‘pca’\nUpdating matrix keys for DimReduc ‘umap’\nUpdating matrix keys for DimReduc ‘tsne’\nWarning: Assay RNA changing from Assay to Assay\nWarning: Assay integrated changing from Assay to Assay\nWarning: Graph integrated_nn changing from Graph to Graph\nWarning: Graph integrated_snn changing from Graph to Graph\nWarning: DimReduc pca changing from DimReduc to DimReduc\nWarning: DimReduc umap changing from DimReduc to DimReduc\nWarning: DimReduc tsne changing from DimReduc to DimReduc\nEnsuring keys are in the proper structure\nEnsuring feature names don't have underscores or pipes\nUpdating slots in RNA\nUpdating slots in integrated\nUpdating slots in integrated_nn\nSetting default assay of integrated_nn to integrated\nUpdating slots in integrated_snn\nSetting default assay of integrated_snn to integrated\nUpdating slots in pca\nUpdating slots in umap\nSetting umap DimReduc to global\nUpdating slots in tsne\nSetting tsne DimReduc to global\nNo assay information could be found for FindIntegrationAnchors\nNo assay information could be found for IntegrateData\nSetting assay used for ScaleData.integrated to integrated\nSetting assay used for RunPCA.integrated to integrated\nSetting assay used for RunUMAP.integrated.pca to integrated\nNo assay information could be found for RunTSNE\nSetting assay used for FindNeighbors.integrated.pca to integrated\nNo assay information could be found for FindClusters\nValidating object structure for Assay ‘RNA’\nValidating object structure for Assay ‘integrated’\nValidating object structure for Graph ‘integrated_nn’\nValidating object structure for Graph ‘integrated_snn’\nValidating object structure for DimReduc ‘pca’\nValidating object structure for DimReduc ‘umap’\nValidating object structure for DimReduc ‘tsne’\nObject representation is consistent with the most current Seurat version\nCreating h5Seurat file for version 3.1.5.9900\nAdding counts for RNA\nAdding data for RNA\nNo variable features found for RNA\nNo feature-level metadata found for RNA\nAdding data for integrated\nAdding scale.data for integrated\nAdding variable features for integrated\nNo feature-level metadata found for integrated\nAdding cell embeddings for pca\nAdding loadings for pca\nNo projected loadings for pca\nAdding standard deviations for pca\nNo JackStraw data for pca\nAdding cell embeddings for umap\nNo loadings for umap\nNo projected loadings for umap\nNo standard deviations for umap\nNo JackStraw data for umap\nAdding cell embeddings for tsne\nNo loadings for tsne\nNo projected loadings for tsne\nNo standard deviations for tsne\nNo JackStraw data for tsne\nValidating h5Seurat file\nError: Destination H5AD file exists\nIn addition: Warning messages:\n1: Adding a command log without an assay associated with it \n2: Adding a command log without an assay associated with it \n3: Adding a command log without an assay associated with it \n4: Adding a command log without an assay associated with it",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRRuntimeError\u001b[0m                             Traceback (most recent call last)",
      "File \u001b[0;32m/usr/local/lib/python3.11/site-packages/rpy2/ipython/rmagic.py:401\u001b[0m, in \u001b[0;36mRMagics.eval\u001b[0;34m(self, code)\u001b[0m\n\u001b[1;32m    400\u001b[0m     r_expr \u001b[38;5;241m=\u001b[39m ri\u001b[38;5;241m.\u001b[39mparse(code)\n\u001b[0;32m--> 401\u001b[0m     value, visible \u001b[38;5;241m=\u001b[39m \u001b[43mri\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mevalr_expr_with_visible\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    402\u001b[0m \u001b[43m        \u001b[49m\u001b[43mr_expr\u001b[49m\n\u001b[1;32m    403\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    404\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (ri\u001b[38;5;241m.\u001b[39membedded\u001b[38;5;241m.\u001b[39mRRuntimeError, \u001b[38;5;167;01mValueError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m exception:\n\u001b[1;32m    405\u001b[0m     \u001b[38;5;66;03m# Otherwise next return seems to have copy of error.\u001b[39;00m\n",
      "File \u001b[0;32m/usr/local/lib/python3.11/site-packages/rpy2/rinterface.py:196\u001b[0m, in \u001b[0;36mevalr_expr_with_visible\u001b[0;34m(expr, envir)\u001b[0m\n\u001b[1;32m    195\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m error_occured[\u001b[38;5;241m0\u001b[39m]:\n\u001b[0;32m--> 196\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m embedded\u001b[38;5;241m.\u001b[39mRRuntimeError(_rinterface\u001b[38;5;241m.\u001b[39m_geterrmessage())\n\u001b[1;32m    197\u001b[0m res \u001b[38;5;241m=\u001b[39m conversion\u001b[38;5;241m.\u001b[39m_cdata_to_rinterface(r_res)\n",
      "\u001b[0;31mRRuntimeError\u001b[0m: Error: Destination H5AD file exists\n",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mRInterpreterError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[3], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mget_ipython\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_cell_magic\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mR\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m-i data_dir\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mlibrary(Seurat)\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mlibrary(SeuratDisk)\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mdeng <- readRDS(paste0(data_dir, \u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m/Deng2024/dengEcoTyper.rds\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m))\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mdeng = UpdateSeuratObject(deng)\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m# After P49 there is no metadata so better delete. Also the ones with N as they are healthy controls,\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mcells_to_exclude <- which(grepl(\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mP05|N\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m, deng@meta.data$orig.ident))\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mdeng <- subset(deng, cells = setdiff(Cells(deng), cells_to_exclude))\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mSaveH5Seurat(deng, filename=paste0(data_dir, \u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m/Deng2024/deng.h5Seurat\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m))\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m# Convert(paste0(data_dir, \u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m/Deng2024/deng.h5Seurat\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m), dest = \u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mh5ad\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m)\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mConvert(paste0(data_dir, \u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m/Deng2024/deng.h5Seurat\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m), dest = paste0(data_dir, \u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m/Deng2024/dengRNA.h5ad\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m), assay = \u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43mRNA\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m)\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/local/lib/python3.11/site-packages/IPython/core/interactiveshell.py:2493\u001b[0m, in \u001b[0;36mInteractiveShell.run_cell_magic\u001b[0;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[1;32m   2491\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbuiltin_trap:\n\u001b[1;32m   2492\u001b[0m     args \u001b[38;5;241m=\u001b[39m (magic_arg_s, cell)\n\u001b[0;32m-> 2493\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mfn\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2495\u001b[0m \u001b[38;5;66;03m# The code below prevents the output from being displayed\u001b[39;00m\n\u001b[1;32m   2496\u001b[0m \u001b[38;5;66;03m# when using magics with decorator @output_can_be_silenced\u001b[39;00m\n\u001b[1;32m   2497\u001b[0m \u001b[38;5;66;03m# when the last Python token in the expression is a ';'.\u001b[39;00m\n\u001b[1;32m   2498\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(fn, magic\u001b[38;5;241m.\u001b[39mMAGIC_OUTPUT_CAN_BE_SILENCED, \u001b[38;5;28;01mFalse\u001b[39;00m):\n",
      "File \u001b[0;32m/usr/local/lib/python3.11/site-packages/rpy2/ipython/rmagic.py:978\u001b[0m, in \u001b[0;36mRMagics.R\u001b[0;34m(self, line, cell, local_ns)\u001b[0m\n\u001b[1;32m    976\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m e\u001b[38;5;241m.\u001b[39mstdout\u001b[38;5;241m.\u001b[39mendswith(e\u001b[38;5;241m.\u001b[39merr):\n\u001b[1;32m    977\u001b[0m         \u001b[38;5;28mprint\u001b[39m(e\u001b[38;5;241m.\u001b[39merr)\n\u001b[0;32m--> 978\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m e\n\u001b[1;32m    979\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m    980\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdevice \u001b[38;5;129;01min\u001b[39;00m DEVICES_STATIC:\n",
      "File \u001b[0;32m/usr/local/lib/python3.11/site-packages/rpy2/ipython/rmagic.py:943\u001b[0m, in \u001b[0;36mRMagics.R\u001b[0;34m(self, line, cell, local_ns)\u001b[0m\n\u001b[1;32m    941\u001b[0m         return_output \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[1;32m    942\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 943\u001b[0m     text_result, result, visible \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43meval\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcode\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    944\u001b[0m     text_output \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m text_result\n\u001b[1;32m    945\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m visible:\n",
      "File \u001b[0;32m/usr/local/lib/python3.11/site-packages/rpy2/ipython/rmagic.py:407\u001b[0m, in \u001b[0;36mRMagics.eval\u001b[0;34m(self, code)\u001b[0m\n\u001b[1;32m    404\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (ri\u001b[38;5;241m.\u001b[39membedded\u001b[38;5;241m.\u001b[39mRRuntimeError, \u001b[38;5;167;01mValueError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m exception:\n\u001b[1;32m    405\u001b[0m     \u001b[38;5;66;03m# Otherwise next return seems to have copy of error.\u001b[39;00m\n\u001b[1;32m    406\u001b[0m     warning_or_other_msg \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mflush()\n\u001b[0;32m--> 407\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m RInterpreterError(code, \u001b[38;5;28mstr\u001b[39m(exception),\n\u001b[1;32m    408\u001b[0m                             warning_or_other_msg)\n\u001b[1;32m    409\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m    410\u001b[0m     ro\u001b[38;5;241m.\u001b[39m_print_deferred_warnings()\n",
      "\u001b[0;31mRInterpreterError\u001b[0m: Failed to parse and evaluate line 'library(Seurat)\\nlibrary(SeuratDisk)\\n\\ndeng <- readRDS(paste0(data_dir, \\'/Deng2024/dengEcoTyper.rds\\'))\\ndeng = UpdateSeuratObject(deng)\\n\\n# After P49 there is no metadata so better delete. Also the ones with N as they are healthy controls,\\ncells_to_exclude <- which(grepl(\"P05|N\", deng@meta.data$orig.ident))\\ndeng <- subset(deng, cells = setdiff(Cells(deng), cells_to_exclude))\\n\\n\\nSaveH5Seurat(deng, filename=paste0(data_dir, \\'/Deng2024/deng.h5Seurat\\'))\\n# Convert(paste0(data_dir, \\'/Deng2024/deng.h5Seurat\\'), dest = \"h5ad\")\\nConvert(paste0(data_dir, \\'/Deng2024/deng.h5Seurat\\'), dest = paste0(data_dir, \\'/Deng2024/dengRNA.h5ad\\'), assay = \\'RNA\\')\\n'.\nR error message: 'Error: Destination H5AD file exists'\nR stdout:\nLoading required package: SeuratObject\nLoading required package: sp\n\nAttaching package: ‘SeuratObject’\n\nThe following object is masked from ‘package:base’:\n\n    intersect\n\nRegistered S3 method overwritten by 'SeuratDisk':\n  method            from  \n  as.sparse.H5Group Seurat\nValidating object structure\nUpdating object slots\nEnsuring keys are in the proper structure\nUpdating matrix keys for DimReduc ‘pca’\nUpdating matrix keys for DimReduc ‘umap’\nUpdating matrix keys for DimReduc ‘tsne’\nWarning: Assay RNA changing from Assay to Assay\nWarning: Assay integrated changing from Assay to Assay\nWarning: Graph integrated_nn changing from Graph to Graph\nWarning: Graph integrated_snn changing from Graph to Graph\nWarning: DimReduc pca changing from DimReduc to DimReduc\nWarning: DimReduc umap changing from DimReduc to DimReduc\nWarning: DimReduc tsne changing from DimReduc to DimReduc\nEnsuring keys are in the proper structure\nEnsuring feature names don't have underscores or pipes\nUpdating slots in RNA\nUpdating slots in integrated\nUpdating slots in integrated_nn\nSetting default assay of integrated_nn to integrated\nUpdating slots in integrated_snn\nSetting default assay of integrated_snn to integrated\nUpdating slots in pca\nUpdating slots in umap\nSetting umap DimReduc to global\nUpdating slots in tsne\nSetting tsne DimReduc to global\nNo assay information could be found for FindIntegrationAnchors\nNo assay information could be found for IntegrateData\nSetting assay used for ScaleData.integrated to integrated\nSetting assay used for RunPCA.integrated to integrated\nSetting assay used for RunUMAP.integrated.pca to integrated\nNo assay information could be found for RunTSNE\nSetting assay used for FindNeighbors.integrated.pca to integrated\nNo assay information could be found for FindClusters\nValidating object structure for Assay ‘RNA’\nValidating object structure for Assay ‘integrated’\nValidating object structure for Graph ‘integrated_nn’\nValidating object structure for Graph ‘integrated_snn’\nValidating object structure for DimReduc ‘pca’\nValidating object structure for DimReduc ‘umap’\nValidating object structure for DimReduc ‘tsne’\nObject representation is consistent with the most current Seurat version\nCreating h5Seurat file for version 3.1.5.9900\nAdding counts for RNA\nAdding data for RNA\nNo variable features found for RNA\nNo feature-level metadata found for RNA\nAdding data for integrated\nAdding scale.data for integrated\nAdding variable features for integrated\nNo feature-level metadata found for integrated\nAdding cell embeddings for pca\nAdding loadings for pca\nNo projected loadings for pca\nAdding standard deviations for pca\nNo JackStraw data for pca\nAdding cell embeddings for umap\nNo loadings for umap\nNo projected loadings for umap\nNo standard deviations for umap\nNo JackStraw data for umap\nAdding cell embeddings for tsne\nNo loadings for tsne\nNo projected loadings for tsne\nNo standard deviations for tsne\nNo JackStraw data for tsne\nValidating h5Seurat file\nError: Destination H5AD file exists\nIn addition: Warning messages:\n1: Adding a command log without an assay associated with it \n2: Adding a command log without an assay associated with it \n3: Adding a command log without an assay associated with it \n4: Adding a command log without an assay associated with it"
     ]
    }
   ],
   "source": [
    "%%R -i data_dir\n",
    "library(Seurat)\n",
    "library(SeuratDisk)\n",
    "\n",
    "deng <- readRDS(paste0(data_dir, '/Deng2024/dengEcoTyper.rds'))\n",
    "deng = UpdateSeuratObject(deng)\n",
    "\n",
    "# After P49 there is no metadata so better delete. Also the ones with N as they are healthy controls,\n",
    "cells_to_exclude <- which(grepl(\"P05|N\", deng@meta.data$orig.ident))\n",
    "deng <- subset(deng, cells = setdiff(Cells(deng), cells_to_exclude))\n",
    "\n",
    "\n",
    "SaveH5Seurat(deng, filename=paste0(data_dir, '/Deng2024/deng.h5Seurat'))\n",
    "# Convert(paste0(data_dir, '/Deng2024/deng.h5Seurat'), dest = \"h5ad\")\n",
    "Convert(paste0(data_dir, '/Deng2024/deng.h5Seurat'), dest = paste0(data_dir, '/Deng2024/dengRNA.h5ad'), assay = 'RNA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6ff2ee1a-a284-48bd-8238-0a0c899b3542",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'adata_integ' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[14], line 13\u001b[0m\n\u001b[1;32m     10\u001b[0m adata\u001b[38;5;241m.\u001b[39mlayers[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdata\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m adata\u001b[38;5;241m.\u001b[39mX\u001b[38;5;241m.\u001b[39mcopy()\n\u001b[1;32m     11\u001b[0m adata\u001b[38;5;241m.\u001b[39mX \u001b[38;5;241m=\u001b[39m adata\u001b[38;5;241m.\u001b[39mraw\u001b[38;5;241m.\u001b[39mX\n\u001b[0;32m---> 13\u001b[0m adata\u001b[38;5;241m.\u001b[39mraw \u001b[38;5;241m=\u001b[39m \u001b[43madata_integ\u001b[49m\n\u001b[1;32m     15\u001b[0m path_metadata \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mdata_dir\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m/Deng2024/metadata.xlsx\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m     16\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mexists(path_metadata):\n",
      "\u001b[0;31mNameError\u001b[0m: name 'adata_integ' is not defined"
     ]
    }
   ],
   "source": [
    "# adata_integ = sc.read_h5ad(f'{data_dir}/Deng2024/deng.h5ad')\n",
    "adata = sc.read_h5ad(f'{data_dir}/Deng2024/dengRNA.h5ad')\n",
    "\n",
    "# After P49 there is no metadata so better delete. Also the ones with N as they are healthy controls,\n",
    "# Should be done before in R to be faster\n",
    "# adata = adata[~adata.obs['orig.ident'].str.contains('P05|N')]\n",
    "\n",
    "# The assignment fo slots and assays is not very good with Convert from SeuratDisk so we make some arrangements\n",
    "# We also arrange the X so that it is like the other filtered datasets in the repo\n",
    "adata.layers['data'] = adata.X.copy()\n",
    "adata.X = adata.raw.X\n",
    "\n",
    "adata.raw = adata_integ\n",
    "\n",
    "path_metadata = f'{data_dir}/Deng2024/metadata.xlsx'\n",
    "if not os.path.exists(path_metadata):\n",
    "    urllib.request.urlretrieve('https://www.ncbi.nlm.nih.gov/pmc/articles/PMC11031428/bin/mmc2.xlsx', path_metadata)\n",
    "xl_zuani = pd.ExcelFile(path_metadata)\n",
    "metada_zuani = xl_zuani.parse(' lung cancer in scRNAseq')\n",
    "adata.obs = pd.merge(adata.obs, metada_zuani, how='left', left_on='orig.ident', right_on='Patient Number')\n",
    "\n",
    "\n",
    "adata.write_h5ad(f'{backup_dir}/filtered_Deng_Liu_LUAD_2024.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "1e1dcce6-f282-4926-8d4c-c8747e7bff66",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['unassigned', 'CD8.T.cells_S04', 'CD4.T.cells_S04',\n",
       "       'CD4.T.cells_S05', 'Mast.cells_S01', 'CD8.T.cells_S03',\n",
       "       'NK.cells_S04', 'CD4.T.cells_S03', 'CD8.T.cells_S01',\n",
       "       'B.cells_S02', 'CD4.T.cells_S01', 'Monocytes.and.Macrophages_S05',\n",
       "       'NK.cells_S02', 'NK.cells_S01', 'CD4.T.cells_S02',\n",
       "       'CD8.T.cells_S02', 'NK.cells_S03', 'Dendritic.cells_S01',\n",
       "       'PMNs_S01', 'Monocytes.and.Macrophages_S04', 'B.cells_S01',\n",
       "       'Monocytes.and.Macrophages_S02', 'B.cells_S03',\n",
       "       'Monocytes.and.Macrophages_S03', 'PMNs_S03', 'Dendritic.cells_S02',\n",
       "       'Mast.cells_S02', 'Monocytes.and.Macrophages_S01',\n",
       "       'Epithelial.cells_S02', 'PMNs_S02', 'Fibroblasts_S03',\n",
       "       'Endothelial.cells_S02', 'Epithelial.cells_S03', 'Fibroblasts_S01',\n",
       "       'Endothelial.cells_S01', 'Endothelial.cells_S04',\n",
       "       'Fibroblasts_S04', 'Epithelial.cells_S01', 'Endothelial.cells_S05',\n",
       "       'Mast.cells_S03', 'Mast.cells_S04', 'Fibroblasts_S02',\n",
       "       'Endothelial.cells_S03'], dtype=object)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.ecoRecovery.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "5f9f9142-0db7-4ed4-8bbb-b906bd4fa659",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 488236 × 33336\n",
       "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'percent.mt', 'assigned_cell_type', 'stage', 'Ecotype', 'ecoRecovery'\n",
       "    var: 'features'\n",
       "    obsm: 'X_tsne', 'X_umap'\n",
       "    layers: 'data'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "id": "315432b2-01ee-42b4-a6a8-67090904b349",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "244372.54s - pydevd: Sending message related to process being replaced timed-out after 5 seconds\n",
      "/usr/bin/sh: 1: /root/datos/maestria/netopaas//Deng2024/dengRNA.h5ad: Permission denied\n"
     ]
    }
   ],
   "source": [
    "! {data_dir}/Deng2024/dengRNA.h5ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "id": "5994f8d5-c661-4401-95a3-bdb3e3523d7b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "244384.51s - pydevd: Sending message related to process being replaced timed-out after 5 seconds\n",
      "/usr/bin/sh: 1: /root/datos/maestria/netopaas//Deng2024/.h5ad: not found\n"
     ]
    }
   ],
   "source": [
    "! {data_dir}/Deng2024/deng.h5ad"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b60d5673-27f4-4ab8-b551-6b84c3ba0aa9",
   "metadata": {},
   "source": [
    "## Alexandra 2023"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "1e7ee161-c989-4081-a668-f06834bef03d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/root/datos/maestria/netopaas//alexandra2023/data.zip',\n",
       " <http.client.HTTPMessage at 0x7f7325a5cf10>)"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "urllib.request.urlretrieve('https://zenodo.org/records/7852154/files/data_code.zip?download=1',f'{data_dir}/Alexandra2023/data.zip')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5628b30-65fc-40ab-abc2-7a9d1a8190b8",
   "metadata": {},
   "source": [
    "## Zhong 2024\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "cc7923b6-6f52-44f5-9135-377b38a3aa69",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>percent.mt</th>\n",
       "      <th>sampleID</th>\n",
       "      <th>cellID</th>\n",
       "      <th>RNA_snn_res.0.8</th>\n",
       "      <th>seurat_clusters</th>\n",
       "      <th>major.cell.type</th>\n",
       "      <th>RNA_snn_res.1</th>\n",
       "      <th>...</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Age</th>\n",
       "      <th>Histology</th>\n",
       "      <th>Cycles</th>\n",
       "      <th>PD1</th>\n",
       "      <th>Pathological Response</th>\n",
       "      <th>EGFR</th>\n",
       "      <th>Pathological Response Rate</th>\n",
       "      <th>Smoking_History</th>\n",
       "      <th>PD-L1 TPS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P343</td>\n",
       "      <td>17606</td>\n",
       "      <td>4574</td>\n",
       "      <td>2.891060</td>\n",
       "      <td>P343</td>\n",
       "      <td>P343_AAACCTGAGCTATGCT-1</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>Fibro</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>F</td>\n",
       "      <td>60</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>L858R</td>\n",
       "      <td>0.85</td>\n",
       "      <td>N</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P343</td>\n",
       "      <td>9542</td>\n",
       "      <td>3230</td>\n",
       "      <td>4.003354</td>\n",
       "      <td>P343</td>\n",
       "      <td>P343_AAACCTGCATGATCCA-1</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>Fibro</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>F</td>\n",
       "      <td>60</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>L858R</td>\n",
       "      <td>0.85</td>\n",
       "      <td>N</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P343</td>\n",
       "      <td>10729</td>\n",
       "      <td>3832</td>\n",
       "      <td>2.302172</td>\n",
       "      <td>P343</td>\n",
       "      <td>P343_AAACCTGGTGAGTATA-1</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>Endo</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>F</td>\n",
       "      <td>60</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>L858R</td>\n",
       "      <td>0.85</td>\n",
       "      <td>N</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P343</td>\n",
       "      <td>7788</td>\n",
       "      <td>2859</td>\n",
       "      <td>5.225989</td>\n",
       "      <td>P343</td>\n",
       "      <td>P343_AAACCTGTCCAGTATG-1</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>Endo</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>F</td>\n",
       "      <td>60</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>L858R</td>\n",
       "      <td>0.85</td>\n",
       "      <td>N</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P343</td>\n",
       "      <td>15449</td>\n",
       "      <td>4074</td>\n",
       "      <td>3.301185</td>\n",
       "      <td>P343</td>\n",
       "      <td>P343_AAACGGGCACGACGAA-1</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>Endo</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>F</td>\n",
       "      <td>60</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>L858R</td>\n",
       "      <td>0.85</td>\n",
       "      <td>N</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78686</th>\n",
       "      <td>P591</td>\n",
       "      <td>6608</td>\n",
       "      <td>2528</td>\n",
       "      <td>3.344431</td>\n",
       "      <td>P591</td>\n",
       "      <td>P591_TGCCCTAGTCAGATAA-1</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>Epi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>51</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>KDD</td>\n",
       "      <td>0.58</td>\n",
       "      <td>Y</td>\n",
       "      <td>&lt;1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78687</th>\n",
       "      <td>P591</td>\n",
       "      <td>1064</td>\n",
       "      <td>707</td>\n",
       "      <td>7.518797</td>\n",
       "      <td>P591</td>\n",
       "      <td>P591_TGGCTGGAGCGTTGCC-1</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>Epi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>51</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>KDD</td>\n",
       "      <td>0.58</td>\n",
       "      <td>Y</td>\n",
       "      <td>&lt;1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78688</th>\n",
       "      <td>P591</td>\n",
       "      <td>13226</td>\n",
       "      <td>3318</td>\n",
       "      <td>1.126569</td>\n",
       "      <td>P591</td>\n",
       "      <td>P591_TTCTCCTAGTGCGTGA-1</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>Epi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>51</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>KDD</td>\n",
       "      <td>0.58</td>\n",
       "      <td>Y</td>\n",
       "      <td>&lt;1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78689</th>\n",
       "      <td>P591</td>\n",
       "      <td>7335</td>\n",
       "      <td>2912</td>\n",
       "      <td>6.693933</td>\n",
       "      <td>P591</td>\n",
       "      <td>P591_TTTCCTCAGCGTAATA-1</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>Epi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>51</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>KDD</td>\n",
       "      <td>0.58</td>\n",
       "      <td>Y</td>\n",
       "      <td>&lt;1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78690</th>\n",
       "      <td>P591</td>\n",
       "      <td>14498</td>\n",
       "      <td>4608</td>\n",
       "      <td>1.448476</td>\n",
       "      <td>P591</td>\n",
       "      <td>P591_TTTGGTTCATGGGACA-1</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>Epi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>51</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>Sintilimab</td>\n",
       "      <td>non-MPR</td>\n",
       "      <td>KDD</td>\n",
       "      <td>0.58</td>\n",
       "      <td>Y</td>\n",
       "      <td>&lt;1%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78691 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      orig.ident  nCount_RNA  nFeature_RNA  percent.mt sampleID  \\\n",
       "0           P343       17606          4574    2.891060     P343   \n",
       "1           P343        9542          3230    4.003354     P343   \n",
       "2           P343       10729          3832    2.302172     P343   \n",
       "3           P343        7788          2859    5.225989     P343   \n",
       "4           P343       15449          4074    3.301185     P343   \n",
       "...          ...         ...           ...         ...      ...   \n",
       "78686       P591        6608          2528    3.344431     P591   \n",
       "78687       P591        1064           707    7.518797     P591   \n",
       "78688       P591       13226          3318    1.126569     P591   \n",
       "78689       P591        7335          2912    6.693933     P591   \n",
       "78690       P591       14498          4608    1.448476     P591   \n",
       "\n",
       "                        cellID  RNA_snn_res.0.8  seurat_clusters  \\\n",
       "0      P343_AAACCTGAGCTATGCT-1                6                6   \n",
       "1      P343_AAACCTGCATGATCCA-1                8                8   \n",
       "2      P343_AAACCTGGTGAGTATA-1               10               10   \n",
       "3      P343_AAACCTGTCCAGTATG-1               10               10   \n",
       "4      P343_AAACGGGCACGACGAA-1               10               10   \n",
       "...                        ...              ...              ...   \n",
       "78686  P591_TGCCCTAGTCAGATAA-1                7                7   \n",
       "78687  P591_TGGCTGGAGCGTTGCC-1                7                7   \n",
       "78688  P591_TTCTCCTAGTGCGTGA-1                7                7   \n",
       "78689  P591_TTTCCTCAGCGTAATA-1                7                7   \n",
       "78690  P591_TTTGGTTCATGGGACA-1               20               20   \n",
       "\n",
       "      major.cell.type  RNA_snn_res.1  ... Gender  Age Histology Cycles  \\\n",
       "0               Fibro            NaN  ...      F   60      LUAD      3   \n",
       "1               Fibro            NaN  ...      F   60      LUAD      3   \n",
       "2                Endo            NaN  ...      F   60      LUAD      3   \n",
       "3                Endo            NaN  ...      F   60      LUAD      3   \n",
       "4                Endo            NaN  ...      F   60      LUAD      3   \n",
       "...               ...            ...  ...    ...  ...       ...    ...   \n",
       "78686             Epi            NaN  ...      M   51      LUAD      3   \n",
       "78687             Epi            NaN  ...      M   51      LUAD      3   \n",
       "78688             Epi            NaN  ...      M   51      LUAD      3   \n",
       "78689             Epi            NaN  ...      M   51      LUAD      3   \n",
       "78690             Epi            NaN  ...      M   51      LUAD      3   \n",
       "\n",
       "              PD1 Pathological Response   EGFR Pathological Response Rate  \\\n",
       "0      Sintilimab               non-MPR  L858R                       0.85   \n",
       "1      Sintilimab               non-MPR  L858R                       0.85   \n",
       "2      Sintilimab               non-MPR  L858R                       0.85   \n",
       "3      Sintilimab               non-MPR  L858R                       0.85   \n",
       "4      Sintilimab               non-MPR  L858R                       0.85   \n",
       "...           ...                   ...    ...                        ...   \n",
       "78686  Sintilimab               non-MPR    KDD                       0.58   \n",
       "78687  Sintilimab               non-MPR    KDD                       0.58   \n",
       "78688  Sintilimab               non-MPR    KDD                       0.58   \n",
       "78689  Sintilimab               non-MPR    KDD                       0.58   \n",
       "78690  Sintilimab               non-MPR    KDD                       0.58   \n",
       "\n",
       "      Smoking_History PD-L1 TPS  \n",
       "0                   N      0.01  \n",
       "1                   N      0.01  \n",
       "2                   N      0.01  \n",
       "3                   N      0.01  \n",
       "4                   N      0.01  \n",
       "...               ...       ...  \n",
       "78686               Y       <1%  \n",
       "78687               Y       <1%  \n",
       "78688               Y       <1%  \n",
       "78689               Y       <1%  \n",
       "78690               Y       <1%  \n",
       "\n",
       "[78691 rows x 23 columns]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zhong_samples = pd.read_csv(f'{data_dir}/Zhong2024/GSE241934_IIT_Meta.txt', sep='\\t')\n",
    "zhong_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c04af7d-8683-4fbb-992c-25da5ea97805",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
